{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "47e2a59f-a59f-489a-a2e7-e8a4a206e397",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!wget http://files.ipd.uw.edu/pub/RFdiffusion/6f5902ac237024bdd0c176cb93063dc4/Base_ckpt.pt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a3f5615c-bc69-4737-9363-0545f7c71504",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4e8d798c-e8b8-4ee8-bb7a-87ca19223a20",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import os, time, pickle\n",
    "import torch\n",
    "from omegaconf import OmegaConf\n",
    "from omegaconf import DictConfig\n",
    "import hydra\n",
    "import logging\n",
    "import numpy as np\n",
    "import random\n",
    "import glob\n",
    "\n",
    "from proteome.models.design.rfdiffusion.util import writepdb_multi, writepdb\n",
    "from proteome.models.design.rfdiffusion.inference import utils as iu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2ae165e1-b171-413e-8173-5d417fb97394",
   "metadata": {},
   "outputs": [],
   "source": [
    "from proteome.models.design.rfdiffusion import config\n",
    "from dataclasses import asdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "94b66275-a850-4432-bf36-30b366f35ef4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_deterministic(seed=0):\n",
    "    torch.manual_seed(seed)\n",
    "    np.random.seed(seed)\n",
    "    random.seed(seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8307578f-9ac6-443b-a868-1fb9b90290ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "device = torch.cuda.current_device()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8a5390fc-611a-4e30-95f4-24d5e9a7123c",
   "metadata": {},
   "outputs": [],
   "source": [
    "make_deterministic()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "22476200-cb35-48c2-af2d-be72ccbdbcb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "cfg = config.RFDiffusionConfig(inference=config.InferenceConfig(input_pdb=\"1qys.pdb\", model_directory_path=\".\"), contigmap=config.ContigMap(contigs=\"90-91\"))\n",
    "conf = DictConfig(asdict(cfg))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "53f4f3ed-d287-4d34-8386-77b8a1245ba3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading models from .\n",
      "This is inf_conf.ckpt_path\n",
      "./Base_ckpt.pt\n",
      "Assembling -model, -diffuser and -preprocess configs from checkpoint\n",
      "USING MODEL CONFIG: self._conf[model][n_extra_block] = 4\n",
      "USING MODEL CONFIG: self._conf[model][n_main_block] = 32\n",
      "USING MODEL CONFIG: self._conf[model][n_ref_block] = 4\n",
      "USING MODEL CONFIG: self._conf[model][d_msa] = 256\n",
      "USING MODEL CONFIG: self._conf[model][d_msa_full] = 64\n",
      "USING MODEL CONFIG: self._conf[model][d_pair] = 128\n",
      "USING MODEL CONFIG: self._conf[model][d_templ] = 64\n",
      "USING MODEL CONFIG: self._conf[model][n_head_msa] = 8\n",
      "USING MODEL CONFIG: self._conf[model][n_head_pair] = 4\n",
      "USING MODEL CONFIG: self._conf[model][n_head_templ] = 4\n",
      "USING MODEL CONFIG: self._conf[model][d_hidden] = 32\n",
      "USING MODEL CONFIG: self._conf[model][d_hidden_templ] = 32\n",
      "USING MODEL CONFIG: self._conf[model][p_drop] = 0.15\n",
      "USING MODEL CONFIG: self._conf[model][SE3_param_full] = {'num_layers': 1, 'num_channels': 32, 'num_degrees': 2, 'n_heads': 4, 'div': 4, 'l0_in_features': 8, 'l0_out_features': 8, 'l1_in_features': 3, 'l1_out_features': 2, 'num_edge_features': 32}\n",
      "USING MODEL CONFIG: self._conf[model][SE3_param_topk] = {'num_layers': 1, 'num_channels': 32, 'num_degrees': 2, 'n_heads': 4, 'div': 4, 'l0_in_features': 64, 'l0_out_features': 64, 'l1_in_features': 3, 'l1_out_features': 2, 'num_edge_features': 64}\n",
      "USING MODEL CONFIG: self._conf[model][freeze_track_motif] = False\n",
      "USING MODEL CONFIG: self._conf[model][use_motif_timestep] = True\n",
      "USING MODEL CONFIG: self._conf[diffuser][T] = 50\n",
      "USING MODEL CONFIG: self._conf[diffuser][b_0] = 0.01\n",
      "USING MODEL CONFIG: self._conf[diffuser][b_T] = 0.07\n",
      "USING MODEL CONFIG: self._conf[diffuser][schedule_type] = linear\n",
      "USING MODEL CONFIG: self._conf[diffuser][so3_type] = igso3\n",
      "USING MODEL CONFIG: self._conf[diffuser][crd_scale] = 0.25\n",
      "USING MODEL CONFIG: self._conf[diffuser][so3_schedule_type] = linear\n",
      "USING MODEL CONFIG: self._conf[diffuser][min_b] = 1.5\n",
      "USING MODEL CONFIG: self._conf[diffuser][max_b] = 2.5\n",
      "USING MODEL CONFIG: self._conf[diffuser][min_sigma] = 0.02\n",
      "USING MODEL CONFIG: self._conf[diffuser][max_sigma] = 1.5\n",
      "USING MODEL CONFIG: self._conf[preprocess][sidechain_input] = False\n",
      "USING MODEL CONFIG: self._conf[preprocess][motif_sidechain_input] = True\n",
      "USING MODEL CONFIG: self._conf[preprocess][d_t1d] = 22\n",
      "USING MODEL CONFIG: self._conf[preprocess][d_t2d] = 44\n",
      "USING MODEL CONFIG: self._conf[preprocess][prob_self_cond] = 0.5\n",
      "USING MODEL CONFIG: self._conf[preprocess][str_self_cond] = True\n",
      "USING MODEL CONFIG: self._conf[preprocess][predict_previous] = False\n",
      "Successful diffuser __init__\n"
     ]
    }
   ],
   "source": [
    "sampler = iu.sampler_selector(conf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "5f7d8c5a-3a3d-459f-972f-ad53f3e2967c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "59808046"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum([p.numel() for p in sampler.model.parameters()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "5474fb61-935f-44f4-8626-610546e2fb5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "design_startnum = sampler.inf_conf.design_startnum\n",
    "if sampler.inf_conf.design_startnum == -1:\n",
    "    existing = glob.glob(sampler.inf_conf.output_prefix + \"*.pdb\")\n",
    "    indices = [-1]\n",
    "    for e in existing:\n",
    "        print(e)\n",
    "        m = re.match(\".*_(\\d+)\\.pdb$\", e)\n",
    "        print(m)\n",
    "        if not m:\n",
    "            continue\n",
    "        m = m.groups()[0]\n",
    "        indices.append(int(m))\n",
    "    design_startnum = max(indices) + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "4ba97034-7fef-45bb-9b2f-915b42ad5b8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "With this beta schedule (linear schedule, beta_0 = 0.04, beta_T = 0.28), alpha_bar_T = 0.00013696050154976547\n",
      "With this beta schedule (linear schedule, beta_0 = 0.04, beta_T = 0.28), alpha_bar_T = 0.00013696050154976547\n",
      "With this beta schedule (linear schedule, beta_0 = 0.04, beta_T = 0.28), alpha_bar_T = 0.00013696050154976547\n",
      "With this beta schedule (linear schedule, beta_0 = 0.04, beta_T = 0.28), alpha_bar_T = 0.00013696050154976547\n",
      "With this beta schedule (linear schedule, beta_0 = 0.04, beta_T = 0.28), alpha_bar_T = 0.00013696050154976547\n",
      "With this beta schedule (linear schedule, beta_0 = 0.04, beta_T = 0.28), alpha_bar_T = 0.00013696050154976547\n",
      "With this beta schedule (linear schedule, beta_0 = 0.04, beta_T = 0.28), alpha_bar_T = 0.00013696050154976547\n",
      "With this beta schedule (linear schedule, beta_0 = 0.04, beta_T = 0.28), alpha_bar_T = 0.00013696050154976547\n",
      "With this beta schedule (linear schedule, beta_0 = 0.04, beta_T = 0.28), alpha_bar_T = 0.00013696050154976547\n",
      "With this beta schedule (linear schedule, beta_0 = 0.04, beta_T = 0.28), alpha_bar_T = 0.00013696050154976547\n"
     ]
    }
   ],
   "source": [
    "for i_des in range(design_startnum, design_startnum + sampler.inf_conf.num_designs):\n",
    "    if conf.inference.deterministic:\n",
    "        make_deterministic(i_des)\n",
    "\n",
    "    start_time = time.time()\n",
    "    out_prefix = f\"{sampler.inf_conf.output_prefix}_{i_des}\"\n",
    "\n",
    "    x_init, seq_init = sampler.sample_init()\n",
    "    denoised_xyz_stack = []\n",
    "    px0_xyz_stack = []\n",
    "    seq_stack = []\n",
    "    plddt_stack = []\n",
    "\n",
    "    x_t = torch.clone(x_init)\n",
    "    seq_t = torch.clone(seq_init)\n",
    "    # Loop over number of reverse diffusion time steps.\n",
    "    for t in range(int(sampler.t_step_input), sampler.inf_conf.final_step - 1, -1):\n",
    "        px0, x_t, seq_t, plddt = sampler.sample_step(\n",
    "            t=t, x_t=x_t, seq_init=seq_t, final_step=sampler.inf_conf.final_step\n",
    "        )\n",
    "        px0_xyz_stack.append(px0)\n",
    "        denoised_xyz_stack.append(x_t)\n",
    "        seq_stack.append(seq_t)\n",
    "        plddt_stack.append(plddt[0])  # remove singleton leading dimension\n",
    "\n",
    "    # Flip order for better visualization in pymol\n",
    "    denoised_xyz_stack = torch.stack(denoised_xyz_stack)\n",
    "    denoised_xyz_stack = torch.flip(\n",
    "        denoised_xyz_stack,\n",
    "        [\n",
    "            0,\n",
    "        ],\n",
    "    )\n",
    "    px0_xyz_stack = torch.stack(px0_xyz_stack)\n",
    "    px0_xyz_stack = torch.flip(\n",
    "        px0_xyz_stack,\n",
    "        [\n",
    "            0,\n",
    "        ],\n",
    "    )\n",
    "\n",
    "    # For logging -- don't flip\n",
    "    plddt_stack = torch.stack(plddt_stack)\n",
    "\n",
    "    # Save outputs\n",
    "    os.makedirs(os.path.dirname(out_prefix), exist_ok=True)\n",
    "    final_seq = seq_stack[-1]\n",
    "\n",
    "    # Output glycines, except for motif region\n",
    "    final_seq = torch.where(\n",
    "        torch.argmax(seq_init, dim=-1) == 21, 7, torch.argmax(seq_init, dim=-1)\n",
    "    )  # 7 is glycine\n",
    "\n",
    "    bfacts = torch.ones_like(final_seq.squeeze())\n",
    "    # make bfact=0 for diffused coordinates\n",
    "    bfacts[torch.where(torch.argmax(seq_init, dim=-1) == 21, True, False)] = 0\n",
    "    # pX0 last step\n",
    "    out = f\"{out_prefix}.pdb\"\n",
    "\n",
    "    # Now don't output sidechains\n",
    "    writepdb(\n",
    "        out,\n",
    "        denoised_xyz_stack[0, :, :4],\n",
    "        final_seq,\n",
    "        sampler.binderlen,\n",
    "        chain_idx=sampler.chain_idx,\n",
    "        bfacts=bfacts,\n",
    "    )\n",
    "\n",
    "    # run metadata\n",
    "    trb = dict(\n",
    "        config=OmegaConf.to_container(sampler._conf, resolve=True),\n",
    "        plddt=plddt_stack.cpu().numpy(),\n",
    "        device=torch.cuda.get_device_name(torch.cuda.current_device())\n",
    "        if torch.cuda.is_available()\n",
    "        else \"CPU\",\n",
    "        time=time.time() - start_time,\n",
    "    )\n",
    "    if hasattr(sampler, \"contig_map\"):\n",
    "        for key, value in sampler.contig_map.get_mappings().items():\n",
    "            trb[key] = value\n",
    "    with open(f\"{out_prefix}.trb\", \"wb\") as f_out:\n",
    "        pickle.dump(trb, f_out)\n",
    "\n",
    "    if sampler.inf_conf.write_trajectory:\n",
    "        # trajectory pdbs\n",
    "        traj_prefix = (\n",
    "            os.path.dirname(out_prefix) + \"/traj/\" + os.path.basename(out_prefix)\n",
    "        )\n",
    "        os.makedirs(os.path.dirname(traj_prefix), exist_ok=True)\n",
    "\n",
    "        out = f\"{traj_prefix}_Xt-1_traj.pdb\"\n",
    "        writepdb_multi(\n",
    "            out,\n",
    "            denoised_xyz_stack,\n",
    "            bfacts,\n",
    "            final_seq.squeeze(),\n",
    "            use_hydrogens=False,\n",
    "            backbone_only=False,\n",
    "            chain_ids=sampler.chain_idx,\n",
    "        )\n",
    "\n",
    "        out = f\"{traj_prefix}_pX0_traj.pdb\"\n",
    "        writepdb_multi(\n",
    "            out,\n",
    "            px0_xyz_stack,\n",
    "            bfacts,\n",
    "            final_seq.squeeze(),\n",
    "            use_hydrogens=False,\n",
    "            backbone_only=False,\n",
    "            chain_ids=sampler.chain_idx,\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "42a1e581-3c84-4df5-86dc-e20aad2e86e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import py3Dmol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "50508be2-0f3d-46ae-98c6-2efb273c60d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"1qys.pdb\", mode=\"r\") as f:\n",
    "    init_pdb = f.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "bbfba248-6e92-440a-b00d-ba4b5ab7fd8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_16908637293794932\"  style=\"position: relative; width: 800px; height: 600px;\">\n        <p id=\"3dmolwarning_16908637293794932\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    //this is to ignore the existence of requirejs amd\n    var savedexports, savedmodule;\n    if (typeof exports !== 'undefined') savedexports = exports;\n    else exports = {}\n    if (typeof module !== 'undefined') savedmodule = module;\n    else module = {}\n\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n        exports = savedexports;\n        module = savedmodule;\n        resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://cdnjs.cloudflare.com/ajax/libs/3Dmol/2.0.3/3Dmol-min.js');\n}\n\nvar viewer_16908637293794932 = null;\nvar warn = document.getElementById(\"3dmolwarning_16908637293794932\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16908637293794932 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_16908637293794932\"),{backgroundColor:\"white\"});\nviewer_16908637293794932.zoomTo();\n\tviewer_16908637293794932.addModelsAsFrames(\"HEADER    DE NOVO PROTEIN                         11-SEP-03   1QYS              \\nTITLE     CRYSTAL STRUCTURE OF TOP7: A COMPUTATIONALLY DESIGNED                 \\nTITLE    2 PROTEIN WITH A NOVEL FOLD                                            \\nCOMPND    MOL_ID: 1;                                                            \\nCOMPND   2 MOLECULE: TOP7;                                                      \\nCOMPND   3 CHAIN: A;                                                            \\nCOMPND   4 ENGINEERED: YES                                                      \\nSOURCE    MOL_ID: 1;                                                            \\nSOURCE   2 ORGANISM_SCIENTIFIC: COMPUTATIONALLY DESIGNED SEQUENCE;              \\nSOURCE   3 EXPRESSION_SYSTEM: ESCHERICHIA COLI;                                 \\nSOURCE   4 EXPRESSION_SYSTEM_TAXID: 562;                                        \\nSOURCE   5 EXPRESSION_SYSTEM_STRAIN: BL21(DE3) PLYSS;                           \\nSOURCE   6 EXPRESSION_SYSTEM_VECTOR_TYPE: PLASMID;                              \\nSOURCE   7 EXPRESSION_SYSTEM_PLASMID: PET29B(+)                                 \\nKEYWDS    ALPHA-BETA, COMPUTATIONALLY DESIGNED, NOVEL FOLD, DE NOVO             \\nKEYWDS   2 PROTEIN                                                              \\nEXPDTA    X-RAY DIFFRACTION                                                     \\nAUTHOR    B.KUHLMAN,G.DANTAS,G.C.IRETON,G.VARANI,B.L.STODDARD,D.BAKER           \\nREVDAT   2   24-FEB-09 1QYS    1       VERSN                                    \\nREVDAT   1   25-NOV-03 1QYS    0                                                \\nJRNL        AUTH   B.KUHLMAN,G.DANTAS,G.C.IRETON,G.VARANI,                      \\nJRNL        AUTH 2 B.L.STODDARD,D.BAKER                                         \\nJRNL        TITL   DESIGN OF A NOVEL GLOBULAR PROTEIN FOLD WITH                 \\nJRNL        TITL 2 ATOMIC-LEVEL ACCURACY                                        \\nJRNL        REF    SCIENCE                       V. 302  1364 2003              \\nJRNL        REFN                   ISSN 0036-8075                               \\nJRNL        PMID   14631033                                                     \\nJRNL        DOI    10.1126/SCIENCE.1089427                                      \\nREMARK   1                                                                      \\nREMARK   2                                                                      \\nREMARK   2 RESOLUTION.    2.50 ANGSTROMS.                                       \\nREMARK   3                                                                      \\nREMARK   3 REFINEMENT.                                                          \\nREMARK   3   PROGRAM     : CNS 1.1                                              \\nREMARK   3   AUTHORS     : BRUNGER,ADAMS,CLORE,DELANO,GROS,GROSSE-              \\nREMARK   3               : KUNSTLEVE,JIANG,KUSZEWSKI,NILGES, PANNU,             \\nREMARK   3               : READ,RICE,SIMONSON,WARREN                            \\nREMARK   3                                                                      \\nREMARK   3  REFINEMENT TARGET : ENGH & HUBER                                    \\nREMARK   3                                                                      \\nREMARK   3  DATA USED IN REFINEMENT.                                            \\nREMARK   3   RESOLUTION RANGE HIGH (ANGSTROMS) : 2.50                           \\nREMARK   3   RESOLUTION RANGE LOW  (ANGSTROMS) : 18.71                          \\nREMARK   3   DATA CUTOFF            (SIGMA(F)) : 0.000                          \\nREMARK   3   DATA CUTOFF HIGH         (ABS(F)) : 1873860.400                    \\nREMARK   3   DATA CUTOFF LOW          (ABS(F)) : 0.0000                         \\nREMARK   3   COMPLETENESS (WORKING+TEST)   (%) : 96.1                           \\nREMARK   3   NUMBER OF REFLECTIONS             : 6736                           \\nREMARK   3                                                                      \\nREMARK   3  FIT TO DATA USED IN REFINEMENT.                                     \\nREMARK   3   CROSS-VALIDATION METHOD          : THROUGHOUT                      \\nREMARK   3   FREE R VALUE TEST SET SELECTION  : RANDOM                          \\nREMARK   3   R VALUE            (WORKING SET) : 0.268                           \\nREMARK   3   FREE R VALUE                     : 0.293                           \\nREMARK   3   FREE R VALUE TEST SET SIZE   (%) : 5.100                           \\nREMARK   3   FREE R VALUE TEST SET COUNT      : 346                             \\nREMARK   3   ESTIMATED ERROR OF FREE R VALUE  : 0.016                           \\nREMARK   3                                                                      \\nREMARK   3  FIT IN THE HIGHEST RESOLUTION BIN.                                  \\nREMARK   3   TOTAL NUMBER OF BINS USED           : 6                            \\nREMARK   3   BIN RESOLUTION RANGE HIGH       (A) : 2.50                         \\nREMARK   3   BIN RESOLUTION RANGE LOW        (A) : 2.66                         \\nREMARK   3   BIN COMPLETENESS (WORKING+TEST) (%) : 91.30                        \\nREMARK   3   REFLECTIONS IN BIN    (WORKING SET) : 1009                         \\nREMARK   3   BIN R VALUE           (WORKING SET) : 0.3530                       \\nREMARK   3   BIN FREE R VALUE                    : 0.3700                       \\nREMARK   3   BIN FREE R VALUE TEST SET SIZE  (%) : 4.30                         \\nREMARK   3   BIN FREE R VALUE TEST SET COUNT     : 45                           \\nREMARK   3   ESTIMATED ERROR OF BIN FREE R VALUE : 0.055                        \\nREMARK   3                                                                      \\nREMARK   3  NUMBER OF NON-HYDROGEN ATOMS USED IN REFINEMENT.                    \\nREMARK   3   PROTEIN ATOMS            : 685                                     \\nREMARK   3   NUCLEIC ACID ATOMS       : 0                                       \\nREMARK   3   HETEROGEN ATOMS          : 0                                       \\nREMARK   3   SOLVENT ATOMS            : 7                                       \\nREMARK   3                                                                      \\nREMARK   3  B VALUES.                                                           \\nREMARK   3   FROM WILSON PLOT           (A**2) : 30.20                          \\nREMARK   3   MEAN B VALUE      (OVERALL, A**2) : 65.50                          \\nREMARK   3   OVERALL ANISOTROPIC B VALUE.                                       \\nREMARK   3    B11 (A**2) : 10.54000                                             \\nREMARK   3    B22 (A**2) : 10.54000                                             \\nREMARK   3    B33 (A**2) : -21.07000                                            \\nREMARK   3    B12 (A**2) : 9.56000                                              \\nREMARK   3    B13 (A**2) : 0.00000                                              \\nREMARK   3    B23 (A**2) : 0.00000                                              \\nREMARK   3                                                                      \\nREMARK   3  ESTIMATED COORDINATE ERROR.                                         \\nREMARK   3   ESD FROM LUZZATI PLOT        (A) : 0.42                            \\nREMARK   3   ESD FROM SIGMAA              (A) : 0.47                            \\nREMARK   3   LOW RESOLUTION CUTOFF        (A) : 5.00                            \\nREMARK   3                                                                      \\nREMARK   3  CROSS-VALIDATED ESTIMATED COORDINATE ERROR.                         \\nREMARK   3   ESD FROM C-V LUZZATI PLOT    (A) : 0.46                            \\nREMARK   3   ESD FROM C-V SIGMAA          (A) : 0.45                            \\nREMARK   3                                                                      \\nREMARK   3  RMS DEVIATIONS FROM IDEAL VALUES.                                   \\nREMARK   3   BOND LENGTHS                 (A) : 0.008                           \\nREMARK   3   BOND ANGLES            (DEGREES) : 1.40                            \\nREMARK   3   DIHEDRAL ANGLES        (DEGREES) : 26.30                           \\nREMARK   3   IMPROPER ANGLES        (DEGREES) : 0.73                            \\nREMARK   3                                                                      \\nREMARK   3  ISOTROPIC THERMAL MODEL : GROUP                                     \\nREMARK   3                                                                      \\nREMARK   3  ISOTROPIC THERMAL FACTOR RESTRAINTS.    RMS    SIGMA                \\nREMARK   3   MAIN-CHAIN BOND              (A**2) : NULL  ; NULL                 \\nREMARK   3   MAIN-CHAIN ANGLE             (A**2) : NULL  ; NULL                 \\nREMARK   3   SIDE-CHAIN BOND              (A**2) : NULL  ; NULL                 \\nREMARK   3   SIDE-CHAIN ANGLE             (A**2) : NULL  ; NULL                 \\nREMARK   3                                                                      \\nREMARK   3  BULK SOLVENT MODELING.                                              \\nREMARK   3   METHOD USED : FLAT MODEL                                           \\nREMARK   3   KSOL        : 0.30                                                 \\nREMARK   3   BSOL        : 55.24                                                \\nREMARK   3                                                                      \\nREMARK   3  NCS MODEL : NULL                                                    \\nREMARK   3                                                                      \\nREMARK   3  NCS RESTRAINTS.                         RMS   SIGMA/WEIGHT          \\nREMARK   3   GROUP  1  POSITIONAL            (A) : NULL  ; NULL                 \\nREMARK   3   GROUP  1  B-FACTOR           (A**2) : NULL  ; NULL                 \\nREMARK   3                                                                      \\nREMARK   3  PARAMETER FILE  1  : PROTEIN_REP.PARAM                              \\nREMARK   3  PARAMETER FILE  2  : WATER_REP.PARAM                                \\nREMARK   3  PARAMETER FILE  3  : NULL                                           \\nREMARK   3  TOPOLOGY FILE  1   : PROTEIN.TOP                                    \\nREMARK   3  TOPOLOGY FILE  2   : WATER.TOP                                      \\nREMARK   3  TOPOLOGY FILE  3   : NULL                                           \\nREMARK   3                                                                      \\nREMARK   3  OTHER REFINEMENT REMARKS: NULL                                      \\nREMARK   4                                                                      \\nREMARK   4 1QYS COMPLIES WITH FORMAT V. 3.15, 01-DEC-08                         \\nREMARK 100                                                                      \\nREMARK 100 THIS ENTRY HAS BEEN PROCESSED BY RCSB ON 12-SEP-03.                  \\nREMARK 100 THE RCSB ID CODE IS RCSB020233.                                      \\nREMARK 200                                                                      \\nREMARK 200 EXPERIMENTAL DETAILS                                                 \\nREMARK 200  EXPERIMENT TYPE                : X-RAY DIFFRACTION                  \\nREMARK 200  DATE OF DATA COLLECTION        : 24-MAR-03                          \\nREMARK 200  TEMPERATURE           (KELVIN) : 100                                \\nREMARK 200  PH                             : 6.6                                \\nREMARK 200  NUMBER OF CRYSTALS USED        : 1                                  \\nREMARK 200                                                                      \\nREMARK 200  SYNCHROTRON              (Y/N) : Y                                  \\nREMARK 200  RADIATION SOURCE               : ALS                                \\nREMARK 200  BEAMLINE                       : 8.2.1                              \\nREMARK 200  X-RAY GENERATOR MODEL          : NULL                               \\nREMARK 200  MONOCHROMATIC OR LAUE    (M/L) : M                                  \\nREMARK 200  WAVELENGTH OR RANGE        (A) : 0.9793                             \\nREMARK 200  MONOCHROMATOR                  : DOUBLE CRYSTAL SI(111)             \\nREMARK 200  OPTICS                         : NULL                               \\nREMARK 200                                                                      \\nREMARK 200  DETECTOR TYPE                  : CCD                                \\nREMARK 200  DETECTOR MANUFACTURER          : ADSC QUANTUM 210                   \\nREMARK 200  INTENSITY-INTEGRATION SOFTWARE : HKL-2000                           \\nREMARK 200  DATA SCALING SOFTWARE          : SCALEPACK                          \\nREMARK 200                                                                      \\nREMARK 200  NUMBER OF UNIQUE REFLECTIONS   : 6979                               \\nREMARK 200  RESOLUTION RANGE HIGH      (A) : 2.500                              \\nREMARK 200  RESOLUTION RANGE LOW       (A) : 50.000                             \\nREMARK 200  REJECTION CRITERIA  (SIGMA(I)) : 0.000                              \\nREMARK 200                                                                      \\nREMARK 200 OVERALL.                                                             \\nREMARK 200  COMPLETENESS FOR RANGE     (%) : 100.0                              \\nREMARK 200  DATA REDUNDANCY                : NULL                               \\nREMARK 200  R MERGE                    (I) : 0.04500                            \\nREMARK 200  R SYM                      (I) : NULL                               \\nREMARK 200  <I/SIGMA(I)> FOR THE DATA SET  : 37.8000                            \\nREMARK 200                                                                      \\nREMARK 200 IN THE HIGHEST RESOLUTION SHELL.                                     \\nREMARK 200  HIGHEST RESOLUTION SHELL, RANGE HIGH (A) : 2.50                     \\nREMARK 200  HIGHEST RESOLUTION SHELL, RANGE LOW  (A) : 2.59                     \\nREMARK 200  COMPLETENESS FOR SHELL     (%) : 100.0                              \\nREMARK 200  DATA REDUNDANCY IN SHELL       : NULL                               \\nREMARK 200  R MERGE FOR SHELL          (I) : 0.34400                            \\nREMARK 200  R SYM FOR SHELL            (I) : NULL                               \\nREMARK 200  <I/SIGMA(I)> FOR SHELL         : 5.000                              \\nREMARK 200                                                                      \\nREMARK 200 DIFFRACTION PROTOCOL: SAD                                            \\nREMARK 200 METHOD USED TO DETERMINE THE STRUCTURE: SAD                          \\nREMARK 200 SOFTWARE USED: CNS                                                   \\nREMARK 200 STARTING MODEL: NULL                                                 \\nREMARK 200                                                                      \\nREMARK 200 REMARK: NULL                                                         \\nREMARK 280                                                                      \\nREMARK 280 CRYSTAL                                                              \\nREMARK 280 SOLVENT CONTENT, VS   (%): 42.94                                     \\nREMARK 280 MATTHEWS COEFFICIENT, VM (ANGSTROMS**3/DA): 2.16                     \\nREMARK 280                                                                      \\nREMARK 280 CRYSTALLIZATION CONDITIONS: 15-20% PEG3350 250MM AMMONIUM            \\nREMARK 280  FORMATE, PH 6.6, VAPOR DIFFUSION, HANGING DROP, STREAK              \\nREMARK 280  SEEDING, TEMPERATURE 298K                                           \\nREMARK 290                                                                      \\nREMARK 290 CRYSTALLOGRAPHIC SYMMETRY                                            \\nREMARK 290 SYMMETRY OPERATORS FOR SPACE GROUP: P 32 2 1                         \\nREMARK 290                                                                      \\nREMARK 290      SYMOP   SYMMETRY                                                \\nREMARK 290     NNNMMM   OPERATOR                                                \\nREMARK 290       1555   X,Y,Z                                                   \\nREMARK 290       2555   -Y,X-Y,Z+2/3                                            \\nREMARK 290       3555   -X+Y,-X,Z+1/3                                           \\nREMARK 290       4555   Y,X,-Z                                                  \\nREMARK 290       5555   X-Y,-Y,-Z+1/3                                           \\nREMARK 290       6555   -X,-X+Y,-Z+2/3                                          \\nREMARK 290                                                                      \\nREMARK 290     WHERE NNN -> OPERATOR NUMBER                                     \\nREMARK 290           MMM -> TRANSLATION VECTOR                                  \\nREMARK 290                                                                      \\nREMARK 290 CRYSTALLOGRAPHIC SYMMETRY TRANSFORMATIONS                            \\nREMARK 290 THE FOLLOWING TRANSFORMATIONS OPERATE ON THE ATOM/HETATM             \\nREMARK 290 RECORDS IN THIS ENTRY TO PRODUCE CRYSTALLOGRAPHICALLY                \\nREMARK 290 RELATED MOLECULES.                                                   \\nREMARK 290   SMTRY1   1  1.000000  0.000000  0.000000        0.00000            \\nREMARK 290   SMTRY2   1  0.000000  1.000000  0.000000        0.00000            \\nREMARK 290   SMTRY3   1  0.000000  0.000000  1.000000        0.00000            \\nREMARK 290   SMTRY1   2 -0.500000 -0.866025  0.000000        0.00000            \\nREMARK 290   SMTRY2   2  0.866025 -0.500000  0.000000        0.00000            \\nREMARK 290   SMTRY3   2  0.000000  0.000000  1.000000       93.70267            \\nREMARK 290   SMTRY1   3 -0.500000  0.866025  0.000000        0.00000            \\nREMARK 290   SMTRY2   3 -0.866025 -0.500000  0.000000        0.00000            \\nREMARK 290   SMTRY3   3  0.000000  0.000000  1.000000       46.85133            \\nREMARK 290   SMTRY1   4 -0.500000  0.866025  0.000000        0.00000            \\nREMARK 290   SMTRY2   4  0.866025  0.500000  0.000000        0.00000            \\nREMARK 290   SMTRY3   4  0.000000  0.000000 -1.000000        0.00000            \\nREMARK 290   SMTRY1   5  1.000000  0.000000  0.000000        0.00000            \\nREMARK 290   SMTRY2   5  0.000000 -1.000000  0.000000        0.00000            \\nREMARK 290   SMTRY3   5  0.000000  0.000000 -1.000000       46.85133            \\nREMARK 290   SMTRY1   6 -0.500000 -0.866025  0.000000        0.00000            \\nREMARK 290   SMTRY2   6 -0.866025  0.500000  0.000000        0.00000            \\nREMARK 290   SMTRY3   6  0.000000  0.000000 -1.000000       93.70267            \\nREMARK 290                                                                      \\nREMARK 290 REMARK: NULL                                                         \\nREMARK 300                                                                      \\nREMARK 300 BIOMOLECULE: 1                                                       \\nREMARK 300 SEE REMARK 350 FOR THE AUTHOR PROVIDED AND/OR PROGRAM                \\nREMARK 300 GENERATED ASSEMBLY INFORMATION FOR THE STRUCTURE IN                  \\nREMARK 300 THIS ENTRY. THE REMARK MAY ALSO PROVIDE INFORMATION ON               \\nREMARK 300 BURIED SURFACE AREA.                                                 \\nREMARK 350                                                                      \\nREMARK 350 COORDINATES FOR A COMPLETE MULTIMER REPRESENTING THE KNOWN           \\nREMARK 350 BIOLOGICALLY SIGNIFICANT OLIGOMERIZATION STATE OF THE                \\nREMARK 350 MOLECULE CAN BE GENERATED BY APPLYING BIOMT TRANSFORMATIONS          \\nREMARK 350 GIVEN BELOW.  BOTH NON-CRYSTALLOGRAPHIC AND                          \\nREMARK 350 CRYSTALLOGRAPHIC OPERATIONS ARE GIVEN.                               \\nREMARK 350                                                                      \\nREMARK 350 BIOMOLECULE: 1                                                       \\nREMARK 350 AUTHOR DETERMINED BIOLOGICAL UNIT: MONOMERIC                         \\nREMARK 350 APPLY THE FOLLOWING TO CHAINS: A                                     \\nREMARK 350   BIOMT1   1  1.000000  0.000000  0.000000        0.00000            \\nREMARK 350   BIOMT2   1  0.000000  1.000000  0.000000        0.00000            \\nREMARK 350   BIOMT3   1  0.000000  0.000000  1.000000        0.00000            \\nREMARK 465                                                                      \\nREMARK 465 MISSING RESIDUES                                                     \\nREMARK 465 THE FOLLOWING RESIDUES WERE NOT LOCATED IN THE                       \\nREMARK 465 EXPERIMENT. (M=MODEL NUMBER; RES=RESIDUE NAME; C=CHAIN               \\nREMARK 465 IDENTIFIER; SSSEQ=SEQUENCE NUMBER; I=INSERTION CODE.)                \\nREMARK 465                                                                      \\nREMARK 465   M RES C SSSEQI                                                     \\nREMARK 465     MSE A     1                                                      \\nREMARK 465     GLY A     2                                                      \\nREMARK 465     GLU A    95                                                      \\nREMARK 465     GLY A    96                                                      \\nREMARK 465     GLY A    97                                                      \\nREMARK 465     SER A    98                                                      \\nREMARK 465     LEU A    99                                                      \\nREMARK 465     GLU A   100                                                      \\nREMARK 465     HIS A   101                                                      \\nREMARK 465     HIS A   102                                                      \\nREMARK 465     HIS A   103                                                      \\nREMARK 465     HIS A   104                                                      \\nREMARK 465     HIS A   105                                                      \\nREMARK 465     HIS A   106                                                      \\nREMARK 470                                                                      \\nREMARK 470 MISSING ATOM                                                         \\nREMARK 470 THE FOLLOWING RESIDUES HAVE MISSING ATOMS(M=MODEL NUMBER;            \\nREMARK 470 RES=RESIDUE NAME; C=CHAIN IDENTIFIER; SSEQ=SEQUENCE NUMBER;          \\nREMARK 470 I=INSERTION CODE):                                                   \\nREMARK 470   M RES CSSEQI  ATOMS                                                \\nREMARK 470     LYS A  15    CG   CD   CE   NZ                                   \\nREMARK 470     PHE A  17    CG   CD1  CD2  CE1  CE2  CZ                         \\nREMARK 470     SER A  27    OG                                                  \\nREMARK 470     GLN A  30    CG   CD   OE1  NE2                                  \\nREMARK 470     LYS A  31    CG   CD   CE   NZ                                   \\nREMARK 470     ASN A  34    CG   OD1  ND2                                       \\nREMARK 470     LEU A  36    CG   CD1  CD2                                       \\nREMARK 470     LYS A  46    CG   CD   CE   NZ                                   \\nREMARK 470     ARG A  47    CG   CD   NE   CZ   NH1  NH2                        \\nREMARK 470     ARG A  55    CG   CD   NE   CZ   NH1  NH2                        \\nREMARK 470     LYS A  62    CG   CD   CE   NZ                                   \\nREMARK 470     GLU A  73    CG   CD   OE1  OE2                                  \\nREMARK 500                                                                      \\nREMARK 500 GEOMETRY AND STEREOCHEMISTRY                                         \\nREMARK 500 SUBTOPIC: CLOSE CONTACTS                                             \\nREMARK 500                                                                      \\nREMARK 500 THE FOLLOWING ATOMS THAT ARE RELATED BY CRYSTALLOGRAPHIC             \\nREMARK 500 SYMMETRY ARE IN CLOSE CONTACT.  AN ATOM LOCATED WITHIN 0.15          \\nREMARK 500 ANGSTROMS OF A SYMMETRY RELATED ATOM IS ASSUMED TO BE ON A           \\nREMARK 500 SPECIAL POSITION AND IS, THEREFORE, LISTED IN REMARK 375             \\nREMARK 500 INSTEAD OF REMARK 500.  ATOMS WITH NON-BLANK ALTERNATE               \\nREMARK 500 LOCATION INDICATORS ARE NOT INCLUDED IN THE CALCULATIONS.            \\nREMARK 500                                                                      \\nREMARK 500 DISTANCE CUTOFF:                                                     \\nREMARK 500 2.2 ANGSTROMS FOR CONTACTS NOT INVOLVING HYDROGEN ATOMS              \\nREMARK 500 1.6 ANGSTROMS FOR CONTACTS INVOLVING HYDROGEN ATOMS                  \\nREMARK 500                                                                      \\nREMARK 500  ATM1  RES C  SSEQI   ATM2  RES C  SSEQI  SSYMOP   DISTANCE          \\nREMARK 500   ND2  ASN A    80     ND2  ASN A    80     4555     1.96            \\nREMARK 500   CD1  ILE A    68     CD1  ILE A    68     4555     1.98            \\nREMARK 500                                                                      \\nREMARK 500 REMARK: NULL                                                         \\nREMARK 500                                                                      \\nREMARK 500 GEOMETRY AND STEREOCHEMISTRY                                         \\nREMARK 500 SUBTOPIC: TORSION ANGLES                                             \\nREMARK 500                                                                      \\nREMARK 500 TORSION ANGLES OUTSIDE THE EXPECTED RAMACHANDRAN REGIONS:            \\nREMARK 500 (M=MODEL NUMBER; RES=RESIDUE NAME; C=CHAIN IDENTIFIER;               \\nREMARK 500 SSEQ=SEQUENCE NUMBER; I=INSERTION CODE).                             \\nREMARK 500                                                                      \\nREMARK 500 STANDARD TABLE:                                                      \\nREMARK 500 FORMAT:(10X,I3,1X,A3,1X,A1,I4,A1,4X,F7.2,3X,F7.2)                    \\nREMARK 500                                                                      \\nREMARK 500 EXPECTED VALUES: GJ KLEYWEGT AND TA JONES (1996). PHI/PSI-           \\nREMARK 500 CHOLOGY: RAMACHANDRAN REVISITED. STRUCTURE 4, 1395 - 1400            \\nREMARK 500                                                                      \\nREMARK 500  M RES CSSEQI        PSI       PHI                                   \\nREMARK 500    ASP A  12       80.22   -168.83                                   \\nREMARK 500    ASN A  13      102.49     43.86                                   \\nREMARK 500    THR A  24       81.00   -158.71                                   \\nREMARK 500    THR A  25      -36.94   -176.78                                   \\nREMARK 500    GLU A  26       -7.45     68.87                                   \\nREMARK 500                                                                      \\nREMARK 500 REMARK: NULL                                                         \\nDBREF  1QYS A    1   106  PDB    1QYS     1QYS             1    106             \\nSEQRES   1 A  106  MSE GLY ASP ILE GLN VAL GLN VAL ASN ILE ASP ASP ASN          \\nSEQRES   2 A  106  GLY LYS ASN PHE ASP TYR THR TYR THR VAL THR THR GLU          \\nSEQRES   3 A  106  SER GLU LEU GLN LYS VAL LEU ASN GLU LEU MSE ASP TYR          \\nSEQRES   4 A  106  ILE LYS LYS GLN GLY ALA LYS ARG VAL ARG ILE SER ILE          \\nSEQRES   5 A  106  THR ALA ARG THR LYS LYS GLU ALA GLU LYS PHE ALA ALA          \\nSEQRES   6 A  106  ILE LEU ILE LYS VAL PHE ALA GLU LEU GLY TYR ASN ASP          \\nSEQRES   7 A  106  ILE ASN VAL THR PHE ASP GLY ASP THR VAL THR VAL GLU          \\nSEQRES   8 A  106  GLY GLN LEU GLU GLY GLY SER LEU GLU HIS HIS HIS HIS          \\nSEQRES   9 A  106  HIS HIS                                                      \\nMODRES 1QYS MSE A   37  MET  SELENOMETHIONINE                                   \\nHET    MSE  A  37       8                                                       \\nHETNAM     MSE SELENOMETHIONINE                                                 \\nFORMUL   1  MSE    C5 H11 N O2 SE                                               \\nFORMUL   2  HOH   *7(H2 O)                                                      \\nHELIX    1   1 SER A   27  GLY A   44  1                                  18    \\nHELIX    2   2 THR A   56  LEU A   74  1                                  19    \\nSHEET    1   A 5 ASN A  16  VAL A  23  0                                        \\nSHEET    2   A 5 ILE A   4  ASP A  11 -1  N  VAL A   8   O  TYR A  19           \\nSHEET    3   A 5 ARG A  47  THR A  53 -1  O  ARG A  49   N  ASN A   9           \\nSHEET    4   A 5 THR A  87  GLN A  93 -1  O  GLY A  92   N  VAL A  48           \\nSHEET    5   A 5 ASP A  78  ASP A  84 -1  N  THR A  82   O  THR A  89           \\nLINK         C   LEU A  36                 N   MSE A  37     1555   1555  1.33  \\nLINK         C   MSE A  37                 N   ASP A  38     1555   1555  1.33  \\nCRYST1   35.900   35.900  140.554  90.00  90.00 120.00 P 32 2 1      6          \\nORIGX1      1.000000  0.000000  0.000000        0.00000                         \\nORIGX2      0.000000  1.000000  0.000000        0.00000                         \\nORIGX3      0.000000  0.000000  1.000000        0.00000                         \\nSCALE1      0.027855  0.016082  0.000000        0.00000                         \\nSCALE2      0.000000  0.032164  0.000000        0.00000                         \\nSCALE3      0.000000  0.000000  0.007115        0.00000                         \\nATOM      1  N   ASP A   3      -4.522  18.306  17.409  1.00174.51           N  \\nATOM      2  CA  ASP A   3      -3.061  18.228  17.122  1.00174.51           C  \\nATOM      3  C   ASP A   3      -2.664  16.993  16.324  1.00174.51           C  \\nATOM      4  O   ASP A   3      -3.515  16.306  15.754  1.00174.51           O  \\nATOM      5  CB  ASP A   3      -2.261  18.246  18.422  1.00 83.84           C  \\nATOM      6  CG  ASP A   3      -1.658  19.600  18.711  1.00 83.84           C  \\nATOM      7  OD1 ASP A   3      -1.169  20.249  17.760  1.00 83.84           O  \\nATOM      8  OD2 ASP A   3      -1.654  20.007  19.892  1.00 83.84           O  \\nATOM      9  N   ILE A   4      -1.360  16.714  16.297  1.00 57.73           N  \\nATOM     10  CA  ILE A   4      -0.823  15.562  15.568  1.00 57.73           C  \\nATOM     11  C   ILE A   4      -0.721  14.309  16.433  1.00 57.73           C  \\nATOM     12  O   ILE A   4       0.091  14.222  17.355  1.00 57.73           O  \\nATOM     13  CB  ILE A   4       0.555  15.888  14.980  1.00 57.14           C  \\nATOM     14  CG1 ILE A   4       0.425  17.108  14.058  1.00 57.14           C  \\nATOM     15  CG2 ILE A   4       1.097  14.674  14.218  1.00 57.14           C  \\nATOM     16  CD1 ILE A   4       1.737  17.831  13.766  1.00 57.14           C  \\nATOM     17  N   GLN A   5      -1.567  13.342  16.105  1.00 49.21           N  \\nATOM     18  CA  GLN A   5      -1.632  12.078  16.814  1.00 49.21           C  \\nATOM     19  C   GLN A   5      -0.907  10.950  16.066  1.00 49.21           C  \\nATOM     20  O   GLN A   5      -1.306  10.567  14.965  1.00 84.12           O  \\nATOM     21  CB  GLN A   5      -3.095  11.691  17.017  1.00 87.59           C  \\nATOM     22  CG  GLN A   5      -3.284  10.472  17.883  1.00 87.59           C  \\nATOM     23  CD  GLN A   5      -2.740  10.671  19.282  1.00 87.59           C  \\nATOM     24  OE1 GLN A   5      -2.801   9.770  20.107  1.00 87.59           O  \\nATOM     25  NE2 GLN A   5      -2.209  11.857  19.556  1.00 87.59           N  \\nATOM     26  N   VAL A   6       0.151  10.426  16.683  1.00 61.01           N  \\nATOM     27  CA  VAL A   6       0.937   9.342  16.107  1.00 61.01           C  \\nATOM     28  C   VAL A   6       0.671   8.008  16.803  1.00 61.01           C  \\nATOM     29  O   VAL A   6       0.957   7.837  17.983  1.00 61.01           O  \\nATOM     30  CB  VAL A   6       2.439   9.624  16.196  1.00 50.93           C  \\nATOM     31  CG1 VAL A   6       3.199   8.503  15.531  1.00 50.93           C  \\nATOM     32  CG2 VAL A   6       2.772  10.922  15.516  1.00 50.93           C  \\nATOM     33  N   GLN A   7       0.118   7.066  16.055  1.00 44.88           N  \\nATOM     34  CA  GLN A   7      -0.196   5.735  16.561  1.00 44.88           C  \\nATOM     35  C   GLN A   7       0.719   4.713  15.881  1.00 44.88           C  \\nATOM     36  O   GLN A   7       0.658   4.513  14.674  1.00 44.88           O  \\nATOM     37  CB  GLN A   7      -1.658   5.390  16.264  1.00 83.92           C  \\nATOM     38  CG  GLN A   7      -2.534   5.302  17.485  1.00 83.92           C  \\nATOM     39  CD  GLN A   7      -2.119   4.173  18.412  1.00 83.92           C  \\nATOM     40  OE1 GLN A   7      -0.990   4.138  18.905  1.00 83.92           O  \\nATOM     41  NE2 GLN A   7      -3.031   3.241  18.653  1.00 83.92           N  \\nATOM     42  N   VAL A   8       1.578   4.084  16.673  1.00 45.42           N  \\nATOM     43  CA  VAL A   8       2.499   3.068  16.186  1.00 45.42           C  \\nATOM     44  C   VAL A   8       2.103   1.700  16.756  1.00 45.42           C  \\nATOM     45  O   VAL A   8       2.114   1.508  17.966  1.00 45.42           O  \\nATOM     46  CB  VAL A   8       3.945   3.377  16.613  1.00 41.58           C  \\nATOM     47  CG1 VAL A   8       4.867   2.293  16.086  1.00 41.58           C  \\nATOM     48  CG2 VAL A   8       4.375   4.731  16.083  1.00 41.58           C  \\nATOM     49  N   ASN A   9       1.746   0.755  15.891  1.00 56.60           N  \\nATOM     50  CA  ASN A   9       1.357  -0.569  16.361  1.00 56.60           C  \\nATOM     51  C   ASN A   9       2.224  -1.684  15.748  1.00 56.60           C  \\nATOM     52  O   ASN A   9       2.140  -1.992  14.559  1.00 56.60           O  \\nATOM     53  CB  ASN A   9      -0.119  -0.818  16.065  1.00105.47           C  \\nATOM     54  CG  ASN A   9      -0.823  -1.519  17.208  1.00105.47           C  \\nATOM     55  OD1 ASN A   9      -1.281  -0.879  18.165  1.00105.47           O  \\nATOM     56  ND2 ASN A   9      -0.887  -2.850  17.133  1.00105.47           N  \\nATOM     57  N   ILE A  10       3.073  -2.269  16.587  1.00 40.88           N  \\nATOM     58  CA  ILE A  10       3.976  -3.337  16.187  1.00 40.88           C  \\nATOM     59  C   ILE A  10       3.531  -4.697  16.730  1.00 40.88           C  \\nATOM     60  O   ILE A  10       3.465  -4.900  17.944  1.00 70.73           O  \\nATOM     61  CB  ILE A  10       5.399  -3.068  16.702  1.00 61.80           C  \\nATOM     62  CG1 ILE A  10       5.870  -1.693  16.253  1.00 61.80           C  \\nATOM     63  CG2 ILE A  10       6.343  -4.127  16.180  1.00 61.80           C  \\nATOM     64  CD1 ILE A  10       7.241  -1.341  16.788  1.00 61.80           C  \\nATOM     65  N   ASP A  11       3.225  -5.621  15.824  1.00 68.81           N  \\nATOM     66  CA  ASP A  11       2.802  -6.960  16.220  1.00 68.81           C  \\nATOM     67  C   ASP A  11       3.892  -7.970  15.878  1.00 68.81           C  \\nATOM     68  O   ASP A  11       4.569  -7.850  14.850  1.00 68.81           O  \\nATOM     69  CB  ASP A  11       1.498  -7.342  15.537  1.00114.14           C  \\nATOM     70  CG  ASP A  11       0.590  -8.125  16.449  1.00114.14           C  \\nATOM     71  OD1 ASP A  11       1.008  -9.205  16.916  1.00114.14           O  \\nATOM     72  OD2 ASP A  11      -0.534  -7.651  16.712  1.00114.14           O  \\nATOM     73  N   ASP A  12       4.055  -8.971  16.737  1.00 70.35           N  \\nATOM     74  CA  ASP A  12       5.107  -9.950  16.527  1.00 70.35           C  \\nATOM     75  C   ASP A  12       4.970 -11.157  17.446  1.00 70.35           C  \\nATOM     76  O   ASP A  12       5.649 -11.243  18.468  1.00 70.35           O  \\nATOM     77  CB  ASP A  12       6.462  -9.264  16.752  1.00 53.31           C  \\nATOM     78  CG  ASP A  12       7.632 -10.096  16.286  1.00 53.31           C  \\nATOM     79  OD1 ASP A  12       7.496 -10.774  15.243  1.00 53.31           O  \\nATOM     80  OD2 ASP A  12       8.697 -10.045  16.944  1.00 53.31           O  \\nATOM     81  N   ASN A  13       4.090 -12.081  17.079  1.00 96.17           N  \\nATOM     82  CA  ASN A  13       3.876 -13.307  17.848  1.00 96.17           C  \\nATOM     83  C   ASN A  13       3.792 -13.129  19.361  1.00 96.17           C  \\nATOM     84  O   ASN A  13       4.817 -12.980  20.028  1.00 96.17           O  \\nATOM     85  CB  ASN A  13       4.988 -14.315  17.550  1.00100.93           C  \\nATOM     86  CG  ASN A  13       5.349 -14.363  16.087  1.00100.93           C  \\nATOM     87  OD1 ASN A  13       5.977 -13.444  15.562  1.00100.93           O  \\nATOM     88  ND2 ASN A  13       4.947 -15.434  15.413  1.00100.93           N  \\nATOM     89  N   GLY A  14       2.572 -13.164  19.895  1.00130.92           N  \\nATOM     90  CA  GLY A  14       2.371 -13.029  21.330  1.00130.92           C  \\nATOM     91  C   GLY A  14       2.938 -11.774  21.965  1.00130.92           C  \\nATOM     92  O   GLY A  14       3.099 -11.711  23.185  1.00130.92           O  \\nATOM     93  N   LYS A  15       3.245 -10.775  21.143  1.00 68.27           N  \\nATOM     94  CA  LYS A  15       3.790  -9.520  21.640  1.00 68.27           C  \\nATOM     95  C   LYS A  15       3.313  -8.363  20.774  1.00 68.27           C  \\nATOM     96  O   LYS A  15       3.754  -8.207  19.639  1.00 68.27           O  \\nATOM     97  CB  LYS A  15       5.329  -9.570  21.663  1.00 55.99           C  \\nATOM     98  N   ASN A  16       2.400  -7.560  21.313  1.00 54.52           N  \\nATOM     99  CA  ASN A  16       1.872  -6.404  20.600  1.00 54.52           C  \\nATOM    100  C   ASN A  16       2.268  -5.109  21.338  1.00 54.52           C  \\nATOM    101  O   ASN A  16       2.022  -4.958  22.538  1.00 54.52           O  \\nATOM    102  CB  ASN A  16       0.347  -6.516  20.481  1.00102.09           C  \\nATOM    103  CG  ASN A  16      -0.260  -5.401  19.643  1.00102.09           C  \\nATOM    104  OD1 ASN A  16       0.179  -5.140  18.522  1.00102.09           O  \\nATOM    105  ND2 ASN A  16      -1.284  -4.745  20.181  1.00102.09           N  \\nATOM    106  N   PHE A  17       2.902  -4.187  20.617  1.00 42.25           N  \\nATOM    107  CA  PHE A  17       3.318  -2.920  21.197  1.00 42.25           C  \\nATOM    108  C   PHE A  17       2.488  -1.820  20.560  1.00 42.25           C  \\nATOM    109  O   PHE A  17       2.326  -1.787  19.346  1.00 42.25           O  \\nATOM    110  CB  PHE A  17       4.810  -2.682  20.947  1.00 34.85           C  \\nATOM    111  N   ASP A  18       1.947  -0.934  21.394  1.00 45.89           N  \\nATOM    112  CA  ASP A  18       1.124   0.181  20.939  1.00 45.89           C  \\nATOM    113  C   ASP A  18       1.671   1.527  21.481  1.00 45.89           C  \\nATOM    114  O   ASP A  18       1.496   1.855  22.652  1.00 45.89           O  \\nATOM    115  CB  ASP A  18      -0.310  -0.037  21.406  1.00 89.81           C  \\nATOM    116  CG  ASP A  18      -1.254   0.966  20.813  1.00 89.81           C  \\nATOM    117  OD1 ASP A  18      -1.419   0.953  19.573  1.00 89.81           O  \\nATOM    118  OD2 ASP A  18      -1.816   1.782  21.577  1.00 89.81           O  \\nATOM    119  N   TYR A  19       2.343   2.290  20.623  1.00 43.94           N  \\nATOM    120  CA  TYR A  19       2.936   3.578  20.999  1.00 43.94           C  \\nATOM    121  C   TYR A  19       2.048   4.728  20.530  1.00 43.94           C  \\nATOM    122  O   TYR A  19       1.694   4.814  19.360  1.00 78.71           O  \\nATOM    123  CB  TYR A  19       4.329   3.734  20.373  1.00 87.94           C  \\nATOM    124  CG  TYR A  19       5.347   2.679  20.778  1.00 87.94           C  \\nATOM    125  CD1 TYR A  19       6.083   2.805  21.954  1.00 87.94           C  \\nATOM    126  CD2 TYR A  19       5.578   1.553  19.977  1.00 87.94           C  \\nATOM    127  CE1 TYR A  19       7.029   1.837  22.327  1.00 87.94           C  \\nATOM    128  CE2 TYR A  19       6.521   0.577  20.342  1.00 87.94           C  \\nATOM    129  CZ  TYR A  19       7.243   0.728  21.518  1.00 87.94           C  \\nATOM    130  OH  TYR A  19       8.180  -0.218  21.886  1.00 87.94           O  \\nATOM    131  N   THR A  20       1.691   5.609  21.457  1.00 50.11           N  \\nATOM    132  CA  THR A  20       0.857   6.756  21.147  1.00 50.11           C  \\nATOM    133  C   THR A  20       1.589   8.007  21.589  1.00 50.11           C  \\nATOM    134  O   THR A  20       1.996   8.119  22.746  1.00 50.11           O  \\nATOM    135  CB  THR A  20      -0.458   6.683  21.889  1.00 57.60           C  \\nATOM    136  OG1 THR A  20      -0.954   5.345  21.830  1.00 57.60           O  \\nATOM    137  CG2 THR A  20      -1.468   7.592  21.246  1.00 57.60           C  \\nATOM    138  N   TYR A  21       1.784   8.929  20.649  1.00 57.65           N  \\nATOM    139  CA  TYR A  21       2.468  10.196  20.916  1.00 57.65           C  \\nATOM    140  C   TYR A  21       1.581  11.325  20.413  1.00 57.65           C  \\nATOM    141  O   TYR A  21       0.683  11.120  19.599  1.00 57.65           O  \\nATOM    142  CB  TYR A  21       3.799  10.331  20.154  1.00 63.74           C  \\nATOM    143  CG  TYR A  21       4.810   9.204  20.258  1.00 63.74           C  \\nATOM    144  CD1 TYR A  21       4.560   7.953  19.684  1.00 63.74           C  \\nATOM    145  CD2 TYR A  21       6.060   9.422  20.845  1.00 63.74           C  \\nATOM    146  CE1 TYR A  21       5.525   6.959  19.691  1.00 63.74           C  \\nATOM    147  CE2 TYR A  21       7.031   8.434  20.854  1.00 63.74           C  \\nATOM    148  CZ  TYR A  21       6.759   7.209  20.271  1.00 63.74           C  \\nATOM    149  OH  TYR A  21       7.735   6.240  20.243  1.00 63.74           O  \\nATOM    150  N   THR A  22       1.867  12.525  20.893  1.00 67.33           N  \\nATOM    151  CA  THR A  22       1.136  13.706  20.489  1.00 67.33           C  \\nATOM    152  C   THR A  22       2.205  14.758  20.356  1.00 67.33           C  \\nATOM    153  O   THR A  22       2.833  15.129  21.342  1.00 67.33           O  \\nATOM    154  CB  THR A  22       0.144  14.146  21.558  1.00 65.03           C  \\nATOM    155  OG1 THR A  22      -0.589  13.007  22.026  1.00 65.03           O  \\nATOM    156  CG2 THR A  22      -0.826  15.159  20.984  1.00 65.03           C  \\nATOM    157  N   VAL A  23       2.444  15.212  19.135  1.00 71.70           N  \\nATOM    158  CA  VAL A  23       3.462  16.223  18.921  1.00 71.70           C  \\nATOM    159  C   VAL A  23       2.881  17.499  18.343  1.00 71.70           C  \\nATOM    160  O   VAL A  23       1.714  17.547  17.935  1.00 71.70           O  \\nATOM    161  CB  VAL A  23       4.574  15.719  17.970  1.00 70.78           C  \\nATOM    162  CG1 VAL A  23       5.298  14.540  18.591  1.00 70.78           C  \\nATOM    163  CG2 VAL A  23       3.976  15.342  16.638  1.00 70.78           C  \\nATOM    164  N   THR A  24       3.717  18.531  18.327  1.00 83.96           N  \\nATOM    165  CA  THR A  24       3.369  19.839  17.793  1.00 83.96           C  \\nATOM    166  C   THR A  24       4.686  20.523  17.473  1.00 83.96           C  \\nATOM    167  O   THR A  24       5.175  21.315  18.270  1.00 83.96           O  \\nATOM    168  CB  THR A  24       2.604  20.696  18.831  1.00 95.26           C  \\nATOM    169  OG1 THR A  24       1.323  20.110  19.084  1.00 95.26           O  \\nATOM    170  CG2 THR A  24       2.399  22.114  18.317  1.00 95.26           C  \\nATOM    171  N   THR A  25       5.280  20.203  16.327  1.00136.13           N  \\nATOM    172  CA  THR A  25       6.552  20.818  15.953  1.00136.13           C  \\nATOM    173  C   THR A  25       7.033  20.385  14.576  1.00136.13           C  \\nATOM    174  O   THR A  25       7.608  21.179  13.834  1.00136.13           O  \\nATOM    175  CB  THR A  25       7.703  20.449  16.943  1.00146.88           C  \\nATOM    176  OG1 THR A  25       7.172  20.180  18.244  1.00146.88           O  \\nATOM    177  CG2 THR A  25       8.697  21.598  17.058  1.00146.88           C  \\nATOM    178  N   GLU A  26       6.796  19.119  14.247  1.00187.15           N  \\nATOM    179  CA  GLU A  26       7.267  18.538  12.994  1.00187.15           C  \\nATOM    180  C   GLU A  26       8.782  18.481  13.170  1.00187.15           C  \\nATOM    181  O   GLU A  26       9.508  17.917  12.351  1.00187.15           O  \\nATOM    182  CB  GLU A  26       6.892  19.396  11.781  1.00134.71           C  \\nATOM    183  CG  GLU A  26       7.383  18.807  10.458  1.00 97.04           C  \\nATOM    184  CD  GLU A  26       6.713  19.422   9.243  1.00 97.04           C  \\nATOM    185  OE1 GLU A  26       6.740  20.662   9.115  1.00 97.04           O  \\nATOM    186  OE2 GLU A  26       6.165  18.662   8.416  1.00 97.04           O  \\nATOM    187  N   SER A  27       9.234  19.085  14.266  1.00114.54           N  \\nATOM    188  CA  SER A  27      10.637  19.110  14.647  1.00114.54           C  \\nATOM    189  C   SER A  27      10.764  18.025  15.711  1.00114.54           C  \\nATOM    190  O   SER A  27      11.634  17.158  15.624  1.00114.54           O  \\nATOM    191  CB  SER A  27      11.008  20.463  15.229  1.00116.53           C  \\nATOM    192  N   GLU A  28       9.885  18.071  16.713  1.00113.18           N  \\nATOM    193  CA  GLU A  28       9.900  17.055  17.762  1.00113.18           C  \\nATOM    194  C   GLU A  28       9.258  15.810  17.168  1.00113.18           C  \\nATOM    195  O   GLU A  28       9.094  14.794  17.842  1.00200.03           O  \\nATOM    196  CB  GLU A  28       9.116  17.505  19.003  1.00112.19           C  \\nATOM    197  CG  GLU A  28       7.596  17.557  18.841  1.00112.19           C  \\nATOM    198  CD  GLU A  28       6.879  17.960  20.129  1.00112.19           C  \\nATOM    199  OE1 GLU A  28       5.778  18.547  20.047  1.00112.19           O  \\nATOM    200  OE2 GLU A  28       7.407  17.679  21.226  1.00112.19           O  \\nATOM    201  N   LEU A  29       8.883  15.919  15.895  1.00 97.20           N  \\nATOM    202  CA  LEU A  29       8.286  14.820  15.147  1.00 97.20           C  \\nATOM    203  C   LEU A  29       9.475  14.012  14.649  1.00 97.20           C  \\nATOM    204  O   LEU A  29       9.456  12.784  14.643  1.00 97.20           O  \\nATOM    205  CB  LEU A  29       7.493  15.350  13.947  1.00 66.59           C  \\nATOM    206  CG  LEU A  29       6.452  14.423  13.296  1.00 66.59           C  \\nATOM    207  CD1 LEU A  29       6.087  14.961  11.915  1.00 66.59           C  \\nATOM    208  CD2 LEU A  29       6.997  13.012  13.162  1.00 66.59           C  \\nATOM    209  N   GLN A  30      10.512  14.723  14.225  1.00113.13           N  \\nATOM    210  CA  GLN A  30      11.724  14.081  13.747  1.00113.13           C  \\nATOM    211  C   GLN A  30      12.310  13.296  14.916  1.00113.13           C  \\nATOM    212  O   GLN A  30      13.031  12.315  14.725  1.00113.13           O  \\nATOM    213  CB  GLN A  30      12.718  15.135  13.256  1.00 67.53           C  \\nATOM    214  N   LYS A  31      11.984  13.738  16.128  1.00 75.23           N  \\nATOM    215  CA  LYS A  31      12.465  13.094  17.348  1.00 75.23           C  \\nATOM    216  C   LYS A  31      11.731  11.776  17.556  1.00 75.23           C  \\nATOM    217  O   LYS A  31      12.331  10.751  17.887  1.00 75.23           O  \\nATOM    218  CB  LYS A  31      12.247  14.016  18.550  1.00 80.64           C  \\nATOM    219  N   VAL A  32      10.420  11.824  17.360  1.00 56.38           N  \\nATOM    220  CA  VAL A  32       9.562  10.658  17.497  1.00 56.38           C  \\nATOM    221  C   VAL A  32       9.898   9.703  16.364  1.00 56.38           C  \\nATOM    222  O   VAL A  32      10.051   8.496  16.569  1.00 56.38           O  \\nATOM    223  CB  VAL A  32       8.090  11.065  17.376  1.00 46.58           C  \\nATOM    224  CG1 VAL A  32       7.202   9.830  17.366  1.00 46.58           C  \\nATOM    225  CG2 VAL A  32       7.725  12.000  18.532  1.00 46.58           C  \\nATOM    226  N   LEU A  33      10.016  10.262  15.162  1.00 55.92           N  \\nATOM    227  CA  LEU A  33      10.340   9.470  13.992  1.00 55.92           C  \\nATOM    228  C   LEU A  33      11.608   8.685  14.235  1.00 55.92           C  \\nATOM    229  O   LEU A  33      11.579   7.454  14.286  1.00 55.92           O  \\nATOM    230  CB  LEU A  33      10.523  10.363  12.773  1.00 52.98           C  \\nATOM    231  CG  LEU A  33       9.242  10.835  12.080  1.00 52.98           C  \\nATOM    232  CD1 LEU A  33       9.641  11.471  10.750  1.00 52.98           C  \\nATOM    233  CD2 LEU A  33       8.273   9.666  11.846  1.00 52.98           C  \\nATOM    234  N   ASN A  34      12.719   9.404  14.389  1.00 69.48           N  \\nATOM    235  CA  ASN A  34      14.011   8.779  14.631  1.00 69.48           C  \\nATOM    236  C   ASN A  34      13.870   7.727  15.722  1.00 69.48           C  \\nATOM    237  O   ASN A  34      14.350   6.605  15.584  1.00 69.48           O  \\nATOM    238  CB  ASN A  34      15.039   9.844  15.038  1.00 46.71           C  \\nATOM    239  N   GLU A  35      13.180   8.098  16.795  1.00 62.44           N  \\nATOM    240  CA  GLU A  35      12.972   7.212  17.938  1.00 62.44           C  \\nATOM    241  C   GLU A  35      12.233   5.917  17.574  1.00 62.44           C  \\nATOM    242  O   GLU A  35      12.580   4.830  18.057  1.00 62.44           O  \\nATOM    243  CB  GLU A  35      12.207   7.956  19.039  1.00129.81           C  \\nATOM    244  CG  GLU A  35      11.750   7.084  20.206  1.00114.46           C  \\nATOM    245  CD  GLU A  35      11.028   7.870  21.289  1.00114.46           C  \\nATOM    246  OE1 GLU A  35      10.356   7.237  22.131  1.00114.46           O  \\nATOM    247  OE2 GLU A  35      11.141   9.116  21.298  1.00114.46           O  \\nATOM    248  N   LEU A  36      11.213   6.021  16.730  1.00 59.14           N  \\nATOM    249  CA  LEU A  36      10.469   4.828  16.342  1.00 59.14           C  \\nATOM    250  C   LEU A  36      11.321   3.923  15.466  1.00 59.14           C  \\nATOM    251  O   LEU A  36      11.229   2.702  15.544  1.00 59.14           O  \\nATOM    252  CB  LEU A  36       9.217   5.227  15.613  1.00 61.28           C  \\nHETATM  253  N   MSE A  37      12.136   4.544  14.621  1.00 45.85           N  \\nHETATM  254  CA  MSE A  37      13.021   3.824  13.725  1.00 45.85           C  \\nHETATM  255  C   MSE A  37      13.869   2.835  14.511  1.00 45.85           C  \\nHETATM  256  O   MSE A  37      14.082   1.700  14.085  1.00 45.85           O  \\nHETATM  257  CB  MSE A  37      13.922   4.804  12.968  1.00 57.27           C  \\nHETATM  258  CG  MSE A  37      13.203   5.576  11.873  1.00 57.27           C  \\nHETATM  259 SE   MSE A  37      14.347   6.471  10.802  1.00 57.27          SE  \\nHETATM  260  CE  MSE A  37      14.638   5.266   9.472  1.00 57.27           C  \\nATOM    261  N   ASP A  38      14.338   3.255  15.679  1.00 56.30           N  \\nATOM    262  CA  ASP A  38      15.156   2.372  16.497  1.00 56.30           C  \\nATOM    263  C   ASP A  38      14.343   1.250  17.135  1.00 56.30           C  \\nATOM    264  O   ASP A  38      14.695   0.070  17.008  1.00 56.30           O  \\nATOM    265  CB  ASP A  38      15.901   3.179  17.556  1.00 94.79           C  \\nATOM    266  CG  ASP A  38      16.859   4.191  16.942  1.00 94.79           C  \\nATOM    267  OD1 ASP A  38      17.680   3.809  16.079  1.00 94.79           O  \\nATOM    268  OD2 ASP A  38      16.791   5.377  17.317  1.00 94.79           O  \\nATOM    269  N   TYR A  39      13.249   1.597  17.806  1.00 57.17           N  \\nATOM    270  CA  TYR A  39      12.420   0.561  18.424  1.00 57.17           C  \\nATOM    271  C   TYR A  39      12.126  -0.564  17.425  1.00 57.17           C  \\nATOM    272  O   TYR A  39      12.195  -1.745  17.771  1.00104.80           O  \\nATOM    273  CB  TYR A  39      11.086   1.135  18.915  1.00 92.26           C  \\nATOM    274  CG  TYR A  39      11.163   2.049  20.123  1.00 92.26           C  \\nATOM    275  CD1 TYR A  39      11.708   1.607  21.333  1.00 92.26           C  \\nATOM    276  CD2 TYR A  39      10.645   3.346  20.069  1.00 92.26           C  \\nATOM    277  CE1 TYR A  39      11.732   2.439  22.460  1.00 92.26           C  \\nATOM    278  CE2 TYR A  39      10.661   4.179  21.186  1.00 92.26           C  \\nATOM    279  CZ  TYR A  39      11.204   3.724  22.373  1.00 92.26           C  \\nATOM    280  OH  TYR A  39      11.218   4.565  23.462  1.00 92.26           O  \\nATOM    281  N   ILE A  40      11.805  -0.192  16.186  1.00 75.79           N  \\nATOM    282  CA  ILE A  40      11.467  -1.188  15.177  1.00 75.79           C  \\nATOM    283  C   ILE A  40      12.677  -1.933  14.605  1.00 75.79           C  \\nATOM    284  O   ILE A  40      12.570  -3.121  14.295  1.00 75.79           O  \\nATOM    285  CB  ILE A  40      10.603  -0.569  14.031  1.00 76.70           C  \\nATOM    286  CG1 ILE A  40      11.482   0.006  12.931  1.00 76.70           C  \\nATOM    287  CG2 ILE A  40       9.709   0.512  14.582  1.00 76.70           C  \\nATOM    288  CD1 ILE A  40      11.700  -0.961  11.782  1.00 76.70           C  \\nATOM    289  N   LYS A  41      13.817  -1.255  14.454  1.00 65.22           N  \\nATOM    290  CA  LYS A  41      15.015  -1.937  13.960  1.00 65.22           C  \\nATOM    291  C   LYS A  41      15.238  -3.089  14.927  1.00 65.22           C  \\nATOM    292  O   LYS A  41      15.336  -4.253  14.534  1.00 65.22           O  \\nATOM    293  CB  LYS A  41      16.245  -1.024  13.999  1.00 66.42           C  \\nATOM    294  CG  LYS A  41      16.441  -0.159  12.762  1.00 66.42           C  \\nATOM    295  CD  LYS A  41      17.635   0.777  12.923  1.00 66.42           C  \\nATOM    296  CE  LYS A  41      17.801   1.692  11.713  1.00 66.42           C  \\nATOM    297  NZ  LYS A  41      18.856   2.725  11.913  1.00 66.42           N  \\nATOM    298  N   LYS A  42      15.287  -2.736  16.207  1.00 55.06           N  \\nATOM    299  CA  LYS A  42      15.498  -3.693  17.278  1.00 55.06           C  \\nATOM    300  C   LYS A  42      14.502  -4.852  17.267  1.00 55.06           C  \\nATOM    301  O   LYS A  42      14.892  -6.013  17.198  1.00 55.06           O  \\nATOM    302  CB  LYS A  42      15.437  -2.966  18.625  1.00200.03           C  \\nATOM    303  CG  LYS A  42      15.692  -3.852  19.830  1.00186.63           C  \\nATOM    304  CD  LYS A  42      15.652  -3.051  21.117  1.00186.63           C  \\nATOM    305  CE  LYS A  42      15.900  -3.945  22.320  1.00139.05           C  \\nATOM    306  NZ  LYS A  42      17.216  -4.640  22.230  1.00139.05           N  \\nATOM    307  N   GLN A  43      13.214  -4.548  17.331  1.00 78.34           N  \\nATOM    308  CA  GLN A  43      12.211  -5.606  17.359  1.00 78.34           C  \\nATOM    309  C   GLN A  43      12.149  -6.428  16.073  1.00 78.34           C  \\nATOM    310  O   GLN A  43      11.838  -7.621  16.111  1.00 78.34           O  \\nATOM    311  CB  GLN A  43      10.832  -5.013  17.677  1.00126.98           C  \\nATOM    312  CG  GLN A  43      10.169  -5.609  18.919  1.00126.98           C  \\nATOM    313  CD  GLN A  43       9.832  -7.084  18.758  1.00126.98           C  \\nATOM    314  OE1 GLN A  43       8.990  -7.452  17.940  1.00126.98           O  \\nATOM    315  NE2 GLN A  43      10.495  -7.935  19.537  1.00126.98           N  \\nATOM    316  N   GLY A  44      12.452  -5.789  14.942  1.00 55.22           N  \\nATOM    317  CA  GLY A  44      12.412  -6.472  13.661  1.00 55.22           C  \\nATOM    318  C   GLY A  44      11.154  -7.305  13.501  1.00 55.22           C  \\nATOM    319  O   GLY A  44      11.230  -8.524  13.317  1.00 55.22           O  \\nATOM    320  N   ALA A  45       9.991  -6.650  13.567  1.00 48.70           N  \\nATOM    321  CA  ALA A  45       8.706  -7.334  13.446  1.00 48.70           C  \\nATOM    322  C   ALA A  45       8.267  -7.491  12.002  1.00 48.70           C  \\nATOM    323  O   ALA A  45       8.625  -6.691  11.135  1.00 48.70           O  \\nATOM    324  CB  ALA A  45       7.651  -6.587  14.227  1.00 42.35           C  \\nATOM    325  N   LYS A  46       7.482  -8.527  11.748  1.00 60.45           N  \\nATOM    326  CA  LYS A  46       7.002  -8.787  10.405  1.00 60.45           C  \\nATOM    327  C   LYS A  46       6.082  -7.662   9.961  1.00 60.45           C  \\nATOM    328  O   LYS A  46       6.256  -7.102   8.881  1.00 60.45           O  \\nATOM    329  CB  LYS A  46       6.266 -10.122  10.362  1.00100.71           C  \\nATOM    330  N   ARG A  47       5.116  -7.324  10.808  1.00 55.30           N  \\nATOM    331  CA  ARG A  47       4.160  -6.279  10.484  1.00 55.30           C  \\nATOM    332  C   ARG A  47       4.244  -5.060  11.406  1.00 55.30           C  \\nATOM    333  O   ARG A  47       4.294  -5.187  12.625  1.00 55.30           O  \\nATOM    334  CB  ARG A  47       2.747  -6.860  10.497  1.00 49.32           C  \\nATOM    335  N   VAL A  48       4.257  -3.875  10.796  1.00 43.03           N  \\nATOM    336  CA  VAL A  48       4.320  -2.600  11.511  1.00 43.03           C  \\nATOM    337  C   VAL A  48       3.326  -1.588  10.915  1.00 43.03           C  \\nATOM    338  O   VAL A  48       3.297  -1.359   9.715  1.00 43.03           O  \\nATOM    339  CB  VAL A  48       5.742  -2.008  11.457  1.00 31.14           C  \\nATOM    340  CG1 VAL A  48       5.702  -0.529  11.831  1.00 31.14           C  \\nATOM    341  CG2 VAL A  48       6.653  -2.775  12.414  1.00 31.14           C  \\nATOM    342  N   ARG A  49       2.520  -0.975  11.776  1.00 36.65           N  \\nATOM    343  CA  ARG A  49       1.516  -0.009  11.362  1.00 36.65           C  \\nATOM    344  C   ARG A  49       1.740   1.333  12.036  1.00 36.65           C  \\nATOM    345  O   ARG A  49       2.013   1.414  13.231  1.00 36.65           O  \\nATOM    346  CB  ARG A  49       0.124  -0.512  11.726  1.00 83.41           C  \\nATOM    347  CG  ARG A  49      -1.002   0.196  10.999  1.00 73.04           C  \\nATOM    348  CD  ARG A  49      -2.287   0.153  11.807  1.00 73.04           C  \\nATOM    349  NE  ARG A  49      -2.459  -1.126  12.495  1.00 73.04           N  \\nATOM    350  CZ  ARG A  49      -3.229  -1.302  13.569  1.00 73.04           C  \\nATOM    351  NH1 ARG A  49      -3.910  -0.278  14.083  1.00 73.04           N  \\nATOM    352  NH2 ARG A  49      -3.304  -2.499  14.147  1.00 73.04           N  \\nATOM    353  N   ILE A  50       1.630   2.395  11.252  1.00 31.16           N  \\nATOM    354  CA  ILE A  50       1.776   3.725  11.785  1.00 31.16           C  \\nATOM    355  C   ILE A  50       0.705   4.599  11.157  1.00 31.16           C  \\nATOM    356  O   ILE A  50       0.578   4.660   9.933  1.00 31.16           O  \\nATOM    357  CB  ILE A  50       3.147   4.316  11.465  1.00 26.84           C  \\nATOM    358  CG1 ILE A  50       4.236   3.406  12.024  1.00 26.84           C  \\nATOM    359  CG2 ILE A  50       3.254   5.722  12.079  1.00 26.84           C  \\nATOM    360  CD1 ILE A  50       5.611   4.014  11.978  1.00 26.84           C  \\nATOM    361  N   SER A  51      -0.093   5.239  12.002  1.00 37.07           N  \\nATOM    362  CA  SER A  51      -1.128   6.139  11.519  1.00 37.07           C  \\nATOM    363  C   SER A  51      -0.951   7.489  12.205  1.00 37.07           C  \\nATOM    364  O   SER A  51      -0.674   7.573  13.402  1.00 37.07           O  \\nATOM    365  CB  SER A  51      -2.522   5.559  11.757  1.00 38.78           C  \\nATOM    366  OG  SER A  51      -2.672   5.197  13.104  1.00 38.78           O  \\nATOM    367  N   ILE A  52      -1.067   8.538  11.408  1.00 47.52           N  \\nATOM    368  CA  ILE A  52      -0.899   9.893  11.887  1.00 47.52           C  \\nATOM    369  C   ILE A  52      -2.116  10.712  11.531  1.00 47.52           C  \\nATOM    370  O   ILE A  52      -2.486  10.825  10.359  1.00 47.52           O  \\nATOM    371  CB  ILE A  52       0.341  10.536  11.257  1.00 41.87           C  \\nATOM    372  CG1 ILE A  52       1.592   9.886  11.839  1.00 41.87           C  \\nATOM    373  CG2 ILE A  52       0.336  12.039  11.479  1.00 41.87           C  \\nATOM    374  CD1 ILE A  52       2.843  10.297  11.141  1.00 41.87           C  \\nATOM    375  N   THR A  53      -2.753  11.261  12.556  1.00 55.53           N  \\nATOM    376  CA  THR A  53      -3.922  12.081  12.351  1.00 55.53           C  \\nATOM    377  C   THR A  53      -3.415  13.493  12.189  1.00 55.53           C  \\nATOM    378  O   THR A  53      -2.919  14.098  13.139  1.00 55.53           O  \\nATOM    379  CB  THR A  53      -4.866  12.023  13.538  1.00 54.19           C  \\nATOM    380  OG1 THR A  53      -5.409  10.702  13.658  1.00 54.19           O  \\nATOM    381  CG2 THR A  53      -5.997  13.003  13.334  1.00 54.19           C  \\nATOM    382  N   ALA A  54      -3.523  13.994  10.965  1.00 53.46           N  \\nATOM    383  CA  ALA A  54      -3.079  15.335  10.624  1.00 53.46           C  \\nATOM    384  C   ALA A  54      -4.249  16.316  10.700  1.00 53.46           C  \\nATOM    385  O   ALA A  54      -5.389  15.916  10.939  1.00 53.46           O  \\nATOM    386  CB  ALA A  54      -2.488  15.331   9.219  1.00 54.30           C  \\nATOM    387  N   ARG A  55      -3.960  17.596  10.490  1.00 74.08           N  \\nATOM    388  CA  ARG A  55      -4.983  18.636  10.528  1.00 74.08           C  \\nATOM    389  C   ARG A  55      -5.731  18.711   9.197  1.00 74.08           C  \\nATOM    390  O   ARG A  55      -6.929  18.979   9.170  1.00 74.08           O  \\nATOM    391  CB  ARG A  55      -4.344  19.985  10.857  1.00 53.00           C  \\nATOM    392  N   THR A  56      -5.017  18.477   8.098  1.00 61.37           N  \\nATOM    393  CA  THR A  56      -5.615  18.505   6.762  1.00 61.37           C  \\nATOM    394  C   THR A  56      -5.102  17.346   5.926  1.00 61.37           C  \\nATOM    395  O   THR A  56      -3.968  16.892   6.107  1.00 61.37           O  \\nATOM    396  CB  THR A  56      -5.281  19.807   6.003  1.00 61.95           C  \\nATOM    397  OG1 THR A  56      -3.861  19.944   5.857  1.00 61.95           O  \\nATOM    398  CG2 THR A  56      -5.818  20.995   6.750  1.00 61.95           C  \\nATOM    399  N   LYS A  57      -5.929  16.874   4.998  1.00 79.80           N  \\nATOM    400  CA  LYS A  57      -5.522  15.766   4.148  1.00 79.80           C  \\nATOM    401  C   LYS A  57      -4.219  16.045   3.395  1.00 79.80           C  \\nATOM    402  O   LYS A  57      -3.422  15.135   3.192  1.00 79.80           O  \\nATOM    403  CB  LYS A  57      -6.635  15.394   3.164  1.00 56.21           C  \\nATOM    404  CG  LYS A  57      -6.161  14.470   2.051  1.00 52.05           C  \\nATOM    405  CD  LYS A  57      -7.236  13.507   1.571  1.00 52.05           C  \\nATOM    406  CE  LYS A  57      -6.718  12.676   0.392  1.00 52.05           C  \\nATOM    407  NZ  LYS A  57      -7.658  11.597  -0.067  1.00 52.05           N  \\nATOM    408  N   LYS A  58      -3.990  17.291   2.986  1.00 61.25           N  \\nATOM    409  CA  LYS A  58      -2.758  17.624   2.269  1.00 61.25           C  \\nATOM    410  C   LYS A  58      -1.566  17.464   3.204  1.00 61.25           C  \\nATOM    411  O   LYS A  58      -0.433  17.274   2.764  1.00 61.25           O  \\nATOM    412  CB  LYS A  58      -2.802  19.055   1.719  1.00120.41           C  \\nATOM    413  CG  LYS A  58      -3.869  19.283   0.655  1.00 90.99           C  \\nATOM    414  CD  LYS A  58      -3.473  20.388  -0.316  1.00 90.99           C  \\nATOM    415  CE  LYS A  58      -2.310  19.959  -1.211  1.00 90.99           C  \\nATOM    416  NZ  LYS A  58      -2.662  18.828  -2.130  1.00 90.99           N  \\nATOM    417  N   GLU A  59      -1.831  17.546   4.500  1.00 55.60           N  \\nATOM    418  CA  GLU A  59      -0.786  17.378   5.496  1.00 55.60           C  \\nATOM    419  C   GLU A  59      -0.515  15.871   5.628  1.00 55.60           C  \\nATOM    420  O   GLU A  59       0.633  15.424   5.622  1.00 55.60           O  \\nATOM    421  CB  GLU A  59      -1.254  17.956   6.835  1.00 68.19           C  \\nATOM    422  CG  GLU A  59      -0.299  17.742   7.985  1.00 68.19           C  \\nATOM    423  CD  GLU A  59      -0.721  18.494   9.232  1.00 68.19           C  \\nATOM    424  OE1 GLU A  59      -1.867  18.296   9.694  1.00 68.19           O  \\nATOM    425  OE2 GLU A  59       0.098  19.287   9.750  1.00 68.19           O  \\nATOM    426  N   ALA A  60      -1.595  15.104   5.738  1.00 49.39           N  \\nATOM    427  CA  ALA A  60      -1.528  13.653   5.853  1.00 49.39           C  \\nATOM    428  C   ALA A  60      -0.674  13.071   4.736  1.00 49.39           C  \\nATOM    429  O   ALA A  60      -0.023  12.038   4.910  1.00 49.39           O  \\nATOM    430  CB  ALA A  60      -2.920  13.069   5.769  1.00 34.98           C  \\nATOM    431  N   GLU A  61      -0.698  13.742   3.588  1.00 62.75           N  \\nATOM    432  CA  GLU A  61       0.058  13.318   2.423  1.00 62.75           C  \\nATOM    433  C   GLU A  61       1.541  13.547   2.665  1.00 62.75           C  \\nATOM    434  O   GLU A  61       2.357  12.657   2.427  1.00 62.75           O  \\nATOM    435  CB  GLU A  61      -0.412  14.092   1.189  1.00 78.01           C  \\nATOM    436  CG  GLU A  61      -1.853  13.778   0.793  1.00 51.13           C  \\nATOM    437  CD  GLU A  61      -2.387  14.677  -0.316  1.00 51.13           C  \\nATOM    438  OE1 GLU A  61      -1.632  15.550  -0.805  1.00 51.13           O  \\nATOM    439  OE2 GLU A  61      -3.567  14.506  -0.695  1.00 51.13           O  \\nATOM    440  N   LYS A  62       1.894  14.733   3.151  1.00 51.31           N  \\nATOM    441  CA  LYS A  62       3.292  15.029   3.420  1.00 51.31           C  \\nATOM    442  C   LYS A  62       3.824  13.998   4.406  1.00 51.31           C  \\nATOM    443  O   LYS A  62       4.972  13.578   4.315  1.00 51.31           O  \\nATOM    444  CB  LYS A  62       3.446  16.445   3.996  1.00 31.92           C  \\nATOM    445  N   PHE A  63       2.977  13.589   5.346  1.00 47.90           N  \\nATOM    446  CA  PHE A  63       3.366  12.610   6.358  1.00 47.90           C  \\nATOM    447  C   PHE A  63       3.498  11.225   5.770  1.00 47.90           C  \\nATOM    448  O   PHE A  63       4.373  10.463   6.164  1.00 47.90           O  \\nATOM    449  CB  PHE A  63       2.343  12.577   7.500  1.00 46.25           C  \\nATOM    450  CG  PHE A  63       2.434  13.748   8.420  1.00 46.25           C  \\nATOM    451  CD1 PHE A  63       1.319  14.174   9.130  1.00 46.25           C  \\nATOM    452  CD2 PHE A  63       3.638  14.420   8.591  1.00 46.25           C  \\nATOM    453  CE1 PHE A  63       1.406  15.259  10.007  1.00 46.25           C  \\nATOM    454  CE2 PHE A  63       3.740  15.496   9.455  1.00 46.25           C  \\nATOM    455  CZ  PHE A  63       2.628  15.922  10.166  1.00 46.25           C  \\nATOM    456  N   ALA A  64       2.610  10.899   4.839  1.00 43.04           N  \\nATOM    457  CA  ALA A  64       2.651   9.602   4.194  1.00 43.04           C  \\nATOM    458  C   ALA A  64       3.960   9.509   3.414  1.00 43.04           C  \\nATOM    459  O   ALA A  64       4.578   8.457   3.367  1.00 43.04           O  \\nATOM    460  CB  ALA A  64       1.441   9.419   3.260  1.00 29.61           C  \\nATOM    461  N   ALA A  65       4.388  10.612   2.815  1.00 46.89           N  \\nATOM    462  CA  ALA A  65       5.642  10.619   2.076  1.00 46.89           C  \\nATOM    463  C   ALA A  65       6.791  10.351   3.055  1.00 46.89           C  \\nATOM    464  O   ALA A  65       7.737   9.629   2.737  1.00 46.89           O  \\nATOM    465  CB  ALA A  65       5.839  11.968   1.367  1.00 31.68           C  \\nATOM    466  N   ILE A  66       6.703  10.929   4.248  1.00 45.51           N  \\nATOM    467  CA  ILE A  66       7.722  10.737   5.275  1.00 45.51           C  \\nATOM    468  C   ILE A  66       7.762   9.285   5.741  1.00 45.51           C  \\nATOM    469  O   ILE A  66       8.816   8.652   5.778  1.00 45.51           O  \\nATOM    470  CB  ILE A  66       7.444  11.631   6.508  1.00 53.42           C  \\nATOM    471  CG1 ILE A  66       7.884  13.069   6.215  1.00 53.42           C  \\nATOM    472  CG2 ILE A  66       8.153  11.075   7.746  1.00 53.42           C  \\nATOM    473  CD1 ILE A  66       7.671  14.034   7.377  1.00 53.42           C  \\nATOM    474  N   LEU A  67       6.599   8.766   6.103  1.00 37.91           N  \\nATOM    475  CA  LEU A  67       6.501   7.412   6.591  1.00 37.91           C  \\nATOM    476  C   LEU A  67       6.912   6.387   5.551  1.00 37.91           C  \\nATOM    477  O   LEU A  67       7.378   5.299   5.887  1.00 37.91           O  \\nATOM    478  CB  LEU A  67       5.079   7.144   7.071  1.00 24.44           C  \\nATOM    479  CG  LEU A  67       4.659   8.016   8.260  1.00 24.44           C  \\nATOM    480  CD1 LEU A  67       3.226   7.626   8.711  1.00 24.44           C  \\nATOM    481  CD2 LEU A  67       5.658   7.854   9.394  1.00 24.44           C  \\nATOM    482  N   ILE A  68       6.748   6.723   4.282  1.00 32.01           N  \\nATOM    483  CA  ILE A  68       7.119   5.776   3.255  1.00 32.01           C  \\nATOM    484  C   ILE A  68       8.615   5.568   3.244  1.00 32.01           C  \\nATOM    485  O   ILE A  68       9.074   4.433   3.201  1.00 32.01           O  \\nATOM    486  CB  ILE A  68       6.622   6.217   1.869  1.00 44.65           C  \\nATOM    487  CG1 ILE A  68       5.223   5.656   1.656  1.00 44.65           C  \\nATOM    488  CG2 ILE A  68       7.533   5.705   0.786  1.00 44.65           C  \\nATOM    489  CD1 ILE A  68       4.580   6.096   0.373  1.00 44.65           C  \\nATOM    490  N   LYS A  69       9.366   6.661   3.310  1.00 46.37           N  \\nATOM    491  CA  LYS A  69      10.819   6.592   3.311  1.00 46.37           C  \\nATOM    492  C   LYS A  69      11.296   5.881   4.574  1.00 46.37           C  \\nATOM    493  O   LYS A  69      12.268   5.125   4.533  1.00 46.37           O  \\nATOM    494  CB  LYS A  69      11.401   8.001   3.201  1.00 60.80           C  \\nATOM    495  CG  LYS A  69      10.814   8.757   2.008  1.00 54.72           C  \\nATOM    496  CD  LYS A  69      11.461  10.116   1.760  1.00 54.72           C  \\nATOM    497  CE  LYS A  69      10.804  10.803   0.552  1.00 54.72           C  \\nATOM    498  NZ  LYS A  69      11.330  12.176   0.230  1.00 54.72           N  \\nATOM    499  N   VAL A  70      10.601   6.102   5.689  1.00 46.67           N  \\nATOM    500  CA  VAL A  70      10.959   5.442   6.942  1.00 46.67           C  \\nATOM    501  C   VAL A  70      10.872   3.920   6.791  1.00 46.67           C  \\nATOM    502  O   VAL A  70      11.842   3.216   7.044  1.00 46.67           O  \\nATOM    503  CB  VAL A  70      10.034   5.890   8.104  1.00 45.21           C  \\nATOM    504  CG1 VAL A  70      10.161   4.929   9.292  1.00 45.21           C  \\nATOM    505  CG2 VAL A  70      10.405   7.301   8.539  1.00 45.21           C  \\nATOM    506  N   PHE A  71       9.705   3.425   6.385  1.00 47.43           N  \\nATOM    507  CA  PHE A  71       9.472   1.995   6.184  1.00 47.43           C  \\nATOM    508  C   PHE A  71      10.400   1.397   5.132  1.00 47.43           C  \\nATOM    509  O   PHE A  71      10.869   0.265   5.276  1.00 47.43           O  \\nATOM    510  CB  PHE A  71       8.043   1.749   5.717  1.00 37.02           C  \\nATOM    511  CG  PHE A  71       7.046   1.645   6.819  1.00 37.02           C  \\nATOM    512  CD1 PHE A  71       6.882   0.462   7.510  1.00 37.02           C  \\nATOM    513  CD2 PHE A  71       6.236   2.722   7.146  1.00 37.02           C  \\nATOM    514  CE1 PHE A  71       5.912   0.350   8.513  1.00 37.02           C  \\nATOM    515  CE2 PHE A  71       5.276   2.619   8.138  1.00 37.02           C  \\nATOM    516  CZ  PHE A  71       5.110   1.440   8.820  1.00 37.02           C  \\nATOM    517  N   ALA A  72      10.631   2.143   4.058  1.00 53.65           N  \\nATOM    518  CA  ALA A  72      11.491   1.668   2.986  1.00 53.65           C  \\nATOM    519  C   ALA A  72      12.911   1.486   3.508  1.00 53.65           C  \\nATOM    520  O   ALA A  72      13.481   0.397   3.410  1.00 53.65           O  \\nATOM    521  CB  ALA A  72      11.476   2.654   1.832  1.00 41.89           C  \\nATOM    522  N   GLU A  73      13.474   2.552   4.070  1.00 50.89           N  \\nATOM    523  CA  GLU A  73      14.818   2.483   4.606  1.00 50.89           C  \\nATOM    524  C   GLU A  73      14.914   1.282   5.565  1.00 50.89           C  \\nATOM    525  O   GLU A  73      15.805   0.457   5.434  1.00 50.89           O  \\nATOM    526  CB  GLU A  73      15.186   3.805   5.314  1.00 20.48           C  \\nATOM    527  N   LEU A  74      13.978   1.145   6.496  1.00 43.89           N  \\nATOM    528  CA  LEU A  74      14.025   0.022   7.431  1.00 43.89           C  \\nATOM    529  C   LEU A  74      13.690  -1.342   6.817  1.00 43.89           C  \\nATOM    530  O   LEU A  74      13.367  -2.282   7.535  1.00 43.89           O  \\nATOM    531  CB  LEU A  74      13.100   0.287   8.621  1.00 51.78           C  \\nATOM    532  CG  LEU A  74      13.297   1.686   9.215  1.00 51.78           C  \\nATOM    533  CD1 LEU A  74      12.449   1.868  10.462  1.00 51.78           C  \\nATOM    534  CD2 LEU A  74      14.769   1.889   9.525  1.00 51.78           C  \\nATOM    535  N   GLY A  75      13.731  -1.453   5.492  1.00 38.15           N  \\nATOM    536  CA  GLY A  75      13.465  -2.739   4.875  1.00 38.15           C  \\nATOM    537  C   GLY A  75      12.039  -3.207   4.697  1.00 38.15           C  \\nATOM    538  O   GLY A  75      11.810  -4.287   4.152  1.00 38.15           O  \\nATOM    539  N   TYR A  76      11.063  -2.447   5.168  1.00 50.24           N  \\nATOM    540  CA  TYR A  76       9.685  -2.857   4.949  1.00 50.24           C  \\nATOM    541  C   TYR A  76       9.433  -2.436   3.506  1.00 50.24           C  \\nATOM    542  O   TYR A  76       9.581  -1.257   3.162  1.00 50.24           O  \\nATOM    543  CB  TYR A  76       8.781  -2.137   5.932  1.00 36.14           C  \\nATOM    544  CG  TYR A  76       9.004  -2.635   7.338  1.00 36.14           C  \\nATOM    545  CD1 TYR A  76       8.162  -3.598   7.898  1.00 36.14           C  \\nATOM    546  CD2 TYR A  76      10.128  -2.235   8.070  1.00 36.14           C  \\nATOM    547  CE1 TYR A  76       8.438  -4.166   9.150  1.00 36.14           C  \\nATOM    548  CE2 TYR A  76      10.413  -2.802   9.322  1.00 36.14           C  \\nATOM    549  CZ  TYR A  76       9.567  -3.770   9.848  1.00 36.14           C  \\nATOM    550  OH  TYR A  76       9.881  -4.392  11.036  1.00 36.14           O  \\nATOM    551  N   ASN A  77       9.073  -3.396   2.657  1.00 50.62           N  \\nATOM    552  CA  ASN A  77       8.890  -3.090   1.246  1.00 50.62           C  \\nATOM    553  C   ASN A  77       7.535  -3.320   0.577  1.00 50.62           C  \\nATOM    554  O   ASN A  77       7.395  -3.014  -0.605  1.00 50.62           O  \\nATOM    555  CB  ASN A  77       9.985  -3.784   0.428  1.00 96.49           C  \\nATOM    556  CG  ASN A  77      11.391  -3.387   0.875  1.00 96.49           C  \\nATOM    557  OD1 ASN A  77      11.710  -2.200   0.989  1.00 96.49           O  \\nATOM    558  ND2 ASN A  77      12.240  -4.384   1.122  1.00 96.49           N  \\nATOM    559  N   ASP A  78       6.556  -3.874   1.285  1.00 72.36           N  \\nATOM    560  CA  ASP A  78       5.219  -4.028   0.700  1.00 72.36           C  \\nATOM    561  C   ASP A  78       4.327  -3.171   1.600  1.00 72.36           C  \\nATOM    562  O   ASP A  78       3.796  -3.643   2.609  1.00 72.36           O  \\nATOM    563  CB  ASP A  78       4.745  -5.479   0.716  1.00108.13           C  \\nATOM    564  CG  ASP A  78       3.377  -5.644   0.065  1.00108.13           C  \\nATOM    565  OD1 ASP A  78       3.225  -5.258  -1.112  1.00108.13           O  \\nATOM    566  OD2 ASP A  78       2.448  -6.153   0.728  1.00108.13           O  \\nATOM    567  N   ILE A  79       4.163  -1.910   1.214  1.00 35.39           N  \\nATOM    568  CA  ILE A  79       3.425  -0.943   2.015  1.00 35.39           C  \\nATOM    569  C   ILE A  79       2.025  -0.548   1.563  1.00 35.39           C  \\nATOM    570  O   ILE A  79       1.779  -0.336   0.382  1.00 35.39           O  \\nATOM    571  CB  ILE A  79       4.260   0.326   2.122  1.00 28.47           C  \\nATOM    572  CG1 ILE A  79       5.672  -0.058   2.552  1.00 28.47           C  \\nATOM    573  CG2 ILE A  79       3.630   1.303   3.070  1.00 28.47           C  \\nATOM    574  CD1 ILE A  79       6.681   1.064   2.358  1.00 28.47           C  \\nATOM    575  N   ASN A  80       1.120  -0.449   2.537  1.00 32.25           N  \\nATOM    576  CA  ASN A  80      -0.255  -0.026   2.299  1.00 32.25           C  \\nATOM    577  C   ASN A  80      -0.386   1.364   2.884  1.00 32.25           C  \\nATOM    578  O   ASN A  80      -0.009   1.599   4.028  1.00 32.25           O  \\nATOM    579  CB  ASN A  80      -1.256  -0.953   2.983  1.00 68.24           C  \\nATOM    580  CG  ASN A  80      -1.409  -2.275   2.265  1.00 68.24           C  \\nATOM    581  OD1 ASN A  80      -1.410  -3.341   2.885  1.00 68.24           O  \\nATOM    582  ND2 ASN A  80      -1.552  -2.215   0.949  1.00 68.24           N  \\nATOM    583  N   VAL A  81      -0.875   2.291   2.074  1.00 31.42           N  \\nATOM    584  CA  VAL A  81      -1.090   3.651   2.511  1.00 31.42           C  \\nATOM    585  C   VAL A  81      -2.586   3.842   2.393  1.00 31.42           C  \\nATOM    586  O   VAL A  81      -3.154   3.663   1.319  1.00 31.42           O  \\nATOM    587  CB  VAL A  81      -0.418   4.651   1.595  1.00 27.86           C  \\nATOM    588  CG1 VAL A  81      -0.828   6.027   1.980  1.00 27.86           C  \\nATOM    589  CG2 VAL A  81       1.089   4.522   1.694  1.00 27.86           C  \\nATOM    590  N   THR A  82      -3.226   4.180   3.499  1.00 39.50           N  \\nATOM    591  CA  THR A  82      -4.666   4.392   3.515  1.00 39.50           C  \\nATOM    592  C   THR A  82      -5.000   5.729   4.145  1.00 39.50           C  \\nATOM    593  O   THR A  82      -4.520   6.055   5.233  1.00 39.50           O  \\nATOM    594  CB  THR A  82      -5.398   3.315   4.339  1.00 46.66           C  \\nATOM    595  OG1 THR A  82      -5.223   2.023   3.741  1.00 46.66           O  \\nATOM    596  CG2 THR A  82      -6.867   3.630   4.400  1.00 46.66           C  \\nATOM    597  N   PHE A  83      -5.802   6.522   3.457  1.00 57.37           N  \\nATOM    598  CA  PHE A  83      -6.197   7.793   4.034  1.00 57.37           C  \\nATOM    599  C   PHE A  83      -7.634   7.636   4.481  1.00 57.37           C  \\nATOM    600  O   PHE A  83      -8.448   7.043   3.776  1.00 57.37           O  \\nATOM    601  CB  PHE A  83      -6.097   8.935   3.025  1.00 35.09           C  \\nATOM    602  CG  PHE A  83      -4.709   9.189   2.529  1.00 35.09           C  \\nATOM    603  CD1 PHE A  83      -4.182   8.445   1.477  1.00 35.09           C  \\nATOM    604  CD2 PHE A  83      -3.930  10.179   3.103  1.00 35.09           C  \\nATOM    605  CE1 PHE A  83      -2.893   8.688   0.998  1.00 35.09           C  \\nATOM    606  CE2 PHE A  83      -2.643  10.433   2.638  1.00 35.09           C  \\nATOM    607  CZ  PHE A  83      -2.123   9.681   1.578  1.00 35.09           C  \\nATOM    608  N   ASP A  84      -7.924   8.130   5.676  1.00 57.44           N  \\nATOM    609  CA  ASP A  84      -9.270   8.081   6.224  1.00 57.44           C  \\nATOM    610  C   ASP A  84      -9.498   9.482   6.767  1.00 57.44           C  \\nATOM    611  O   ASP A  84      -9.176   9.781   7.915  1.00 57.44           O  \\nATOM    612  CB  ASP A  84      -9.365   7.052   7.349  1.00115.57           C  \\nATOM    613  CG  ASP A  84     -10.794   6.628   7.625  1.00115.57           C  \\nATOM    614  OD1 ASP A  84     -11.645   7.511   7.866  1.00115.57           O  \\nATOM    615  OD2 ASP A  84     -11.068   5.410   7.597  1.00115.57           O  \\nATOM    616  N   GLY A  85     -10.026  10.352   5.924  1.00 51.12           N  \\nATOM    617  CA  GLY A  85     -10.250  11.713   6.358  1.00 51.12           C  \\nATOM    618  C   GLY A  85      -8.911  12.403   6.503  1.00 51.12           C  \\nATOM    619  O   GLY A  85      -8.212  12.608   5.510  1.00 51.12           O  \\nATOM    620  N   ASP A  86      -8.545  12.754   7.733  1.00 53.00           N  \\nATOM    621  CA  ASP A  86      -7.275  13.427   7.979  1.00 53.00           C  \\nATOM    622  C   ASP A  86      -6.266  12.493   8.617  1.00 53.00           C  \\nATOM    623  O   ASP A  86      -5.171  12.907   9.004  1.00 53.00           O  \\nATOM    624  CB  ASP A  86      -7.497  14.652   8.856  1.00 84.41           C  \\nATOM    625  CG  ASP A  86      -7.893  15.864   8.047  1.00 84.41           C  \\nATOM    626  OD1 ASP A  86      -8.499  16.790   8.622  1.00 84.41           O  \\nATOM    627  OD2 ASP A  86      -7.589  15.892   6.833  1.00 84.41           O  \\nATOM    628  N   THR A  87      -6.652  11.227   8.713  1.00 54.91           N  \\nATOM    629  CA  THR A  87      -5.799  10.202   9.280  1.00 54.91           C  \\nATOM    630  C   THR A  87      -5.142   9.365   8.190  1.00 54.91           C  \\nATOM    631  O   THR A  87      -5.826   8.797   7.346  1.00 54.91           O  \\nATOM    632  CB  THR A  87      -6.600   9.245  10.167  1.00 59.15           C  \\nATOM    633  OG1 THR A  87      -7.103   9.953  11.304  1.00 59.15           O  \\nATOM    634  CG2 THR A  87      -5.728   8.095  10.628  1.00 59.15           C  \\nATOM    635  N   VAL A  88      -3.816   9.303   8.191  1.00 36.77           N  \\nATOM    636  CA  VAL A  88      -3.125   8.459   7.218  1.00 36.77           C  \\nATOM    637  C   VAL A  88      -2.580   7.225   7.935  1.00 36.77           C  \\nATOM    638  O   VAL A  88      -1.997   7.316   9.019  1.00 36.77           O  \\nATOM    639  CB  VAL A  88      -1.953   9.176   6.533  1.00 32.82           C  \\nATOM    640  CG1 VAL A  88      -0.849   9.457   7.518  1.00 32.82           C  \\nATOM    641  CG2 VAL A  88      -1.459   8.320   5.381  1.00 32.82           C  \\nATOM    642  N   THR A  89      -2.790   6.067   7.330  1.00 49.29           N  \\nATOM    643  CA  THR A  89      -2.314   4.826   7.917  1.00 49.29           C  \\nATOM    644  C   THR A  89      -1.395   4.069   6.971  1.00 49.29           C  \\nATOM    645  O   THR A  89      -1.844   3.485   5.978  1.00 49.29           O  \\nATOM    646  CB  THR A  89      -3.476   3.934   8.283  1.00 51.33           C  \\nATOM    647  OG1 THR A  89      -4.392   4.694   9.073  1.00 51.33           O  \\nATOM    648  CG2 THR A  89      -2.999   2.722   9.060  1.00 51.33           C  \\nATOM    649  N   VAL A  90      -0.101   4.112   7.284  1.00 33.06           N  \\nATOM    650  CA  VAL A  90       0.909   3.414   6.508  1.00 33.06           C  \\nATOM    651  C   VAL A  90       1.260   2.131   7.257  1.00 33.06           C  \\nATOM    652  O   VAL A  90       1.736   2.164   8.392  1.00 33.06           O  \\nATOM    653  CB  VAL A  90       2.156   4.280   6.343  1.00 31.66           C  \\nATOM    654  CG1 VAL A  90       3.173   3.579   5.443  1.00 31.66           C  \\nATOM    655  CG2 VAL A  90       1.762   5.626   5.741  1.00 31.66           C  \\nATOM    656  N   GLU A  91       0.969   0.996   6.638  1.00 39.68           N  \\nATOM    657  CA  GLU A  91       1.283  -0.296   7.233  1.00 39.68           C  \\nATOM    658  C   GLU A  91       2.266  -1.019   6.321  1.00 39.68           C  \\nATOM    659  O   GLU A  91       2.032  -1.166   5.114  1.00 39.68           O  \\nATOM    660  CB  GLU A  91       0.045  -1.172   7.391  1.00115.49           C  \\nATOM    661  CG  GLU A  91       0.407  -2.556   7.905  1.00111.17           C  \\nATOM    662  CD  GLU A  91      -0.442  -3.656   7.308  1.00111.17           C  \\nATOM    663  OE1 GLU A  91      -1.618  -3.782   7.708  1.00111.17           O  \\nATOM    664  OE2 GLU A  91       0.066  -4.394   6.432  1.00111.17           O  \\nATOM    665  N   GLY A  92       3.369  -1.469   6.897  1.00 57.29           N  \\nATOM    666  CA  GLY A  92       4.345  -2.181   6.106  1.00 57.29           C  \\nATOM    667  C   GLY A  92       4.665  -3.567   6.629  1.00 57.29           C  \\nATOM    668  O   GLY A  92       4.434  -3.886   7.799  1.00 57.29           O  \\nATOM    669  N   GLN A  93       5.184  -4.402   5.733  1.00 51.61           N  \\nATOM    670  CA  GLN A  93       5.609  -5.756   6.069  1.00 51.61           C  \\nATOM    671  C   GLN A  93       7.071  -5.898   5.659  1.00 51.61           C  \\nATOM    672  O   GLN A  93       7.498  -5.373   4.628  1.00 51.61           O  \\nATOM    673  CB  GLN A  93       4.724  -6.792   5.376  1.00 87.10           C  \\nATOM    674  CG  GLN A  93       3.505  -7.152   6.212  1.00 87.10           C  \\nATOM    675  CD  GLN A  93       2.449  -7.895   5.433  1.00 87.10           C  \\nATOM    676  OE1 GLN A  93       1.866  -7.357   4.489  1.00 87.10           O  \\nATOM    677  NE2 GLN A  93       2.189  -9.139   5.823  1.00 87.10           N  \\nATOM    678  N   LEU A  94       7.836  -6.589   6.494  1.00 78.23           N  \\nATOM    679  CA  LEU A  94       9.264  -6.783   6.273  1.00 78.23           C  \\nATOM    680  C   LEU A  94       9.604  -7.603   5.021  1.00 78.23           C  \\nATOM    681  O   LEU A  94      10.488  -7.171   4.243  1.00 78.23           O  \\nATOM    682  CB  LEU A  94       9.865  -7.422   7.524  1.00 64.38           C  \\nATOM    683  CG  LEU A  94      11.295  -7.028   7.888  1.00 64.38           C  \\nATOM    684  CD1 LEU A  94      11.525  -5.554   7.645  1.00 64.38           C  \\nATOM    685  CD2 LEU A  94      11.541  -7.378   9.338  1.00 64.38           C  \\nTER     686      LEU A  94                                                      \\nHETATM  687  O   HOH A 107      10.102  -4.251  13.943  1.00 44.75           O  \\nHETATM  688  O   HOH A 108       1.041  -4.800   4.090  1.00 52.54           O  \\nHETATM  689  O   HOH A 109      -0.433   3.575  23.826  1.00 38.09           O  \\nHETATM  690  O   HOH A 110       2.878  19.537   9.279  1.00 61.82           O  \\nHETATM  691  O   HOH A 111      18.925   5.360  10.377  1.00 60.69           O  \\nHETATM  692  O   HOH A 112      -3.917   2.311  13.588  1.00 49.83           O  \\nHETATM  693  O   HOH A 113       4.748  12.641  22.041  1.00 58.74           O  \\nCONECT  250  253                                                                \\nCONECT  253  250  254                                                           \\nCONECT  254  253  255  257                                                      \\nCONECT  255  254  256  261                                                      \\nCONECT  256  255                                                                \\nCONECT  257  254  258                                                           \\nCONECT  258  257  259                                                           \\nCONECT  259  258  260                                                           \\nCONECT  260  259                                                                \\nCONECT  261  255                                                                \\nMASTER      303    0    1    2    5    0    0    6  692    1   10    9          \\nEND                                                                             \\n\");\n\tviewer_16908637293794932.setStyle({\"model\": -1},{\"cartoon\": {\"colorscheme\": {\"prop\": \"b\", \"map\": {\"0\": \"#FF7D45\", \"1\": \"#FFDB13\", \"2\": \"#65CBF3\", \"3\": \"#0053D6\"}}}, \"stick\": {}});\n\tviewer_16908637293794932.zoomTo();\nviewer_16908637293794932.render();\n});\n</script>",
      "text/html": [
       "<div id=\"3dmolviewer_16908637293794932\"  style=\"position: relative; width: 800px; height: 600px;\">\n",
       "        <p id=\"3dmolwarning_16908637293794932\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
       "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
       "        </div>\n",
       "<script>\n",
       "\n",
       "var loadScriptAsync = function(uri){\n",
       "  return new Promise((resolve, reject) => {\n",
       "    //this is to ignore the existence of requirejs amd\n",
       "    var savedexports, savedmodule;\n",
       "    if (typeof exports !== 'undefined') savedexports = exports;\n",
       "    else exports = {}\n",
       "    if (typeof module !== 'undefined') savedmodule = module;\n",
       "    else module = {}\n",
       "\n",
       "    var tag = document.createElement('script');\n",
       "    tag.src = uri;\n",
       "    tag.async = true;\n",
       "    tag.onload = () => {\n",
       "        exports = savedexports;\n",
       "        module = savedmodule;\n",
       "        resolve();\n",
       "    };\n",
       "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
       "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
       "});\n",
       "};\n",
       "\n",
       "if(typeof $3Dmolpromise === 'undefined') {\n",
       "$3Dmolpromise = null;\n",
       "  $3Dmolpromise = loadScriptAsync('https://cdnjs.cloudflare.com/ajax/libs/3Dmol/2.0.3/3Dmol-min.js');\n",
       "}\n",
       "\n",
       "var viewer_16908637293794932 = null;\n",
       "var warn = document.getElementById(\"3dmolwarning_16908637293794932\");\n",
       "if(warn) {\n",
       "    warn.parentNode.removeChild(warn);\n",
       "}\n",
       "$3Dmolpromise.then(function() {\n",
       "viewer_16908637293794932 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_16908637293794932\"),{backgroundColor:\"white\"});\n",
       "viewer_16908637293794932.zoomTo();\n",
       "\tviewer_16908637293794932.addModelsAsFrames(\"HEADER    DE NOVO PROTEIN                         11-SEP-03   1QYS              \\nTITLE     CRYSTAL STRUCTURE OF TOP7: A COMPUTATIONALLY DESIGNED                 \\nTITLE    2 PROTEIN WITH A NOVEL FOLD                                            \\nCOMPND    MOL_ID: 1;                                                            \\nCOMPND   2 MOLECULE: TOP7;                                                      \\nCOMPND   3 CHAIN: A;                                                            \\nCOMPND   4 ENGINEERED: YES                                                      \\nSOURCE    MOL_ID: 1;                                                            \\nSOURCE   2 ORGANISM_SCIENTIFIC: COMPUTATIONALLY DESIGNED SEQUENCE;              \\nSOURCE   3 EXPRESSION_SYSTEM: ESCHERICHIA COLI;                                 \\nSOURCE   4 EXPRESSION_SYSTEM_TAXID: 562;                                        \\nSOURCE   5 EXPRESSION_SYSTEM_STRAIN: BL21(DE3) PLYSS;                           \\nSOURCE   6 EXPRESSION_SYSTEM_VECTOR_TYPE: PLASMID;                              \\nSOURCE   7 EXPRESSION_SYSTEM_PLASMID: PET29B(+)                                 \\nKEYWDS    ALPHA-BETA, COMPUTATIONALLY DESIGNED, NOVEL FOLD, DE NOVO             \\nKEYWDS   2 PROTEIN                                                              \\nEXPDTA    X-RAY DIFFRACTION                                                     \\nAUTHOR    B.KUHLMAN,G.DANTAS,G.C.IRETON,G.VARANI,B.L.STODDARD,D.BAKER           \\nREVDAT   2   24-FEB-09 1QYS    1       VERSN                                    \\nREVDAT   1   25-NOV-03 1QYS    0                                                \\nJRNL        AUTH   B.KUHLMAN,G.DANTAS,G.C.IRETON,G.VARANI,                      \\nJRNL        AUTH 2 B.L.STODDARD,D.BAKER                                         \\nJRNL        TITL   DESIGN OF A NOVEL GLOBULAR PROTEIN FOLD WITH                 \\nJRNL        TITL 2 ATOMIC-LEVEL ACCURACY                                        \\nJRNL        REF    SCIENCE                       V. 302  1364 2003              \\nJRNL        REFN                   ISSN 0036-8075                               \\nJRNL        PMID   14631033                                                     \\nJRNL        DOI    10.1126/SCIENCE.1089427                                      \\nREMARK   1                                                                      \\nREMARK   2                                                                      \\nREMARK   2 RESOLUTION.    2.50 ANGSTROMS.                                       \\nREMARK   3                                                                      \\nREMARK   3 REFINEMENT.                                                          \\nREMARK   3   PROGRAM     : CNS 1.1                                              \\nREMARK   3   AUTHORS     : BRUNGER,ADAMS,CLORE,DELANO,GROS,GROSSE-              \\nREMARK   3               : KUNSTLEVE,JIANG,KUSZEWSKI,NILGES, PANNU,             \\nREMARK   3               : READ,RICE,SIMONSON,WARREN                            \\nREMARK   3                                                                      \\nREMARK   3  REFINEMENT TARGET : ENGH & HUBER                                    \\nREMARK   3                                                                      \\nREMARK   3  DATA USED IN REFINEMENT.                                            \\nREMARK   3   RESOLUTION RANGE HIGH (ANGSTROMS) : 2.50                           \\nREMARK   3   RESOLUTION RANGE LOW  (ANGSTROMS) : 18.71                          \\nREMARK   3   DATA CUTOFF            (SIGMA(F)) : 0.000                          \\nREMARK   3   DATA CUTOFF HIGH         (ABS(F)) : 1873860.400                    \\nREMARK   3   DATA CUTOFF LOW          (ABS(F)) : 0.0000                         \\nREMARK   3   COMPLETENESS (WORKING+TEST)   (%) : 96.1                           \\nREMARK   3   NUMBER OF REFLECTIONS             : 6736                           \\nREMARK   3                                                                      \\nREMARK   3  FIT TO DATA USED IN REFINEMENT.                                     \\nREMARK   3   CROSS-VALIDATION METHOD          : THROUGHOUT                      \\nREMARK   3   FREE R VALUE TEST SET SELECTION  : RANDOM                          \\nREMARK   3   R VALUE            (WORKING SET) : 0.268                           \\nREMARK   3   FREE R VALUE                     : 0.293                           \\nREMARK   3   FREE R VALUE TEST SET SIZE   (%) : 5.100                           \\nREMARK   3   FREE R VALUE TEST SET COUNT      : 346                             \\nREMARK   3   ESTIMATED ERROR OF FREE R VALUE  : 0.016                           \\nREMARK   3                                                                      \\nREMARK   3  FIT IN THE HIGHEST RESOLUTION BIN.                                  \\nREMARK   3   TOTAL NUMBER OF BINS USED           : 6                            \\nREMARK   3   BIN RESOLUTION RANGE HIGH       (A) : 2.50                         \\nREMARK   3   BIN RESOLUTION RANGE LOW        (A) : 2.66                         \\nREMARK   3   BIN COMPLETENESS (WORKING+TEST) (%) : 91.30                        \\nREMARK   3   REFLECTIONS IN BIN    (WORKING SET) : 1009                         \\nREMARK   3   BIN R VALUE           (WORKING SET) : 0.3530                       \\nREMARK   3   BIN FREE R VALUE                    : 0.3700                       \\nREMARK   3   BIN FREE R VALUE TEST SET SIZE  (%) : 4.30                         \\nREMARK   3   BIN FREE R VALUE TEST SET COUNT     : 45                           \\nREMARK   3   ESTIMATED ERROR OF BIN FREE R VALUE : 0.055                        \\nREMARK   3                                                                      \\nREMARK   3  NUMBER OF NON-HYDROGEN ATOMS USED IN REFINEMENT.                    \\nREMARK   3   PROTEIN ATOMS            : 685                                     \\nREMARK   3   NUCLEIC ACID ATOMS       : 0                                       \\nREMARK   3   HETEROGEN ATOMS          : 0                                       \\nREMARK   3   SOLVENT ATOMS            : 7                                       \\nREMARK   3                                                                      \\nREMARK   3  B VALUES.                                                           \\nREMARK   3   FROM WILSON PLOT           (A**2) : 30.20                          \\nREMARK   3   MEAN B VALUE      (OVERALL, A**2) : 65.50                          \\nREMARK   3   OVERALL ANISOTROPIC B VALUE.                                       \\nREMARK   3    B11 (A**2) : 10.54000                                             \\nREMARK   3    B22 (A**2) : 10.54000                                             \\nREMARK   3    B33 (A**2) : -21.07000                                            \\nREMARK   3    B12 (A**2) : 9.56000                                              \\nREMARK   3    B13 (A**2) : 0.00000                                              \\nREMARK   3    B23 (A**2) : 0.00000                                              \\nREMARK   3                                                                      \\nREMARK   3  ESTIMATED COORDINATE ERROR.                                         \\nREMARK   3   ESD FROM LUZZATI PLOT        (A) : 0.42                            \\nREMARK   3   ESD FROM SIGMAA              (A) : 0.47                            \\nREMARK   3   LOW RESOLUTION CUTOFF        (A) : 5.00                            \\nREMARK   3                                                                      \\nREMARK   3  CROSS-VALIDATED ESTIMATED COORDINATE ERROR.                         \\nREMARK   3   ESD FROM C-V LUZZATI PLOT    (A) : 0.46                            \\nREMARK   3   ESD FROM C-V SIGMAA          (A) : 0.45                            \\nREMARK   3                                                                      \\nREMARK   3  RMS DEVIATIONS FROM IDEAL VALUES.                                   \\nREMARK   3   BOND LENGTHS                 (A) : 0.008                           \\nREMARK   3   BOND ANGLES            (DEGREES) : 1.40                            \\nREMARK   3   DIHEDRAL ANGLES        (DEGREES) : 26.30                           \\nREMARK   3   IMPROPER ANGLES        (DEGREES) : 0.73                            \\nREMARK   3                                                                      \\nREMARK   3  ISOTROPIC THERMAL MODEL : GROUP                                     \\nREMARK   3                                                                      \\nREMARK   3  ISOTROPIC THERMAL FACTOR RESTRAINTS.    RMS    SIGMA                \\nREMARK   3   MAIN-CHAIN BOND              (A**2) : NULL  ; NULL                 \\nREMARK   3   MAIN-CHAIN ANGLE             (A**2) : NULL  ; NULL                 \\nREMARK   3   SIDE-CHAIN BOND              (A**2) : NULL  ; NULL                 \\nREMARK   3   SIDE-CHAIN ANGLE             (A**2) : NULL  ; NULL                 \\nREMARK   3                                                                      \\nREMARK   3  BULK SOLVENT MODELING.                                              \\nREMARK   3   METHOD USED : FLAT MODEL                                           \\nREMARK   3   KSOL        : 0.30                                                 \\nREMARK   3   BSOL        : 55.24                                                \\nREMARK   3                                                                      \\nREMARK   3  NCS MODEL : NULL                                                    \\nREMARK   3                                                                      \\nREMARK   3  NCS RESTRAINTS.                         RMS   SIGMA/WEIGHT          \\nREMARK   3   GROUP  1  POSITIONAL            (A) : NULL  ; NULL                 \\nREMARK   3   GROUP  1  B-FACTOR           (A**2) : NULL  ; NULL                 \\nREMARK   3                                                                      \\nREMARK   3  PARAMETER FILE  1  : PROTEIN_REP.PARAM                              \\nREMARK   3  PARAMETER FILE  2  : WATER_REP.PARAM                                \\nREMARK   3  PARAMETER FILE  3  : NULL                                           \\nREMARK   3  TOPOLOGY FILE  1   : PROTEIN.TOP                                    \\nREMARK   3  TOPOLOGY FILE  2   : WATER.TOP                                      \\nREMARK   3  TOPOLOGY FILE  3   : NULL                                           \\nREMARK   3                                                                      \\nREMARK   3  OTHER REFINEMENT REMARKS: NULL                                      \\nREMARK   4                                                                      \\nREMARK   4 1QYS COMPLIES WITH FORMAT V. 3.15, 01-DEC-08                         \\nREMARK 100                                                                      \\nREMARK 100 THIS ENTRY HAS BEEN PROCESSED BY RCSB ON 12-SEP-03.                  \\nREMARK 100 THE RCSB ID CODE IS RCSB020233.                                      \\nREMARK 200                                                                      \\nREMARK 200 EXPERIMENTAL DETAILS                                                 \\nREMARK 200  EXPERIMENT TYPE                : X-RAY DIFFRACTION                  \\nREMARK 200  DATE OF DATA COLLECTION        : 24-MAR-03                          \\nREMARK 200  TEMPERATURE           (KELVIN) : 100                                \\nREMARK 200  PH                             : 6.6                                \\nREMARK 200  NUMBER OF CRYSTALS USED        : 1                                  \\nREMARK 200                                                                      \\nREMARK 200  SYNCHROTRON              (Y/N) : Y                                  \\nREMARK 200  RADIATION SOURCE               : ALS                                \\nREMARK 200  BEAMLINE                       : 8.2.1                              \\nREMARK 200  X-RAY GENERATOR MODEL          : NULL                               \\nREMARK 200  MONOCHROMATIC OR LAUE    (M/L) : M                                  \\nREMARK 200  WAVELENGTH OR RANGE        (A) : 0.9793                             \\nREMARK 200  MONOCHROMATOR                  : DOUBLE CRYSTAL SI(111)             \\nREMARK 200  OPTICS                         : NULL                               \\nREMARK 200                                                                      \\nREMARK 200  DETECTOR TYPE                  : CCD                                \\nREMARK 200  DETECTOR MANUFACTURER          : ADSC QUANTUM 210                   \\nREMARK 200  INTENSITY-INTEGRATION SOFTWARE : HKL-2000                           \\nREMARK 200  DATA SCALING SOFTWARE          : SCALEPACK                          \\nREMARK 200                                                                      \\nREMARK 200  NUMBER OF UNIQUE REFLECTIONS   : 6979                               \\nREMARK 200  RESOLUTION RANGE HIGH      (A) : 2.500                              \\nREMARK 200  RESOLUTION RANGE LOW       (A) : 50.000                             \\nREMARK 200  REJECTION CRITERIA  (SIGMA(I)) : 0.000                              \\nREMARK 200                                                                      \\nREMARK 200 OVERALL.                                                             \\nREMARK 200  COMPLETENESS FOR RANGE     (%) : 100.0                              \\nREMARK 200  DATA REDUNDANCY                : NULL                               \\nREMARK 200  R MERGE                    (I) : 0.04500                            \\nREMARK 200  R SYM                      (I) : NULL                               \\nREMARK 200  <I/SIGMA(I)> FOR THE DATA SET  : 37.8000                            \\nREMARK 200                                                                      \\nREMARK 200 IN THE HIGHEST RESOLUTION SHELL.                                     \\nREMARK 200  HIGHEST RESOLUTION SHELL, RANGE HIGH (A) : 2.50                     \\nREMARK 200  HIGHEST RESOLUTION SHELL, RANGE LOW  (A) : 2.59                     \\nREMARK 200  COMPLETENESS FOR SHELL     (%) : 100.0                              \\nREMARK 200  DATA REDUNDANCY IN SHELL       : NULL                               \\nREMARK 200  R MERGE FOR SHELL          (I) : 0.34400                            \\nREMARK 200  R SYM FOR SHELL            (I) : NULL                               \\nREMARK 200  <I/SIGMA(I)> FOR SHELL         : 5.000                              \\nREMARK 200                                                                      \\nREMARK 200 DIFFRACTION PROTOCOL: SAD                                            \\nREMARK 200 METHOD USED TO DETERMINE THE STRUCTURE: SAD                          \\nREMARK 200 SOFTWARE USED: CNS                                                   \\nREMARK 200 STARTING MODEL: NULL                                                 \\nREMARK 200                                                                      \\nREMARK 200 REMARK: NULL                                                         \\nREMARK 280                                                                      \\nREMARK 280 CRYSTAL                                                              \\nREMARK 280 SOLVENT CONTENT, VS   (%): 42.94                                     \\nREMARK 280 MATTHEWS COEFFICIENT, VM (ANGSTROMS**3/DA): 2.16                     \\nREMARK 280                                                                      \\nREMARK 280 CRYSTALLIZATION CONDITIONS: 15-20% PEG3350 250MM AMMONIUM            \\nREMARK 280  FORMATE, PH 6.6, VAPOR DIFFUSION, HANGING DROP, STREAK              \\nREMARK 280  SEEDING, TEMPERATURE 298K                                           \\nREMARK 290                                                                      \\nREMARK 290 CRYSTALLOGRAPHIC SYMMETRY                                            \\nREMARK 290 SYMMETRY OPERATORS FOR SPACE GROUP: P 32 2 1                         \\nREMARK 290                                                                      \\nREMARK 290      SYMOP   SYMMETRY                                                \\nREMARK 290     NNNMMM   OPERATOR                                                \\nREMARK 290       1555   X,Y,Z                                                   \\nREMARK 290       2555   -Y,X-Y,Z+2/3                                            \\nREMARK 290       3555   -X+Y,-X,Z+1/3                                           \\nREMARK 290       4555   Y,X,-Z                                                  \\nREMARK 290       5555   X-Y,-Y,-Z+1/3                                           \\nREMARK 290       6555   -X,-X+Y,-Z+2/3                                          \\nREMARK 290                                                                      \\nREMARK 290     WHERE NNN -> OPERATOR NUMBER                                     \\nREMARK 290           MMM -> TRANSLATION VECTOR                                  \\nREMARK 290                                                                      \\nREMARK 290 CRYSTALLOGRAPHIC SYMMETRY TRANSFORMATIONS                            \\nREMARK 290 THE FOLLOWING TRANSFORMATIONS OPERATE ON THE ATOM/HETATM             \\nREMARK 290 RECORDS IN THIS ENTRY TO PRODUCE CRYSTALLOGRAPHICALLY                \\nREMARK 290 RELATED MOLECULES.                                                   \\nREMARK 290   SMTRY1   1  1.000000  0.000000  0.000000        0.00000            \\nREMARK 290   SMTRY2   1  0.000000  1.000000  0.000000        0.00000            \\nREMARK 290   SMTRY3   1  0.000000  0.000000  1.000000        0.00000            \\nREMARK 290   SMTRY1   2 -0.500000 -0.866025  0.000000        0.00000            \\nREMARK 290   SMTRY2   2  0.866025 -0.500000  0.000000        0.00000            \\nREMARK 290   SMTRY3   2  0.000000  0.000000  1.000000       93.70267            \\nREMARK 290   SMTRY1   3 -0.500000  0.866025  0.000000        0.00000            \\nREMARK 290   SMTRY2   3 -0.866025 -0.500000  0.000000        0.00000            \\nREMARK 290   SMTRY3   3  0.000000  0.000000  1.000000       46.85133            \\nREMARK 290   SMTRY1   4 -0.500000  0.866025  0.000000        0.00000            \\nREMARK 290   SMTRY2   4  0.866025  0.500000  0.000000        0.00000            \\nREMARK 290   SMTRY3   4  0.000000  0.000000 -1.000000        0.00000            \\nREMARK 290   SMTRY1   5  1.000000  0.000000  0.000000        0.00000            \\nREMARK 290   SMTRY2   5  0.000000 -1.000000  0.000000        0.00000            \\nREMARK 290   SMTRY3   5  0.000000  0.000000 -1.000000       46.85133            \\nREMARK 290   SMTRY1   6 -0.500000 -0.866025  0.000000        0.00000            \\nREMARK 290   SMTRY2   6 -0.866025  0.500000  0.000000        0.00000            \\nREMARK 290   SMTRY3   6  0.000000  0.000000 -1.000000       93.70267            \\nREMARK 290                                                                      \\nREMARK 290 REMARK: NULL                                                         \\nREMARK 300                                                                      \\nREMARK 300 BIOMOLECULE: 1                                                       \\nREMARK 300 SEE REMARK 350 FOR THE AUTHOR PROVIDED AND/OR PROGRAM                \\nREMARK 300 GENERATED ASSEMBLY INFORMATION FOR THE STRUCTURE IN                  \\nREMARK 300 THIS ENTRY. THE REMARK MAY ALSO PROVIDE INFORMATION ON               \\nREMARK 300 BURIED SURFACE AREA.                                                 \\nREMARK 350                                                                      \\nREMARK 350 COORDINATES FOR A COMPLETE MULTIMER REPRESENTING THE KNOWN           \\nREMARK 350 BIOLOGICALLY SIGNIFICANT OLIGOMERIZATION STATE OF THE                \\nREMARK 350 MOLECULE CAN BE GENERATED BY APPLYING BIOMT TRANSFORMATIONS          \\nREMARK 350 GIVEN BELOW.  BOTH NON-CRYSTALLOGRAPHIC AND                          \\nREMARK 350 CRYSTALLOGRAPHIC OPERATIONS ARE GIVEN.                               \\nREMARK 350                                                                      \\nREMARK 350 BIOMOLECULE: 1                                                       \\nREMARK 350 AUTHOR DETERMINED BIOLOGICAL UNIT: MONOMERIC                         \\nREMARK 350 APPLY THE FOLLOWING TO CHAINS: A                                     \\nREMARK 350   BIOMT1   1  1.000000  0.000000  0.000000        0.00000            \\nREMARK 350   BIOMT2   1  0.000000  1.000000  0.000000        0.00000            \\nREMARK 350   BIOMT3   1  0.000000  0.000000  1.000000        0.00000            \\nREMARK 465                                                                      \\nREMARK 465 MISSING RESIDUES                                                     \\nREMARK 465 THE FOLLOWING RESIDUES WERE NOT LOCATED IN THE                       \\nREMARK 465 EXPERIMENT. (M=MODEL NUMBER; RES=RESIDUE NAME; C=CHAIN               \\nREMARK 465 IDENTIFIER; SSSEQ=SEQUENCE NUMBER; I=INSERTION CODE.)                \\nREMARK 465                                                                      \\nREMARK 465   M RES C SSSEQI                                                     \\nREMARK 465     MSE A     1                                                      \\nREMARK 465     GLY A     2                                                      \\nREMARK 465     GLU A    95                                                      \\nREMARK 465     GLY A    96                                                      \\nREMARK 465     GLY A    97                                                      \\nREMARK 465     SER A    98                                                      \\nREMARK 465     LEU A    99                                                      \\nREMARK 465     GLU A   100                                                      \\nREMARK 465     HIS A   101                                                      \\nREMARK 465     HIS A   102                                                      \\nREMARK 465     HIS A   103                                                      \\nREMARK 465     HIS A   104                                                      \\nREMARK 465     HIS A   105                                                      \\nREMARK 465     HIS A   106                                                      \\nREMARK 470                                                                      \\nREMARK 470 MISSING ATOM                                                         \\nREMARK 470 THE FOLLOWING RESIDUES HAVE MISSING ATOMS(M=MODEL NUMBER;            \\nREMARK 470 RES=RESIDUE NAME; C=CHAIN IDENTIFIER; SSEQ=SEQUENCE NUMBER;          \\nREMARK 470 I=INSERTION CODE):                                                   \\nREMARK 470   M RES CSSEQI  ATOMS                                                \\nREMARK 470     LYS A  15    CG   CD   CE   NZ                                   \\nREMARK 470     PHE A  17    CG   CD1  CD2  CE1  CE2  CZ                         \\nREMARK 470     SER A  27    OG                                                  \\nREMARK 470     GLN A  30    CG   CD   OE1  NE2                                  \\nREMARK 470     LYS A  31    CG   CD   CE   NZ                                   \\nREMARK 470     ASN A  34    CG   OD1  ND2                                       \\nREMARK 470     LEU A  36    CG   CD1  CD2                                       \\nREMARK 470     LYS A  46    CG   CD   CE   NZ                                   \\nREMARK 470     ARG A  47    CG   CD   NE   CZ   NH1  NH2                        \\nREMARK 470     ARG A  55    CG   CD   NE   CZ   NH1  NH2                        \\nREMARK 470     LYS A  62    CG   CD   CE   NZ                                   \\nREMARK 470     GLU A  73    CG   CD   OE1  OE2                                  \\nREMARK 500                                                                      \\nREMARK 500 GEOMETRY AND STEREOCHEMISTRY                                         \\nREMARK 500 SUBTOPIC: CLOSE CONTACTS                                             \\nREMARK 500                                                                      \\nREMARK 500 THE FOLLOWING ATOMS THAT ARE RELATED BY CRYSTALLOGRAPHIC             \\nREMARK 500 SYMMETRY ARE IN CLOSE CONTACT.  AN ATOM LOCATED WITHIN 0.15          \\nREMARK 500 ANGSTROMS OF A SYMMETRY RELATED ATOM IS ASSUMED TO BE ON A           \\nREMARK 500 SPECIAL POSITION AND IS, THEREFORE, LISTED IN REMARK 375             \\nREMARK 500 INSTEAD OF REMARK 500.  ATOMS WITH NON-BLANK ALTERNATE               \\nREMARK 500 LOCATION INDICATORS ARE NOT INCLUDED IN THE CALCULATIONS.            \\nREMARK 500                                                                      \\nREMARK 500 DISTANCE CUTOFF:                                                     \\nREMARK 500 2.2 ANGSTROMS FOR CONTACTS NOT INVOLVING HYDROGEN ATOMS              \\nREMARK 500 1.6 ANGSTROMS FOR CONTACTS INVOLVING HYDROGEN ATOMS                  \\nREMARK 500                                                                      \\nREMARK 500  ATM1  RES C  SSEQI   ATM2  RES C  SSEQI  SSYMOP   DISTANCE          \\nREMARK 500   ND2  ASN A    80     ND2  ASN A    80     4555     1.96            \\nREMARK 500   CD1  ILE A    68     CD1  ILE A    68     4555     1.98            \\nREMARK 500                                                                      \\nREMARK 500 REMARK: NULL                                                         \\nREMARK 500                                                                      \\nREMARK 500 GEOMETRY AND STEREOCHEMISTRY                                         \\nREMARK 500 SUBTOPIC: TORSION ANGLES                                             \\nREMARK 500                                                                      \\nREMARK 500 TORSION ANGLES OUTSIDE THE EXPECTED RAMACHANDRAN REGIONS:            \\nREMARK 500 (M=MODEL NUMBER; RES=RESIDUE NAME; C=CHAIN IDENTIFIER;               \\nREMARK 500 SSEQ=SEQUENCE NUMBER; I=INSERTION CODE).                             \\nREMARK 500                                                                      \\nREMARK 500 STANDARD TABLE:                                                      \\nREMARK 500 FORMAT:(10X,I3,1X,A3,1X,A1,I4,A1,4X,F7.2,3X,F7.2)                    \\nREMARK 500                                                                      \\nREMARK 500 EXPECTED VALUES: GJ KLEYWEGT AND TA JONES (1996). PHI/PSI-           \\nREMARK 500 CHOLOGY: RAMACHANDRAN REVISITED. STRUCTURE 4, 1395 - 1400            \\nREMARK 500                                                                      \\nREMARK 500  M RES CSSEQI        PSI       PHI                                   \\nREMARK 500    ASP A  12       80.22   -168.83                                   \\nREMARK 500    ASN A  13      102.49     43.86                                   \\nREMARK 500    THR A  24       81.00   -158.71                                   \\nREMARK 500    THR A  25      -36.94   -176.78                                   \\nREMARK 500    GLU A  26       -7.45     68.87                                   \\nREMARK 500                                                                      \\nREMARK 500 REMARK: NULL                                                         \\nDBREF  1QYS A    1   106  PDB    1QYS     1QYS             1    106             \\nSEQRES   1 A  106  MSE GLY ASP ILE GLN VAL GLN VAL ASN ILE ASP ASP ASN          \\nSEQRES   2 A  106  GLY LYS ASN PHE ASP TYR THR TYR THR VAL THR THR GLU          \\nSEQRES   3 A  106  SER GLU LEU GLN LYS VAL LEU ASN GLU LEU MSE ASP TYR          \\nSEQRES   4 A  106  ILE LYS LYS GLN GLY ALA LYS ARG VAL ARG ILE SER ILE          \\nSEQRES   5 A  106  THR ALA ARG THR LYS LYS GLU ALA GLU LYS PHE ALA ALA          \\nSEQRES   6 A  106  ILE LEU ILE LYS VAL PHE ALA GLU LEU GLY TYR ASN ASP          \\nSEQRES   7 A  106  ILE ASN VAL THR PHE ASP GLY ASP THR VAL THR VAL GLU          \\nSEQRES   8 A  106  GLY GLN LEU GLU GLY GLY SER LEU GLU HIS HIS HIS HIS          \\nSEQRES   9 A  106  HIS HIS                                                      \\nMODRES 1QYS MSE A   37  MET  SELENOMETHIONINE                                   \\nHET    MSE  A  37       8                                                       \\nHETNAM     MSE SELENOMETHIONINE                                                 \\nFORMUL   1  MSE    C5 H11 N O2 SE                                               \\nFORMUL   2  HOH   *7(H2 O)                                                      \\nHELIX    1   1 SER A   27  GLY A   44  1                                  18    \\nHELIX    2   2 THR A   56  LEU A   74  1                                  19    \\nSHEET    1   A 5 ASN A  16  VAL A  23  0                                        \\nSHEET    2   A 5 ILE A   4  ASP A  11 -1  N  VAL A   8   O  TYR A  19           \\nSHEET    3   A 5 ARG A  47  THR A  53 -1  O  ARG A  49   N  ASN A   9           \\nSHEET    4   A 5 THR A  87  GLN A  93 -1  O  GLY A  92   N  VAL A  48           \\nSHEET    5   A 5 ASP A  78  ASP A  84 -1  N  THR A  82   O  THR A  89           \\nLINK         C   LEU A  36                 N   MSE A  37     1555   1555  1.33  \\nLINK         C   MSE A  37                 N   ASP A  38     1555   1555  1.33  \\nCRYST1   35.900   35.900  140.554  90.00  90.00 120.00 P 32 2 1      6          \\nORIGX1      1.000000  0.000000  0.000000        0.00000                         \\nORIGX2      0.000000  1.000000  0.000000        0.00000                         \\nORIGX3      0.000000  0.000000  1.000000        0.00000                         \\nSCALE1      0.027855  0.016082  0.000000        0.00000                         \\nSCALE2      0.000000  0.032164  0.000000        0.00000                         \\nSCALE3      0.000000  0.000000  0.007115        0.00000                         \\nATOM      1  N   ASP A   3      -4.522  18.306  17.409  1.00174.51           N  \\nATOM      2  CA  ASP A   3      -3.061  18.228  17.122  1.00174.51           C  \\nATOM      3  C   ASP A   3      -2.664  16.993  16.324  1.00174.51           C  \\nATOM      4  O   ASP A   3      -3.515  16.306  15.754  1.00174.51           O  \\nATOM      5  CB  ASP A   3      -2.261  18.246  18.422  1.00 83.84           C  \\nATOM      6  CG  ASP A   3      -1.658  19.600  18.711  1.00 83.84           C  \\nATOM      7  OD1 ASP A   3      -1.169  20.249  17.760  1.00 83.84           O  \\nATOM      8  OD2 ASP A   3      -1.654  20.007  19.892  1.00 83.84           O  \\nATOM      9  N   ILE A   4      -1.360  16.714  16.297  1.00 57.73           N  \\nATOM     10  CA  ILE A   4      -0.823  15.562  15.568  1.00 57.73           C  \\nATOM     11  C   ILE A   4      -0.721  14.309  16.433  1.00 57.73           C  \\nATOM     12  O   ILE A   4       0.091  14.222  17.355  1.00 57.73           O  \\nATOM     13  CB  ILE A   4       0.555  15.888  14.980  1.00 57.14           C  \\nATOM     14  CG1 ILE A   4       0.425  17.108  14.058  1.00 57.14           C  \\nATOM     15  CG2 ILE A   4       1.097  14.674  14.218  1.00 57.14           C  \\nATOM     16  CD1 ILE A   4       1.737  17.831  13.766  1.00 57.14           C  \\nATOM     17  N   GLN A   5      -1.567  13.342  16.105  1.00 49.21           N  \\nATOM     18  CA  GLN A   5      -1.632  12.078  16.814  1.00 49.21           C  \\nATOM     19  C   GLN A   5      -0.907  10.950  16.066  1.00 49.21           C  \\nATOM     20  O   GLN A   5      -1.306  10.567  14.965  1.00 84.12           O  \\nATOM     21  CB  GLN A   5      -3.095  11.691  17.017  1.00 87.59           C  \\nATOM     22  CG  GLN A   5      -3.284  10.472  17.883  1.00 87.59           C  \\nATOM     23  CD  GLN A   5      -2.740  10.671  19.282  1.00 87.59           C  \\nATOM     24  OE1 GLN A   5      -2.801   9.770  20.107  1.00 87.59           O  \\nATOM     25  NE2 GLN A   5      -2.209  11.857  19.556  1.00 87.59           N  \\nATOM     26  N   VAL A   6       0.151  10.426  16.683  1.00 61.01           N  \\nATOM     27  CA  VAL A   6       0.937   9.342  16.107  1.00 61.01           C  \\nATOM     28  C   VAL A   6       0.671   8.008  16.803  1.00 61.01           C  \\nATOM     29  O   VAL A   6       0.957   7.837  17.983  1.00 61.01           O  \\nATOM     30  CB  VAL A   6       2.439   9.624  16.196  1.00 50.93           C  \\nATOM     31  CG1 VAL A   6       3.199   8.503  15.531  1.00 50.93           C  \\nATOM     32  CG2 VAL A   6       2.772  10.922  15.516  1.00 50.93           C  \\nATOM     33  N   GLN A   7       0.118   7.066  16.055  1.00 44.88           N  \\nATOM     34  CA  GLN A   7      -0.196   5.735  16.561  1.00 44.88           C  \\nATOM     35  C   GLN A   7       0.719   4.713  15.881  1.00 44.88           C  \\nATOM     36  O   GLN A   7       0.658   4.513  14.674  1.00 44.88           O  \\nATOM     37  CB  GLN A   7      -1.658   5.390  16.264  1.00 83.92           C  \\nATOM     38  CG  GLN A   7      -2.534   5.302  17.485  1.00 83.92           C  \\nATOM     39  CD  GLN A   7      -2.119   4.173  18.412  1.00 83.92           C  \\nATOM     40  OE1 GLN A   7      -0.990   4.138  18.905  1.00 83.92           O  \\nATOM     41  NE2 GLN A   7      -3.031   3.241  18.653  1.00 83.92           N  \\nATOM     42  N   VAL A   8       1.578   4.084  16.673  1.00 45.42           N  \\nATOM     43  CA  VAL A   8       2.499   3.068  16.186  1.00 45.42           C  \\nATOM     44  C   VAL A   8       2.103   1.700  16.756  1.00 45.42           C  \\nATOM     45  O   VAL A   8       2.114   1.508  17.966  1.00 45.42           O  \\nATOM     46  CB  VAL A   8       3.945   3.377  16.613  1.00 41.58           C  \\nATOM     47  CG1 VAL A   8       4.867   2.293  16.086  1.00 41.58           C  \\nATOM     48  CG2 VAL A   8       4.375   4.731  16.083  1.00 41.58           C  \\nATOM     49  N   ASN A   9       1.746   0.755  15.891  1.00 56.60           N  \\nATOM     50  CA  ASN A   9       1.357  -0.569  16.361  1.00 56.60           C  \\nATOM     51  C   ASN A   9       2.224  -1.684  15.748  1.00 56.60           C  \\nATOM     52  O   ASN A   9       2.140  -1.992  14.559  1.00 56.60           O  \\nATOM     53  CB  ASN A   9      -0.119  -0.818  16.065  1.00105.47           C  \\nATOM     54  CG  ASN A   9      -0.823  -1.519  17.208  1.00105.47           C  \\nATOM     55  OD1 ASN A   9      -1.281  -0.879  18.165  1.00105.47           O  \\nATOM     56  ND2 ASN A   9      -0.887  -2.850  17.133  1.00105.47           N  \\nATOM     57  N   ILE A  10       3.073  -2.269  16.587  1.00 40.88           N  \\nATOM     58  CA  ILE A  10       3.976  -3.337  16.187  1.00 40.88           C  \\nATOM     59  C   ILE A  10       3.531  -4.697  16.730  1.00 40.88           C  \\nATOM     60  O   ILE A  10       3.465  -4.900  17.944  1.00 70.73           O  \\nATOM     61  CB  ILE A  10       5.399  -3.068  16.702  1.00 61.80           C  \\nATOM     62  CG1 ILE A  10       5.870  -1.693  16.253  1.00 61.80           C  \\nATOM     63  CG2 ILE A  10       6.343  -4.127  16.180  1.00 61.80           C  \\nATOM     64  CD1 ILE A  10       7.241  -1.341  16.788  1.00 61.80           C  \\nATOM     65  N   ASP A  11       3.225  -5.621  15.824  1.00 68.81           N  \\nATOM     66  CA  ASP A  11       2.802  -6.960  16.220  1.00 68.81           C  \\nATOM     67  C   ASP A  11       3.892  -7.970  15.878  1.00 68.81           C  \\nATOM     68  O   ASP A  11       4.569  -7.850  14.850  1.00 68.81           O  \\nATOM     69  CB  ASP A  11       1.498  -7.342  15.537  1.00114.14           C  \\nATOM     70  CG  ASP A  11       0.590  -8.125  16.449  1.00114.14           C  \\nATOM     71  OD1 ASP A  11       1.008  -9.205  16.916  1.00114.14           O  \\nATOM     72  OD2 ASP A  11      -0.534  -7.651  16.712  1.00114.14           O  \\nATOM     73  N   ASP A  12       4.055  -8.971  16.737  1.00 70.35           N  \\nATOM     74  CA  ASP A  12       5.107  -9.950  16.527  1.00 70.35           C  \\nATOM     75  C   ASP A  12       4.970 -11.157  17.446  1.00 70.35           C  \\nATOM     76  O   ASP A  12       5.649 -11.243  18.468  1.00 70.35           O  \\nATOM     77  CB  ASP A  12       6.462  -9.264  16.752  1.00 53.31           C  \\nATOM     78  CG  ASP A  12       7.632 -10.096  16.286  1.00 53.31           C  \\nATOM     79  OD1 ASP A  12       7.496 -10.774  15.243  1.00 53.31           O  \\nATOM     80  OD2 ASP A  12       8.697 -10.045  16.944  1.00 53.31           O  \\nATOM     81  N   ASN A  13       4.090 -12.081  17.079  1.00 96.17           N  \\nATOM     82  CA  ASN A  13       3.876 -13.307  17.848  1.00 96.17           C  \\nATOM     83  C   ASN A  13       3.792 -13.129  19.361  1.00 96.17           C  \\nATOM     84  O   ASN A  13       4.817 -12.980  20.028  1.00 96.17           O  \\nATOM     85  CB  ASN A  13       4.988 -14.315  17.550  1.00100.93           C  \\nATOM     86  CG  ASN A  13       5.349 -14.363  16.087  1.00100.93           C  \\nATOM     87  OD1 ASN A  13       5.977 -13.444  15.562  1.00100.93           O  \\nATOM     88  ND2 ASN A  13       4.947 -15.434  15.413  1.00100.93           N  \\nATOM     89  N   GLY A  14       2.572 -13.164  19.895  1.00130.92           N  \\nATOM     90  CA  GLY A  14       2.371 -13.029  21.330  1.00130.92           C  \\nATOM     91  C   GLY A  14       2.938 -11.774  21.965  1.00130.92           C  \\nATOM     92  O   GLY A  14       3.099 -11.711  23.185  1.00130.92           O  \\nATOM     93  N   LYS A  15       3.245 -10.775  21.143  1.00 68.27           N  \\nATOM     94  CA  LYS A  15       3.790  -9.520  21.640  1.00 68.27           C  \\nATOM     95  C   LYS A  15       3.313  -8.363  20.774  1.00 68.27           C  \\nATOM     96  O   LYS A  15       3.754  -8.207  19.639  1.00 68.27           O  \\nATOM     97  CB  LYS A  15       5.329  -9.570  21.663  1.00 55.99           C  \\nATOM     98  N   ASN A  16       2.400  -7.560  21.313  1.00 54.52           N  \\nATOM     99  CA  ASN A  16       1.872  -6.404  20.600  1.00 54.52           C  \\nATOM    100  C   ASN A  16       2.268  -5.109  21.338  1.00 54.52           C  \\nATOM    101  O   ASN A  16       2.022  -4.958  22.538  1.00 54.52           O  \\nATOM    102  CB  ASN A  16       0.347  -6.516  20.481  1.00102.09           C  \\nATOM    103  CG  ASN A  16      -0.260  -5.401  19.643  1.00102.09           C  \\nATOM    104  OD1 ASN A  16       0.179  -5.140  18.522  1.00102.09           O  \\nATOM    105  ND2 ASN A  16      -1.284  -4.745  20.181  1.00102.09           N  \\nATOM    106  N   PHE A  17       2.902  -4.187  20.617  1.00 42.25           N  \\nATOM    107  CA  PHE A  17       3.318  -2.920  21.197  1.00 42.25           C  \\nATOM    108  C   PHE A  17       2.488  -1.820  20.560  1.00 42.25           C  \\nATOM    109  O   PHE A  17       2.326  -1.787  19.346  1.00 42.25           O  \\nATOM    110  CB  PHE A  17       4.810  -2.682  20.947  1.00 34.85           C  \\nATOM    111  N   ASP A  18       1.947  -0.934  21.394  1.00 45.89           N  \\nATOM    112  CA  ASP A  18       1.124   0.181  20.939  1.00 45.89           C  \\nATOM    113  C   ASP A  18       1.671   1.527  21.481  1.00 45.89           C  \\nATOM    114  O   ASP A  18       1.496   1.855  22.652  1.00 45.89           O  \\nATOM    115  CB  ASP A  18      -0.310  -0.037  21.406  1.00 89.81           C  \\nATOM    116  CG  ASP A  18      -1.254   0.966  20.813  1.00 89.81           C  \\nATOM    117  OD1 ASP A  18      -1.419   0.953  19.573  1.00 89.81           O  \\nATOM    118  OD2 ASP A  18      -1.816   1.782  21.577  1.00 89.81           O  \\nATOM    119  N   TYR A  19       2.343   2.290  20.623  1.00 43.94           N  \\nATOM    120  CA  TYR A  19       2.936   3.578  20.999  1.00 43.94           C  \\nATOM    121  C   TYR A  19       2.048   4.728  20.530  1.00 43.94           C  \\nATOM    122  O   TYR A  19       1.694   4.814  19.360  1.00 78.71           O  \\nATOM    123  CB  TYR A  19       4.329   3.734  20.373  1.00 87.94           C  \\nATOM    124  CG  TYR A  19       5.347   2.679  20.778  1.00 87.94           C  \\nATOM    125  CD1 TYR A  19       6.083   2.805  21.954  1.00 87.94           C  \\nATOM    126  CD2 TYR A  19       5.578   1.553  19.977  1.00 87.94           C  \\nATOM    127  CE1 TYR A  19       7.029   1.837  22.327  1.00 87.94           C  \\nATOM    128  CE2 TYR A  19       6.521   0.577  20.342  1.00 87.94           C  \\nATOM    129  CZ  TYR A  19       7.243   0.728  21.518  1.00 87.94           C  \\nATOM    130  OH  TYR A  19       8.180  -0.218  21.886  1.00 87.94           O  \\nATOM    131  N   THR A  20       1.691   5.609  21.457  1.00 50.11           N  \\nATOM    132  CA  THR A  20       0.857   6.756  21.147  1.00 50.11           C  \\nATOM    133  C   THR A  20       1.589   8.007  21.589  1.00 50.11           C  \\nATOM    134  O   THR A  20       1.996   8.119  22.746  1.00 50.11           O  \\nATOM    135  CB  THR A  20      -0.458   6.683  21.889  1.00 57.60           C  \\nATOM    136  OG1 THR A  20      -0.954   5.345  21.830  1.00 57.60           O  \\nATOM    137  CG2 THR A  20      -1.468   7.592  21.246  1.00 57.60           C  \\nATOM    138  N   TYR A  21       1.784   8.929  20.649  1.00 57.65           N  \\nATOM    139  CA  TYR A  21       2.468  10.196  20.916  1.00 57.65           C  \\nATOM    140  C   TYR A  21       1.581  11.325  20.413  1.00 57.65           C  \\nATOM    141  O   TYR A  21       0.683  11.120  19.599  1.00 57.65           O  \\nATOM    142  CB  TYR A  21       3.799  10.331  20.154  1.00 63.74           C  \\nATOM    143  CG  TYR A  21       4.810   9.204  20.258  1.00 63.74           C  \\nATOM    144  CD1 TYR A  21       4.560   7.953  19.684  1.00 63.74           C  \\nATOM    145  CD2 TYR A  21       6.060   9.422  20.845  1.00 63.74           C  \\nATOM    146  CE1 TYR A  21       5.525   6.959  19.691  1.00 63.74           C  \\nATOM    147  CE2 TYR A  21       7.031   8.434  20.854  1.00 63.74           C  \\nATOM    148  CZ  TYR A  21       6.759   7.209  20.271  1.00 63.74           C  \\nATOM    149  OH  TYR A  21       7.735   6.240  20.243  1.00 63.74           O  \\nATOM    150  N   THR A  22       1.867  12.525  20.893  1.00 67.33           N  \\nATOM    151  CA  THR A  22       1.136  13.706  20.489  1.00 67.33           C  \\nATOM    152  C   THR A  22       2.205  14.758  20.356  1.00 67.33           C  \\nATOM    153  O   THR A  22       2.833  15.129  21.342  1.00 67.33           O  \\nATOM    154  CB  THR A  22       0.144  14.146  21.558  1.00 65.03           C  \\nATOM    155  OG1 THR A  22      -0.589  13.007  22.026  1.00 65.03           O  \\nATOM    156  CG2 THR A  22      -0.826  15.159  20.984  1.00 65.03           C  \\nATOM    157  N   VAL A  23       2.444  15.212  19.135  1.00 71.70           N  \\nATOM    158  CA  VAL A  23       3.462  16.223  18.921  1.00 71.70           C  \\nATOM    159  C   VAL A  23       2.881  17.499  18.343  1.00 71.70           C  \\nATOM    160  O   VAL A  23       1.714  17.547  17.935  1.00 71.70           O  \\nATOM    161  CB  VAL A  23       4.574  15.719  17.970  1.00 70.78           C  \\nATOM    162  CG1 VAL A  23       5.298  14.540  18.591  1.00 70.78           C  \\nATOM    163  CG2 VAL A  23       3.976  15.342  16.638  1.00 70.78           C  \\nATOM    164  N   THR A  24       3.717  18.531  18.327  1.00 83.96           N  \\nATOM    165  CA  THR A  24       3.369  19.839  17.793  1.00 83.96           C  \\nATOM    166  C   THR A  24       4.686  20.523  17.473  1.00 83.96           C  \\nATOM    167  O   THR A  24       5.175  21.315  18.270  1.00 83.96           O  \\nATOM    168  CB  THR A  24       2.604  20.696  18.831  1.00 95.26           C  \\nATOM    169  OG1 THR A  24       1.323  20.110  19.084  1.00 95.26           O  \\nATOM    170  CG2 THR A  24       2.399  22.114  18.317  1.00 95.26           C  \\nATOM    171  N   THR A  25       5.280  20.203  16.327  1.00136.13           N  \\nATOM    172  CA  THR A  25       6.552  20.818  15.953  1.00136.13           C  \\nATOM    173  C   THR A  25       7.033  20.385  14.576  1.00136.13           C  \\nATOM    174  O   THR A  25       7.608  21.179  13.834  1.00136.13           O  \\nATOM    175  CB  THR A  25       7.703  20.449  16.943  1.00146.88           C  \\nATOM    176  OG1 THR A  25       7.172  20.180  18.244  1.00146.88           O  \\nATOM    177  CG2 THR A  25       8.697  21.598  17.058  1.00146.88           C  \\nATOM    178  N   GLU A  26       6.796  19.119  14.247  1.00187.15           N  \\nATOM    179  CA  GLU A  26       7.267  18.538  12.994  1.00187.15           C  \\nATOM    180  C   GLU A  26       8.782  18.481  13.170  1.00187.15           C  \\nATOM    181  O   GLU A  26       9.508  17.917  12.351  1.00187.15           O  \\nATOM    182  CB  GLU A  26       6.892  19.396  11.781  1.00134.71           C  \\nATOM    183  CG  GLU A  26       7.383  18.807  10.458  1.00 97.04           C  \\nATOM    184  CD  GLU A  26       6.713  19.422   9.243  1.00 97.04           C  \\nATOM    185  OE1 GLU A  26       6.740  20.662   9.115  1.00 97.04           O  \\nATOM    186  OE2 GLU A  26       6.165  18.662   8.416  1.00 97.04           O  \\nATOM    187  N   SER A  27       9.234  19.085  14.266  1.00114.54           N  \\nATOM    188  CA  SER A  27      10.637  19.110  14.647  1.00114.54           C  \\nATOM    189  C   SER A  27      10.764  18.025  15.711  1.00114.54           C  \\nATOM    190  O   SER A  27      11.634  17.158  15.624  1.00114.54           O  \\nATOM    191  CB  SER A  27      11.008  20.463  15.229  1.00116.53           C  \\nATOM    192  N   GLU A  28       9.885  18.071  16.713  1.00113.18           N  \\nATOM    193  CA  GLU A  28       9.900  17.055  17.762  1.00113.18           C  \\nATOM    194  C   GLU A  28       9.258  15.810  17.168  1.00113.18           C  \\nATOM    195  O   GLU A  28       9.094  14.794  17.842  1.00200.03           O  \\nATOM    196  CB  GLU A  28       9.116  17.505  19.003  1.00112.19           C  \\nATOM    197  CG  GLU A  28       7.596  17.557  18.841  1.00112.19           C  \\nATOM    198  CD  GLU A  28       6.879  17.960  20.129  1.00112.19           C  \\nATOM    199  OE1 GLU A  28       5.778  18.547  20.047  1.00112.19           O  \\nATOM    200  OE2 GLU A  28       7.407  17.679  21.226  1.00112.19           O  \\nATOM    201  N   LEU A  29       8.883  15.919  15.895  1.00 97.20           N  \\nATOM    202  CA  LEU A  29       8.286  14.820  15.147  1.00 97.20           C  \\nATOM    203  C   LEU A  29       9.475  14.012  14.649  1.00 97.20           C  \\nATOM    204  O   LEU A  29       9.456  12.784  14.643  1.00 97.20           O  \\nATOM    205  CB  LEU A  29       7.493  15.350  13.947  1.00 66.59           C  \\nATOM    206  CG  LEU A  29       6.452  14.423  13.296  1.00 66.59           C  \\nATOM    207  CD1 LEU A  29       6.087  14.961  11.915  1.00 66.59           C  \\nATOM    208  CD2 LEU A  29       6.997  13.012  13.162  1.00 66.59           C  \\nATOM    209  N   GLN A  30      10.512  14.723  14.225  1.00113.13           N  \\nATOM    210  CA  GLN A  30      11.724  14.081  13.747  1.00113.13           C  \\nATOM    211  C   GLN A  30      12.310  13.296  14.916  1.00113.13           C  \\nATOM    212  O   GLN A  30      13.031  12.315  14.725  1.00113.13           O  \\nATOM    213  CB  GLN A  30      12.718  15.135  13.256  1.00 67.53           C  \\nATOM    214  N   LYS A  31      11.984  13.738  16.128  1.00 75.23           N  \\nATOM    215  CA  LYS A  31      12.465  13.094  17.348  1.00 75.23           C  \\nATOM    216  C   LYS A  31      11.731  11.776  17.556  1.00 75.23           C  \\nATOM    217  O   LYS A  31      12.331  10.751  17.887  1.00 75.23           O  \\nATOM    218  CB  LYS A  31      12.247  14.016  18.550  1.00 80.64           C  \\nATOM    219  N   VAL A  32      10.420  11.824  17.360  1.00 56.38           N  \\nATOM    220  CA  VAL A  32       9.562  10.658  17.497  1.00 56.38           C  \\nATOM    221  C   VAL A  32       9.898   9.703  16.364  1.00 56.38           C  \\nATOM    222  O   VAL A  32      10.051   8.496  16.569  1.00 56.38           O  \\nATOM    223  CB  VAL A  32       8.090  11.065  17.376  1.00 46.58           C  \\nATOM    224  CG1 VAL A  32       7.202   9.830  17.366  1.00 46.58           C  \\nATOM    225  CG2 VAL A  32       7.725  12.000  18.532  1.00 46.58           C  \\nATOM    226  N   LEU A  33      10.016  10.262  15.162  1.00 55.92           N  \\nATOM    227  CA  LEU A  33      10.340   9.470  13.992  1.00 55.92           C  \\nATOM    228  C   LEU A  33      11.608   8.685  14.235  1.00 55.92           C  \\nATOM    229  O   LEU A  33      11.579   7.454  14.286  1.00 55.92           O  \\nATOM    230  CB  LEU A  33      10.523  10.363  12.773  1.00 52.98           C  \\nATOM    231  CG  LEU A  33       9.242  10.835  12.080  1.00 52.98           C  \\nATOM    232  CD1 LEU A  33       9.641  11.471  10.750  1.00 52.98           C  \\nATOM    233  CD2 LEU A  33       8.273   9.666  11.846  1.00 52.98           C  \\nATOM    234  N   ASN A  34      12.719   9.404  14.389  1.00 69.48           N  \\nATOM    235  CA  ASN A  34      14.011   8.779  14.631  1.00 69.48           C  \\nATOM    236  C   ASN A  34      13.870   7.727  15.722  1.00 69.48           C  \\nATOM    237  O   ASN A  34      14.350   6.605  15.584  1.00 69.48           O  \\nATOM    238  CB  ASN A  34      15.039   9.844  15.038  1.00 46.71           C  \\nATOM    239  N   GLU A  35      13.180   8.098  16.795  1.00 62.44           N  \\nATOM    240  CA  GLU A  35      12.972   7.212  17.938  1.00 62.44           C  \\nATOM    241  C   GLU A  35      12.233   5.917  17.574  1.00 62.44           C  \\nATOM    242  O   GLU A  35      12.580   4.830  18.057  1.00 62.44           O  \\nATOM    243  CB  GLU A  35      12.207   7.956  19.039  1.00129.81           C  \\nATOM    244  CG  GLU A  35      11.750   7.084  20.206  1.00114.46           C  \\nATOM    245  CD  GLU A  35      11.028   7.870  21.289  1.00114.46           C  \\nATOM    246  OE1 GLU A  35      10.356   7.237  22.131  1.00114.46           O  \\nATOM    247  OE2 GLU A  35      11.141   9.116  21.298  1.00114.46           O  \\nATOM    248  N   LEU A  36      11.213   6.021  16.730  1.00 59.14           N  \\nATOM    249  CA  LEU A  36      10.469   4.828  16.342  1.00 59.14           C  \\nATOM    250  C   LEU A  36      11.321   3.923  15.466  1.00 59.14           C  \\nATOM    251  O   LEU A  36      11.229   2.702  15.544  1.00 59.14           O  \\nATOM    252  CB  LEU A  36       9.217   5.227  15.613  1.00 61.28           C  \\nHETATM  253  N   MSE A  37      12.136   4.544  14.621  1.00 45.85           N  \\nHETATM  254  CA  MSE A  37      13.021   3.824  13.725  1.00 45.85           C  \\nHETATM  255  C   MSE A  37      13.869   2.835  14.511  1.00 45.85           C  \\nHETATM  256  O   MSE A  37      14.082   1.700  14.085  1.00 45.85           O  \\nHETATM  257  CB  MSE A  37      13.922   4.804  12.968  1.00 57.27           C  \\nHETATM  258  CG  MSE A  37      13.203   5.576  11.873  1.00 57.27           C  \\nHETATM  259 SE   MSE A  37      14.347   6.471  10.802  1.00 57.27          SE  \\nHETATM  260  CE  MSE A  37      14.638   5.266   9.472  1.00 57.27           C  \\nATOM    261  N   ASP A  38      14.338   3.255  15.679  1.00 56.30           N  \\nATOM    262  CA  ASP A  38      15.156   2.372  16.497  1.00 56.30           C  \\nATOM    263  C   ASP A  38      14.343   1.250  17.135  1.00 56.30           C  \\nATOM    264  O   ASP A  38      14.695   0.070  17.008  1.00 56.30           O  \\nATOM    265  CB  ASP A  38      15.901   3.179  17.556  1.00 94.79           C  \\nATOM    266  CG  ASP A  38      16.859   4.191  16.942  1.00 94.79           C  \\nATOM    267  OD1 ASP A  38      17.680   3.809  16.079  1.00 94.79           O  \\nATOM    268  OD2 ASP A  38      16.791   5.377  17.317  1.00 94.79           O  \\nATOM    269  N   TYR A  39      13.249   1.597  17.806  1.00 57.17           N  \\nATOM    270  CA  TYR A  39      12.420   0.561  18.424  1.00 57.17           C  \\nATOM    271  C   TYR A  39      12.126  -0.564  17.425  1.00 57.17           C  \\nATOM    272  O   TYR A  39      12.195  -1.745  17.771  1.00104.80           O  \\nATOM    273  CB  TYR A  39      11.086   1.135  18.915  1.00 92.26           C  \\nATOM    274  CG  TYR A  39      11.163   2.049  20.123  1.00 92.26           C  \\nATOM    275  CD1 TYR A  39      11.708   1.607  21.333  1.00 92.26           C  \\nATOM    276  CD2 TYR A  39      10.645   3.346  20.069  1.00 92.26           C  \\nATOM    277  CE1 TYR A  39      11.732   2.439  22.460  1.00 92.26           C  \\nATOM    278  CE2 TYR A  39      10.661   4.179  21.186  1.00 92.26           C  \\nATOM    279  CZ  TYR A  39      11.204   3.724  22.373  1.00 92.26           C  \\nATOM    280  OH  TYR A  39      11.218   4.565  23.462  1.00 92.26           O  \\nATOM    281  N   ILE A  40      11.805  -0.192  16.186  1.00 75.79           N  \\nATOM    282  CA  ILE A  40      11.467  -1.188  15.177  1.00 75.79           C  \\nATOM    283  C   ILE A  40      12.677  -1.933  14.605  1.00 75.79           C  \\nATOM    284  O   ILE A  40      12.570  -3.121  14.295  1.00 75.79           O  \\nATOM    285  CB  ILE A  40      10.603  -0.569  14.031  1.00 76.70           C  \\nATOM    286  CG1 ILE A  40      11.482   0.006  12.931  1.00 76.70           C  \\nATOM    287  CG2 ILE A  40       9.709   0.512  14.582  1.00 76.70           C  \\nATOM    288  CD1 ILE A  40      11.700  -0.961  11.782  1.00 76.70           C  \\nATOM    289  N   LYS A  41      13.817  -1.255  14.454  1.00 65.22           N  \\nATOM    290  CA  LYS A  41      15.015  -1.937  13.960  1.00 65.22           C  \\nATOM    291  C   LYS A  41      15.238  -3.089  14.927  1.00 65.22           C  \\nATOM    292  O   LYS A  41      15.336  -4.253  14.534  1.00 65.22           O  \\nATOM    293  CB  LYS A  41      16.245  -1.024  13.999  1.00 66.42           C  \\nATOM    294  CG  LYS A  41      16.441  -0.159  12.762  1.00 66.42           C  \\nATOM    295  CD  LYS A  41      17.635   0.777  12.923  1.00 66.42           C  \\nATOM    296  CE  LYS A  41      17.801   1.692  11.713  1.00 66.42           C  \\nATOM    297  NZ  LYS A  41      18.856   2.725  11.913  1.00 66.42           N  \\nATOM    298  N   LYS A  42      15.287  -2.736  16.207  1.00 55.06           N  \\nATOM    299  CA  LYS A  42      15.498  -3.693  17.278  1.00 55.06           C  \\nATOM    300  C   LYS A  42      14.502  -4.852  17.267  1.00 55.06           C  \\nATOM    301  O   LYS A  42      14.892  -6.013  17.198  1.00 55.06           O  \\nATOM    302  CB  LYS A  42      15.437  -2.966  18.625  1.00200.03           C  \\nATOM    303  CG  LYS A  42      15.692  -3.852  19.830  1.00186.63           C  \\nATOM    304  CD  LYS A  42      15.652  -3.051  21.117  1.00186.63           C  \\nATOM    305  CE  LYS A  42      15.900  -3.945  22.320  1.00139.05           C  \\nATOM    306  NZ  LYS A  42      17.216  -4.640  22.230  1.00139.05           N  \\nATOM    307  N   GLN A  43      13.214  -4.548  17.331  1.00 78.34           N  \\nATOM    308  CA  GLN A  43      12.211  -5.606  17.359  1.00 78.34           C  \\nATOM    309  C   GLN A  43      12.149  -6.428  16.073  1.00 78.34           C  \\nATOM    310  O   GLN A  43      11.838  -7.621  16.111  1.00 78.34           O  \\nATOM    311  CB  GLN A  43      10.832  -5.013  17.677  1.00126.98           C  \\nATOM    312  CG  GLN A  43      10.169  -5.609  18.919  1.00126.98           C  \\nATOM    313  CD  GLN A  43       9.832  -7.084  18.758  1.00126.98           C  \\nATOM    314  OE1 GLN A  43       8.990  -7.452  17.940  1.00126.98           O  \\nATOM    315  NE2 GLN A  43      10.495  -7.935  19.537  1.00126.98           N  \\nATOM    316  N   GLY A  44      12.452  -5.789  14.942  1.00 55.22           N  \\nATOM    317  CA  GLY A  44      12.412  -6.472  13.661  1.00 55.22           C  \\nATOM    318  C   GLY A  44      11.154  -7.305  13.501  1.00 55.22           C  \\nATOM    319  O   GLY A  44      11.230  -8.524  13.317  1.00 55.22           O  \\nATOM    320  N   ALA A  45       9.991  -6.650  13.567  1.00 48.70           N  \\nATOM    321  CA  ALA A  45       8.706  -7.334  13.446  1.00 48.70           C  \\nATOM    322  C   ALA A  45       8.267  -7.491  12.002  1.00 48.70           C  \\nATOM    323  O   ALA A  45       8.625  -6.691  11.135  1.00 48.70           O  \\nATOM    324  CB  ALA A  45       7.651  -6.587  14.227  1.00 42.35           C  \\nATOM    325  N   LYS A  46       7.482  -8.527  11.748  1.00 60.45           N  \\nATOM    326  CA  LYS A  46       7.002  -8.787  10.405  1.00 60.45           C  \\nATOM    327  C   LYS A  46       6.082  -7.662   9.961  1.00 60.45           C  \\nATOM    328  O   LYS A  46       6.256  -7.102   8.881  1.00 60.45           O  \\nATOM    329  CB  LYS A  46       6.266 -10.122  10.362  1.00100.71           C  \\nATOM    330  N   ARG A  47       5.116  -7.324  10.808  1.00 55.30           N  \\nATOM    331  CA  ARG A  47       4.160  -6.279  10.484  1.00 55.30           C  \\nATOM    332  C   ARG A  47       4.244  -5.060  11.406  1.00 55.30           C  \\nATOM    333  O   ARG A  47       4.294  -5.187  12.625  1.00 55.30           O  \\nATOM    334  CB  ARG A  47       2.747  -6.860  10.497  1.00 49.32           C  \\nATOM    335  N   VAL A  48       4.257  -3.875  10.796  1.00 43.03           N  \\nATOM    336  CA  VAL A  48       4.320  -2.600  11.511  1.00 43.03           C  \\nATOM    337  C   VAL A  48       3.326  -1.588  10.915  1.00 43.03           C  \\nATOM    338  O   VAL A  48       3.297  -1.359   9.715  1.00 43.03           O  \\nATOM    339  CB  VAL A  48       5.742  -2.008  11.457  1.00 31.14           C  \\nATOM    340  CG1 VAL A  48       5.702  -0.529  11.831  1.00 31.14           C  \\nATOM    341  CG2 VAL A  48       6.653  -2.775  12.414  1.00 31.14           C  \\nATOM    342  N   ARG A  49       2.520  -0.975  11.776  1.00 36.65           N  \\nATOM    343  CA  ARG A  49       1.516  -0.009  11.362  1.00 36.65           C  \\nATOM    344  C   ARG A  49       1.740   1.333  12.036  1.00 36.65           C  \\nATOM    345  O   ARG A  49       2.013   1.414  13.231  1.00 36.65           O  \\nATOM    346  CB  ARG A  49       0.124  -0.512  11.726  1.00 83.41           C  \\nATOM    347  CG  ARG A  49      -1.002   0.196  10.999  1.00 73.04           C  \\nATOM    348  CD  ARG A  49      -2.287   0.153  11.807  1.00 73.04           C  \\nATOM    349  NE  ARG A  49      -2.459  -1.126  12.495  1.00 73.04           N  \\nATOM    350  CZ  ARG A  49      -3.229  -1.302  13.569  1.00 73.04           C  \\nATOM    351  NH1 ARG A  49      -3.910  -0.278  14.083  1.00 73.04           N  \\nATOM    352  NH2 ARG A  49      -3.304  -2.499  14.147  1.00 73.04           N  \\nATOM    353  N   ILE A  50       1.630   2.395  11.252  1.00 31.16           N  \\nATOM    354  CA  ILE A  50       1.776   3.725  11.785  1.00 31.16           C  \\nATOM    355  C   ILE A  50       0.705   4.599  11.157  1.00 31.16           C  \\nATOM    356  O   ILE A  50       0.578   4.660   9.933  1.00 31.16           O  \\nATOM    357  CB  ILE A  50       3.147   4.316  11.465  1.00 26.84           C  \\nATOM    358  CG1 ILE A  50       4.236   3.406  12.024  1.00 26.84           C  \\nATOM    359  CG2 ILE A  50       3.254   5.722  12.079  1.00 26.84           C  \\nATOM    360  CD1 ILE A  50       5.611   4.014  11.978  1.00 26.84           C  \\nATOM    361  N   SER A  51      -0.093   5.239  12.002  1.00 37.07           N  \\nATOM    362  CA  SER A  51      -1.128   6.139  11.519  1.00 37.07           C  \\nATOM    363  C   SER A  51      -0.951   7.489  12.205  1.00 37.07           C  \\nATOM    364  O   SER A  51      -0.674   7.573  13.402  1.00 37.07           O  \\nATOM    365  CB  SER A  51      -2.522   5.559  11.757  1.00 38.78           C  \\nATOM    366  OG  SER A  51      -2.672   5.197  13.104  1.00 38.78           O  \\nATOM    367  N   ILE A  52      -1.067   8.538  11.408  1.00 47.52           N  \\nATOM    368  CA  ILE A  52      -0.899   9.893  11.887  1.00 47.52           C  \\nATOM    369  C   ILE A  52      -2.116  10.712  11.531  1.00 47.52           C  \\nATOM    370  O   ILE A  52      -2.486  10.825  10.359  1.00 47.52           O  \\nATOM    371  CB  ILE A  52       0.341  10.536  11.257  1.00 41.87           C  \\nATOM    372  CG1 ILE A  52       1.592   9.886  11.839  1.00 41.87           C  \\nATOM    373  CG2 ILE A  52       0.336  12.039  11.479  1.00 41.87           C  \\nATOM    374  CD1 ILE A  52       2.843  10.297  11.141  1.00 41.87           C  \\nATOM    375  N   THR A  53      -2.753  11.261  12.556  1.00 55.53           N  \\nATOM    376  CA  THR A  53      -3.922  12.081  12.351  1.00 55.53           C  \\nATOM    377  C   THR A  53      -3.415  13.493  12.189  1.00 55.53           C  \\nATOM    378  O   THR A  53      -2.919  14.098  13.139  1.00 55.53           O  \\nATOM    379  CB  THR A  53      -4.866  12.023  13.538  1.00 54.19           C  \\nATOM    380  OG1 THR A  53      -5.409  10.702  13.658  1.00 54.19           O  \\nATOM    381  CG2 THR A  53      -5.997  13.003  13.334  1.00 54.19           C  \\nATOM    382  N   ALA A  54      -3.523  13.994  10.965  1.00 53.46           N  \\nATOM    383  CA  ALA A  54      -3.079  15.335  10.624  1.00 53.46           C  \\nATOM    384  C   ALA A  54      -4.249  16.316  10.700  1.00 53.46           C  \\nATOM    385  O   ALA A  54      -5.389  15.916  10.939  1.00 53.46           O  \\nATOM    386  CB  ALA A  54      -2.488  15.331   9.219  1.00 54.30           C  \\nATOM    387  N   ARG A  55      -3.960  17.596  10.490  1.00 74.08           N  \\nATOM    388  CA  ARG A  55      -4.983  18.636  10.528  1.00 74.08           C  \\nATOM    389  C   ARG A  55      -5.731  18.711   9.197  1.00 74.08           C  \\nATOM    390  O   ARG A  55      -6.929  18.979   9.170  1.00 74.08           O  \\nATOM    391  CB  ARG A  55      -4.344  19.985  10.857  1.00 53.00           C  \\nATOM    392  N   THR A  56      -5.017  18.477   8.098  1.00 61.37           N  \\nATOM    393  CA  THR A  56      -5.615  18.505   6.762  1.00 61.37           C  \\nATOM    394  C   THR A  56      -5.102  17.346   5.926  1.00 61.37           C  \\nATOM    395  O   THR A  56      -3.968  16.892   6.107  1.00 61.37           O  \\nATOM    396  CB  THR A  56      -5.281  19.807   6.003  1.00 61.95           C  \\nATOM    397  OG1 THR A  56      -3.861  19.944   5.857  1.00 61.95           O  \\nATOM    398  CG2 THR A  56      -5.818  20.995   6.750  1.00 61.95           C  \\nATOM    399  N   LYS A  57      -5.929  16.874   4.998  1.00 79.80           N  \\nATOM    400  CA  LYS A  57      -5.522  15.766   4.148  1.00 79.80           C  \\nATOM    401  C   LYS A  57      -4.219  16.045   3.395  1.00 79.80           C  \\nATOM    402  O   LYS A  57      -3.422  15.135   3.192  1.00 79.80           O  \\nATOM    403  CB  LYS A  57      -6.635  15.394   3.164  1.00 56.21           C  \\nATOM    404  CG  LYS A  57      -6.161  14.470   2.051  1.00 52.05           C  \\nATOM    405  CD  LYS A  57      -7.236  13.507   1.571  1.00 52.05           C  \\nATOM    406  CE  LYS A  57      -6.718  12.676   0.392  1.00 52.05           C  \\nATOM    407  NZ  LYS A  57      -7.658  11.597  -0.067  1.00 52.05           N  \\nATOM    408  N   LYS A  58      -3.990  17.291   2.986  1.00 61.25           N  \\nATOM    409  CA  LYS A  58      -2.758  17.624   2.269  1.00 61.25           C  \\nATOM    410  C   LYS A  58      -1.566  17.464   3.204  1.00 61.25           C  \\nATOM    411  O   LYS A  58      -0.433  17.274   2.764  1.00 61.25           O  \\nATOM    412  CB  LYS A  58      -2.802  19.055   1.719  1.00120.41           C  \\nATOM    413  CG  LYS A  58      -3.869  19.283   0.655  1.00 90.99           C  \\nATOM    414  CD  LYS A  58      -3.473  20.388  -0.316  1.00 90.99           C  \\nATOM    415  CE  LYS A  58      -2.310  19.959  -1.211  1.00 90.99           C  \\nATOM    416  NZ  LYS A  58      -2.662  18.828  -2.130  1.00 90.99           N  \\nATOM    417  N   GLU A  59      -1.831  17.546   4.500  1.00 55.60           N  \\nATOM    418  CA  GLU A  59      -0.786  17.378   5.496  1.00 55.60           C  \\nATOM    419  C   GLU A  59      -0.515  15.871   5.628  1.00 55.60           C  \\nATOM    420  O   GLU A  59       0.633  15.424   5.622  1.00 55.60           O  \\nATOM    421  CB  GLU A  59      -1.254  17.956   6.835  1.00 68.19           C  \\nATOM    422  CG  GLU A  59      -0.299  17.742   7.985  1.00 68.19           C  \\nATOM    423  CD  GLU A  59      -0.721  18.494   9.232  1.00 68.19           C  \\nATOM    424  OE1 GLU A  59      -1.867  18.296   9.694  1.00 68.19           O  \\nATOM    425  OE2 GLU A  59       0.098  19.287   9.750  1.00 68.19           O  \\nATOM    426  N   ALA A  60      -1.595  15.104   5.738  1.00 49.39           N  \\nATOM    427  CA  ALA A  60      -1.528  13.653   5.853  1.00 49.39           C  \\nATOM    428  C   ALA A  60      -0.674  13.071   4.736  1.00 49.39           C  \\nATOM    429  O   ALA A  60      -0.023  12.038   4.910  1.00 49.39           O  \\nATOM    430  CB  ALA A  60      -2.920  13.069   5.769  1.00 34.98           C  \\nATOM    431  N   GLU A  61      -0.698  13.742   3.588  1.00 62.75           N  \\nATOM    432  CA  GLU A  61       0.058  13.318   2.423  1.00 62.75           C  \\nATOM    433  C   GLU A  61       1.541  13.547   2.665  1.00 62.75           C  \\nATOM    434  O   GLU A  61       2.357  12.657   2.427  1.00 62.75           O  \\nATOM    435  CB  GLU A  61      -0.412  14.092   1.189  1.00 78.01           C  \\nATOM    436  CG  GLU A  61      -1.853  13.778   0.793  1.00 51.13           C  \\nATOM    437  CD  GLU A  61      -2.387  14.677  -0.316  1.00 51.13           C  \\nATOM    438  OE1 GLU A  61      -1.632  15.550  -0.805  1.00 51.13           O  \\nATOM    439  OE2 GLU A  61      -3.567  14.506  -0.695  1.00 51.13           O  \\nATOM    440  N   LYS A  62       1.894  14.733   3.151  1.00 51.31           N  \\nATOM    441  CA  LYS A  62       3.292  15.029   3.420  1.00 51.31           C  \\nATOM    442  C   LYS A  62       3.824  13.998   4.406  1.00 51.31           C  \\nATOM    443  O   LYS A  62       4.972  13.578   4.315  1.00 51.31           O  \\nATOM    444  CB  LYS A  62       3.446  16.445   3.996  1.00 31.92           C  \\nATOM    445  N   PHE A  63       2.977  13.589   5.346  1.00 47.90           N  \\nATOM    446  CA  PHE A  63       3.366  12.610   6.358  1.00 47.90           C  \\nATOM    447  C   PHE A  63       3.498  11.225   5.770  1.00 47.90           C  \\nATOM    448  O   PHE A  63       4.373  10.463   6.164  1.00 47.90           O  \\nATOM    449  CB  PHE A  63       2.343  12.577   7.500  1.00 46.25           C  \\nATOM    450  CG  PHE A  63       2.434  13.748   8.420  1.00 46.25           C  \\nATOM    451  CD1 PHE A  63       1.319  14.174   9.130  1.00 46.25           C  \\nATOM    452  CD2 PHE A  63       3.638  14.420   8.591  1.00 46.25           C  \\nATOM    453  CE1 PHE A  63       1.406  15.259  10.007  1.00 46.25           C  \\nATOM    454  CE2 PHE A  63       3.740  15.496   9.455  1.00 46.25           C  \\nATOM    455  CZ  PHE A  63       2.628  15.922  10.166  1.00 46.25           C  \\nATOM    456  N   ALA A  64       2.610  10.899   4.839  1.00 43.04           N  \\nATOM    457  CA  ALA A  64       2.651   9.602   4.194  1.00 43.04           C  \\nATOM    458  C   ALA A  64       3.960   9.509   3.414  1.00 43.04           C  \\nATOM    459  O   ALA A  64       4.578   8.457   3.367  1.00 43.04           O  \\nATOM    460  CB  ALA A  64       1.441   9.419   3.260  1.00 29.61           C  \\nATOM    461  N   ALA A  65       4.388  10.612   2.815  1.00 46.89           N  \\nATOM    462  CA  ALA A  65       5.642  10.619   2.076  1.00 46.89           C  \\nATOM    463  C   ALA A  65       6.791  10.351   3.055  1.00 46.89           C  \\nATOM    464  O   ALA A  65       7.737   9.629   2.737  1.00 46.89           O  \\nATOM    465  CB  ALA A  65       5.839  11.968   1.367  1.00 31.68           C  \\nATOM    466  N   ILE A  66       6.703  10.929   4.248  1.00 45.51           N  \\nATOM    467  CA  ILE A  66       7.722  10.737   5.275  1.00 45.51           C  \\nATOM    468  C   ILE A  66       7.762   9.285   5.741  1.00 45.51           C  \\nATOM    469  O   ILE A  66       8.816   8.652   5.778  1.00 45.51           O  \\nATOM    470  CB  ILE A  66       7.444  11.631   6.508  1.00 53.42           C  \\nATOM    471  CG1 ILE A  66       7.884  13.069   6.215  1.00 53.42           C  \\nATOM    472  CG2 ILE A  66       8.153  11.075   7.746  1.00 53.42           C  \\nATOM    473  CD1 ILE A  66       7.671  14.034   7.377  1.00 53.42           C  \\nATOM    474  N   LEU A  67       6.599   8.766   6.103  1.00 37.91           N  \\nATOM    475  CA  LEU A  67       6.501   7.412   6.591  1.00 37.91           C  \\nATOM    476  C   LEU A  67       6.912   6.387   5.551  1.00 37.91           C  \\nATOM    477  O   LEU A  67       7.378   5.299   5.887  1.00 37.91           O  \\nATOM    478  CB  LEU A  67       5.079   7.144   7.071  1.00 24.44           C  \\nATOM    479  CG  LEU A  67       4.659   8.016   8.260  1.00 24.44           C  \\nATOM    480  CD1 LEU A  67       3.226   7.626   8.711  1.00 24.44           C  \\nATOM    481  CD2 LEU A  67       5.658   7.854   9.394  1.00 24.44           C  \\nATOM    482  N   ILE A  68       6.748   6.723   4.282  1.00 32.01           N  \\nATOM    483  CA  ILE A  68       7.119   5.776   3.255  1.00 32.01           C  \\nATOM    484  C   ILE A  68       8.615   5.568   3.244  1.00 32.01           C  \\nATOM    485  O   ILE A  68       9.074   4.433   3.201  1.00 32.01           O  \\nATOM    486  CB  ILE A  68       6.622   6.217   1.869  1.00 44.65           C  \\nATOM    487  CG1 ILE A  68       5.223   5.656   1.656  1.00 44.65           C  \\nATOM    488  CG2 ILE A  68       7.533   5.705   0.786  1.00 44.65           C  \\nATOM    489  CD1 ILE A  68       4.580   6.096   0.373  1.00 44.65           C  \\nATOM    490  N   LYS A  69       9.366   6.661   3.310  1.00 46.37           N  \\nATOM    491  CA  LYS A  69      10.819   6.592   3.311  1.00 46.37           C  \\nATOM    492  C   LYS A  69      11.296   5.881   4.574  1.00 46.37           C  \\nATOM    493  O   LYS A  69      12.268   5.125   4.533  1.00 46.37           O  \\nATOM    494  CB  LYS A  69      11.401   8.001   3.201  1.00 60.80           C  \\nATOM    495  CG  LYS A  69      10.814   8.757   2.008  1.00 54.72           C  \\nATOM    496  CD  LYS A  69      11.461  10.116   1.760  1.00 54.72           C  \\nATOM    497  CE  LYS A  69      10.804  10.803   0.552  1.00 54.72           C  \\nATOM    498  NZ  LYS A  69      11.330  12.176   0.230  1.00 54.72           N  \\nATOM    499  N   VAL A  70      10.601   6.102   5.689  1.00 46.67           N  \\nATOM    500  CA  VAL A  70      10.959   5.442   6.942  1.00 46.67           C  \\nATOM    501  C   VAL A  70      10.872   3.920   6.791  1.00 46.67           C  \\nATOM    502  O   VAL A  70      11.842   3.216   7.044  1.00 46.67           O  \\nATOM    503  CB  VAL A  70      10.034   5.890   8.104  1.00 45.21           C  \\nATOM    504  CG1 VAL A  70      10.161   4.929   9.292  1.00 45.21           C  \\nATOM    505  CG2 VAL A  70      10.405   7.301   8.539  1.00 45.21           C  \\nATOM    506  N   PHE A  71       9.705   3.425   6.385  1.00 47.43           N  \\nATOM    507  CA  PHE A  71       9.472   1.995   6.184  1.00 47.43           C  \\nATOM    508  C   PHE A  71      10.400   1.397   5.132  1.00 47.43           C  \\nATOM    509  O   PHE A  71      10.869   0.265   5.276  1.00 47.43           O  \\nATOM    510  CB  PHE A  71       8.043   1.749   5.717  1.00 37.02           C  \\nATOM    511  CG  PHE A  71       7.046   1.645   6.819  1.00 37.02           C  \\nATOM    512  CD1 PHE A  71       6.882   0.462   7.510  1.00 37.02           C  \\nATOM    513  CD2 PHE A  71       6.236   2.722   7.146  1.00 37.02           C  \\nATOM    514  CE1 PHE A  71       5.912   0.350   8.513  1.00 37.02           C  \\nATOM    515  CE2 PHE A  71       5.276   2.619   8.138  1.00 37.02           C  \\nATOM    516  CZ  PHE A  71       5.110   1.440   8.820  1.00 37.02           C  \\nATOM    517  N   ALA A  72      10.631   2.143   4.058  1.00 53.65           N  \\nATOM    518  CA  ALA A  72      11.491   1.668   2.986  1.00 53.65           C  \\nATOM    519  C   ALA A  72      12.911   1.486   3.508  1.00 53.65           C  \\nATOM    520  O   ALA A  72      13.481   0.397   3.410  1.00 53.65           O  \\nATOM    521  CB  ALA A  72      11.476   2.654   1.832  1.00 41.89           C  \\nATOM    522  N   GLU A  73      13.474   2.552   4.070  1.00 50.89           N  \\nATOM    523  CA  GLU A  73      14.818   2.483   4.606  1.00 50.89           C  \\nATOM    524  C   GLU A  73      14.914   1.282   5.565  1.00 50.89           C  \\nATOM    525  O   GLU A  73      15.805   0.457   5.434  1.00 50.89           O  \\nATOM    526  CB  GLU A  73      15.186   3.805   5.314  1.00 20.48           C  \\nATOM    527  N   LEU A  74      13.978   1.145   6.496  1.00 43.89           N  \\nATOM    528  CA  LEU A  74      14.025   0.022   7.431  1.00 43.89           C  \\nATOM    529  C   LEU A  74      13.690  -1.342   6.817  1.00 43.89           C  \\nATOM    530  O   LEU A  74      13.367  -2.282   7.535  1.00 43.89           O  \\nATOM    531  CB  LEU A  74      13.100   0.287   8.621  1.00 51.78           C  \\nATOM    532  CG  LEU A  74      13.297   1.686   9.215  1.00 51.78           C  \\nATOM    533  CD1 LEU A  74      12.449   1.868  10.462  1.00 51.78           C  \\nATOM    534  CD2 LEU A  74      14.769   1.889   9.525  1.00 51.78           C  \\nATOM    535  N   GLY A  75      13.731  -1.453   5.492  1.00 38.15           N  \\nATOM    536  CA  GLY A  75      13.465  -2.739   4.875  1.00 38.15           C  \\nATOM    537  C   GLY A  75      12.039  -3.207   4.697  1.00 38.15           C  \\nATOM    538  O   GLY A  75      11.810  -4.287   4.152  1.00 38.15           O  \\nATOM    539  N   TYR A  76      11.063  -2.447   5.168  1.00 50.24           N  \\nATOM    540  CA  TYR A  76       9.685  -2.857   4.949  1.00 50.24           C  \\nATOM    541  C   TYR A  76       9.433  -2.436   3.506  1.00 50.24           C  \\nATOM    542  O   TYR A  76       9.581  -1.257   3.162  1.00 50.24           O  \\nATOM    543  CB  TYR A  76       8.781  -2.137   5.932  1.00 36.14           C  \\nATOM    544  CG  TYR A  76       9.004  -2.635   7.338  1.00 36.14           C  \\nATOM    545  CD1 TYR A  76       8.162  -3.598   7.898  1.00 36.14           C  \\nATOM    546  CD2 TYR A  76      10.128  -2.235   8.070  1.00 36.14           C  \\nATOM    547  CE1 TYR A  76       8.438  -4.166   9.150  1.00 36.14           C  \\nATOM    548  CE2 TYR A  76      10.413  -2.802   9.322  1.00 36.14           C  \\nATOM    549  CZ  TYR A  76       9.567  -3.770   9.848  1.00 36.14           C  \\nATOM    550  OH  TYR A  76       9.881  -4.392  11.036  1.00 36.14           O  \\nATOM    551  N   ASN A  77       9.073  -3.396   2.657  1.00 50.62           N  \\nATOM    552  CA  ASN A  77       8.890  -3.090   1.246  1.00 50.62           C  \\nATOM    553  C   ASN A  77       7.535  -3.320   0.577  1.00 50.62           C  \\nATOM    554  O   ASN A  77       7.395  -3.014  -0.605  1.00 50.62           O  \\nATOM    555  CB  ASN A  77       9.985  -3.784   0.428  1.00 96.49           C  \\nATOM    556  CG  ASN A  77      11.391  -3.387   0.875  1.00 96.49           C  \\nATOM    557  OD1 ASN A  77      11.710  -2.200   0.989  1.00 96.49           O  \\nATOM    558  ND2 ASN A  77      12.240  -4.384   1.122  1.00 96.49           N  \\nATOM    559  N   ASP A  78       6.556  -3.874   1.285  1.00 72.36           N  \\nATOM    560  CA  ASP A  78       5.219  -4.028   0.700  1.00 72.36           C  \\nATOM    561  C   ASP A  78       4.327  -3.171   1.600  1.00 72.36           C  \\nATOM    562  O   ASP A  78       3.796  -3.643   2.609  1.00 72.36           O  \\nATOM    563  CB  ASP A  78       4.745  -5.479   0.716  1.00108.13           C  \\nATOM    564  CG  ASP A  78       3.377  -5.644   0.065  1.00108.13           C  \\nATOM    565  OD1 ASP A  78       3.225  -5.258  -1.112  1.00108.13           O  \\nATOM    566  OD2 ASP A  78       2.448  -6.153   0.728  1.00108.13           O  \\nATOM    567  N   ILE A  79       4.163  -1.910   1.214  1.00 35.39           N  \\nATOM    568  CA  ILE A  79       3.425  -0.943   2.015  1.00 35.39           C  \\nATOM    569  C   ILE A  79       2.025  -0.548   1.563  1.00 35.39           C  \\nATOM    570  O   ILE A  79       1.779  -0.336   0.382  1.00 35.39           O  \\nATOM    571  CB  ILE A  79       4.260   0.326   2.122  1.00 28.47           C  \\nATOM    572  CG1 ILE A  79       5.672  -0.058   2.552  1.00 28.47           C  \\nATOM    573  CG2 ILE A  79       3.630   1.303   3.070  1.00 28.47           C  \\nATOM    574  CD1 ILE A  79       6.681   1.064   2.358  1.00 28.47           C  \\nATOM    575  N   ASN A  80       1.120  -0.449   2.537  1.00 32.25           N  \\nATOM    576  CA  ASN A  80      -0.255  -0.026   2.299  1.00 32.25           C  \\nATOM    577  C   ASN A  80      -0.386   1.364   2.884  1.00 32.25           C  \\nATOM    578  O   ASN A  80      -0.009   1.599   4.028  1.00 32.25           O  \\nATOM    579  CB  ASN A  80      -1.256  -0.953   2.983  1.00 68.24           C  \\nATOM    580  CG  ASN A  80      -1.409  -2.275   2.265  1.00 68.24           C  \\nATOM    581  OD1 ASN A  80      -1.410  -3.341   2.885  1.00 68.24           O  \\nATOM    582  ND2 ASN A  80      -1.552  -2.215   0.949  1.00 68.24           N  \\nATOM    583  N   VAL A  81      -0.875   2.291   2.074  1.00 31.42           N  \\nATOM    584  CA  VAL A  81      -1.090   3.651   2.511  1.00 31.42           C  \\nATOM    585  C   VAL A  81      -2.586   3.842   2.393  1.00 31.42           C  \\nATOM    586  O   VAL A  81      -3.154   3.663   1.319  1.00 31.42           O  \\nATOM    587  CB  VAL A  81      -0.418   4.651   1.595  1.00 27.86           C  \\nATOM    588  CG1 VAL A  81      -0.828   6.027   1.980  1.00 27.86           C  \\nATOM    589  CG2 VAL A  81       1.089   4.522   1.694  1.00 27.86           C  \\nATOM    590  N   THR A  82      -3.226   4.180   3.499  1.00 39.50           N  \\nATOM    591  CA  THR A  82      -4.666   4.392   3.515  1.00 39.50           C  \\nATOM    592  C   THR A  82      -5.000   5.729   4.145  1.00 39.50           C  \\nATOM    593  O   THR A  82      -4.520   6.055   5.233  1.00 39.50           O  \\nATOM    594  CB  THR A  82      -5.398   3.315   4.339  1.00 46.66           C  \\nATOM    595  OG1 THR A  82      -5.223   2.023   3.741  1.00 46.66           O  \\nATOM    596  CG2 THR A  82      -6.867   3.630   4.400  1.00 46.66           C  \\nATOM    597  N   PHE A  83      -5.802   6.522   3.457  1.00 57.37           N  \\nATOM    598  CA  PHE A  83      -6.197   7.793   4.034  1.00 57.37           C  \\nATOM    599  C   PHE A  83      -7.634   7.636   4.481  1.00 57.37           C  \\nATOM    600  O   PHE A  83      -8.448   7.043   3.776  1.00 57.37           O  \\nATOM    601  CB  PHE A  83      -6.097   8.935   3.025  1.00 35.09           C  \\nATOM    602  CG  PHE A  83      -4.709   9.189   2.529  1.00 35.09           C  \\nATOM    603  CD1 PHE A  83      -4.182   8.445   1.477  1.00 35.09           C  \\nATOM    604  CD2 PHE A  83      -3.930  10.179   3.103  1.00 35.09           C  \\nATOM    605  CE1 PHE A  83      -2.893   8.688   0.998  1.00 35.09           C  \\nATOM    606  CE2 PHE A  83      -2.643  10.433   2.638  1.00 35.09           C  \\nATOM    607  CZ  PHE A  83      -2.123   9.681   1.578  1.00 35.09           C  \\nATOM    608  N   ASP A  84      -7.924   8.130   5.676  1.00 57.44           N  \\nATOM    609  CA  ASP A  84      -9.270   8.081   6.224  1.00 57.44           C  \\nATOM    610  C   ASP A  84      -9.498   9.482   6.767  1.00 57.44           C  \\nATOM    611  O   ASP A  84      -9.176   9.781   7.915  1.00 57.44           O  \\nATOM    612  CB  ASP A  84      -9.365   7.052   7.349  1.00115.57           C  \\nATOM    613  CG  ASP A  84     -10.794   6.628   7.625  1.00115.57           C  \\nATOM    614  OD1 ASP A  84     -11.645   7.511   7.866  1.00115.57           O  \\nATOM    615  OD2 ASP A  84     -11.068   5.410   7.597  1.00115.57           O  \\nATOM    616  N   GLY A  85     -10.026  10.352   5.924  1.00 51.12           N  \\nATOM    617  CA  GLY A  85     -10.250  11.713   6.358  1.00 51.12           C  \\nATOM    618  C   GLY A  85      -8.911  12.403   6.503  1.00 51.12           C  \\nATOM    619  O   GLY A  85      -8.212  12.608   5.510  1.00 51.12           O  \\nATOM    620  N   ASP A  86      -8.545  12.754   7.733  1.00 53.00           N  \\nATOM    621  CA  ASP A  86      -7.275  13.427   7.979  1.00 53.00           C  \\nATOM    622  C   ASP A  86      -6.266  12.493   8.617  1.00 53.00           C  \\nATOM    623  O   ASP A  86      -5.171  12.907   9.004  1.00 53.00           O  \\nATOM    624  CB  ASP A  86      -7.497  14.652   8.856  1.00 84.41           C  \\nATOM    625  CG  ASP A  86      -7.893  15.864   8.047  1.00 84.41           C  \\nATOM    626  OD1 ASP A  86      -8.499  16.790   8.622  1.00 84.41           O  \\nATOM    627  OD2 ASP A  86      -7.589  15.892   6.833  1.00 84.41           O  \\nATOM    628  N   THR A  87      -6.652  11.227   8.713  1.00 54.91           N  \\nATOM    629  CA  THR A  87      -5.799  10.202   9.280  1.00 54.91           C  \\nATOM    630  C   THR A  87      -5.142   9.365   8.190  1.00 54.91           C  \\nATOM    631  O   THR A  87      -5.826   8.797   7.346  1.00 54.91           O  \\nATOM    632  CB  THR A  87      -6.600   9.245  10.167  1.00 59.15           C  \\nATOM    633  OG1 THR A  87      -7.103   9.953  11.304  1.00 59.15           O  \\nATOM    634  CG2 THR A  87      -5.728   8.095  10.628  1.00 59.15           C  \\nATOM    635  N   VAL A  88      -3.816   9.303   8.191  1.00 36.77           N  \\nATOM    636  CA  VAL A  88      -3.125   8.459   7.218  1.00 36.77           C  \\nATOM    637  C   VAL A  88      -2.580   7.225   7.935  1.00 36.77           C  \\nATOM    638  O   VAL A  88      -1.997   7.316   9.019  1.00 36.77           O  \\nATOM    639  CB  VAL A  88      -1.953   9.176   6.533  1.00 32.82           C  \\nATOM    640  CG1 VAL A  88      -0.849   9.457   7.518  1.00 32.82           C  \\nATOM    641  CG2 VAL A  88      -1.459   8.320   5.381  1.00 32.82           C  \\nATOM    642  N   THR A  89      -2.790   6.067   7.330  1.00 49.29           N  \\nATOM    643  CA  THR A  89      -2.314   4.826   7.917  1.00 49.29           C  \\nATOM    644  C   THR A  89      -1.395   4.069   6.971  1.00 49.29           C  \\nATOM    645  O   THR A  89      -1.844   3.485   5.978  1.00 49.29           O  \\nATOM    646  CB  THR A  89      -3.476   3.934   8.283  1.00 51.33           C  \\nATOM    647  OG1 THR A  89      -4.392   4.694   9.073  1.00 51.33           O  \\nATOM    648  CG2 THR A  89      -2.999   2.722   9.060  1.00 51.33           C  \\nATOM    649  N   VAL A  90      -0.101   4.112   7.284  1.00 33.06           N  \\nATOM    650  CA  VAL A  90       0.909   3.414   6.508  1.00 33.06           C  \\nATOM    651  C   VAL A  90       1.260   2.131   7.257  1.00 33.06           C  \\nATOM    652  O   VAL A  90       1.736   2.164   8.392  1.00 33.06           O  \\nATOM    653  CB  VAL A  90       2.156   4.280   6.343  1.00 31.66           C  \\nATOM    654  CG1 VAL A  90       3.173   3.579   5.443  1.00 31.66           C  \\nATOM    655  CG2 VAL A  90       1.762   5.626   5.741  1.00 31.66           C  \\nATOM    656  N   GLU A  91       0.969   0.996   6.638  1.00 39.68           N  \\nATOM    657  CA  GLU A  91       1.283  -0.296   7.233  1.00 39.68           C  \\nATOM    658  C   GLU A  91       2.266  -1.019   6.321  1.00 39.68           C  \\nATOM    659  O   GLU A  91       2.032  -1.166   5.114  1.00 39.68           O  \\nATOM    660  CB  GLU A  91       0.045  -1.172   7.391  1.00115.49           C  \\nATOM    661  CG  GLU A  91       0.407  -2.556   7.905  1.00111.17           C  \\nATOM    662  CD  GLU A  91      -0.442  -3.656   7.308  1.00111.17           C  \\nATOM    663  OE1 GLU A  91      -1.618  -3.782   7.708  1.00111.17           O  \\nATOM    664  OE2 GLU A  91       0.066  -4.394   6.432  1.00111.17           O  \\nATOM    665  N   GLY A  92       3.369  -1.469   6.897  1.00 57.29           N  \\nATOM    666  CA  GLY A  92       4.345  -2.181   6.106  1.00 57.29           C  \\nATOM    667  C   GLY A  92       4.665  -3.567   6.629  1.00 57.29           C  \\nATOM    668  O   GLY A  92       4.434  -3.886   7.799  1.00 57.29           O  \\nATOM    669  N   GLN A  93       5.184  -4.402   5.733  1.00 51.61           N  \\nATOM    670  CA  GLN A  93       5.609  -5.756   6.069  1.00 51.61           C  \\nATOM    671  C   GLN A  93       7.071  -5.898   5.659  1.00 51.61           C  \\nATOM    672  O   GLN A  93       7.498  -5.373   4.628  1.00 51.61           O  \\nATOM    673  CB  GLN A  93       4.724  -6.792   5.376  1.00 87.10           C  \\nATOM    674  CG  GLN A  93       3.505  -7.152   6.212  1.00 87.10           C  \\nATOM    675  CD  GLN A  93       2.449  -7.895   5.433  1.00 87.10           C  \\nATOM    676  OE1 GLN A  93       1.866  -7.357   4.489  1.00 87.10           O  \\nATOM    677  NE2 GLN A  93       2.189  -9.139   5.823  1.00 87.10           N  \\nATOM    678  N   LEU A  94       7.836  -6.589   6.494  1.00 78.23           N  \\nATOM    679  CA  LEU A  94       9.264  -6.783   6.273  1.00 78.23           C  \\nATOM    680  C   LEU A  94       9.604  -7.603   5.021  1.00 78.23           C  \\nATOM    681  O   LEU A  94      10.488  -7.171   4.243  1.00 78.23           O  \\nATOM    682  CB  LEU A  94       9.865  -7.422   7.524  1.00 64.38           C  \\nATOM    683  CG  LEU A  94      11.295  -7.028   7.888  1.00 64.38           C  \\nATOM    684  CD1 LEU A  94      11.525  -5.554   7.645  1.00 64.38           C  \\nATOM    685  CD2 LEU A  94      11.541  -7.378   9.338  1.00 64.38           C  \\nTER     686      LEU A  94                                                      \\nHETATM  687  O   HOH A 107      10.102  -4.251  13.943  1.00 44.75           O  \\nHETATM  688  O   HOH A 108       1.041  -4.800   4.090  1.00 52.54           O  \\nHETATM  689  O   HOH A 109      -0.433   3.575  23.826  1.00 38.09           O  \\nHETATM  690  O   HOH A 110       2.878  19.537   9.279  1.00 61.82           O  \\nHETATM  691  O   HOH A 111      18.925   5.360  10.377  1.00 60.69           O  \\nHETATM  692  O   HOH A 112      -3.917   2.311  13.588  1.00 49.83           O  \\nHETATM  693  O   HOH A 113       4.748  12.641  22.041  1.00 58.74           O  \\nCONECT  250  253                                                                \\nCONECT  253  250  254                                                           \\nCONECT  254  253  255  257                                                      \\nCONECT  255  254  256  261                                                      \\nCONECT  256  255                                                                \\nCONECT  257  254  258                                                           \\nCONECT  258  257  259                                                           \\nCONECT  259  258  260                                                           \\nCONECT  260  259                                                                \\nCONECT  261  255                                                                \\nMASTER      303    0    1    2    5    0    0    6  692    1   10    9          \\nEND                                                                             \\n\");\n",
       "\tviewer_16908637293794932.setStyle({\"model\": -1},{\"cartoon\": {\"colorscheme\": {\"prop\": \"b\", \"map\": {\"0\": \"#FF7D45\", \"1\": \"#FFDB13\", \"2\": \"#65CBF3\", \"3\": \"#0053D6\"}}}, \"stick\": {}});\n",
       "\tviewer_16908637293794932.zoomTo();\n",
       "viewer_16908637293794932.render();\n",
       "});\n",
       "</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<py3Dmol.view at 0x7f5c5d3dc7c0>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PLDDT_BANDS = [\n",
    "  (0, 50, '#FF7D45'),\n",
    "  (50, 70, '#FFDB13'),\n",
    "  (70, 90, '#65CBF3'),\n",
    "  (90, 100, '#0053D6')\n",
    "]\n",
    "view = py3Dmol.view(width=800, height=600)\n",
    "view.addModelsAsFrames(init_pdb)\n",
    "\n",
    "color_map = {i: bands[2] for i, bands in enumerate(PLDDT_BANDS)}\n",
    "style = {'cartoon': {'colorscheme': {'prop': 'b', 'map': color_map}}}\n",
    "\n",
    "style['stick'] = {}\n",
    "\n",
    "view.setStyle({'model': -1}, style)\n",
    "view.zoomTo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "80441f06-5a39-41aa-9559-94eae97ac5ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"./samples/design_9.pdb\", mode=\"r\") as f:\n",
    "    designed_pdb = f.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "61cadcfa-4679-4894-932c-b7b771f576e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_16908637703553598\"  style=\"position: relative; width: 800px; height: 600px;\">\n        <p id=\"3dmolwarning_16908637703553598\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    //this is to ignore the existence of requirejs amd\n    var savedexports, savedmodule;\n    if (typeof exports !== 'undefined') savedexports = exports;\n    else exports = {}\n    if (typeof module !== 'undefined') savedmodule = module;\n    else module = {}\n\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n        exports = savedexports;\n        module = savedmodule;\n        resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://cdnjs.cloudflare.com/ajax/libs/3Dmol/2.0.3/3Dmol-min.js');\n}\n\nvar viewer_16908637703553598 = null;\nvar warn = document.getElementById(\"3dmolwarning_16908637703553598\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16908637703553598 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_16908637703553598\"),{backgroundColor:\"white\"});\nviewer_16908637703553598.zoomTo();\n\tviewer_16908637703553598.addModelsAsFrames(\"ATOM      1  N   GLY A   1      11.309   0.598   1.723  1.00  0.00\\nATOM      2  CA  GLY A   1      10.408   1.213   2.690  1.00  0.00\\nATOM      3  C   GLY A   1       8.966   0.783   2.451  1.00  0.00\\nATOM      4  O   GLY A   1       8.596   0.411   1.338  1.00  0.00\\nATOM      5  N   GLY A   2       8.077   1.637   2.685  1.00  0.00\\nATOM      6  CA  GLY A   2       6.786   0.962   2.631  1.00  0.00\\nATOM      7  C   GLY A   2       6.405   0.614   1.198  1.00  0.00\\nATOM      8  O   GLY A   2       6.515   1.446   0.297  1.00  0.00\\nATOM      9  N   GLY A   3       6.154  -0.391   1.211  1.00  0.00\\nATOM     10  CA  GLY A   3       5.779  -0.911  -0.098  1.00  0.00\\nATOM     11  C   GLY A   3       4.285  -0.751  -0.347  1.00  0.00\\nATOM     12  O   GLY A   3       3.462  -1.307   0.380  1.00  0.00\\nATOM     13  N   GLY A   4       4.043   0.138  -0.532  1.00  0.00\\nATOM     14  CA  GLY A   4       2.637   0.395  -0.821  1.00  0.00\\nATOM     15  C   GLY A   4       2.365   0.349  -2.319  1.00  0.00\\nATOM     16  O   GLY A   4       1.213   0.287  -2.749  1.00  0.00\\nATOM     17  N   GLY A   5       3.104   0.160  -3.048  1.00  0.00\\nATOM     18  CA  GLY A   5       2.936   0.231  -4.495  1.00  0.00\\nATOM     19  C   GLY A   5       1.816  -0.689  -4.965  1.00  0.00\\nATOM     20  O   GLY A   5       0.993  -0.305  -5.796  1.00  0.00\\nATOM     21  N   GLY A   6       1.920  -1.767  -4.681  1.00  0.00\\nATOM     22  CA  GLY A   6       0.838  -2.646  -5.108  1.00  0.00\\nATOM     23  C   GLY A   6      -0.428  -2.393  -4.299  1.00  0.00\\nATOM     24  O   GLY A   6      -1.539  -2.623  -4.779  1.00  0.00\\nATOM     25  N   GLY A   7      -0.050  -2.160  -3.316  1.00  0.00\\nATOM     26  CA  GLY A   7      -1.269  -1.895  -2.561  1.00  0.00\\nATOM     27  C   GLY A   7      -2.026  -0.705  -3.137  1.00  0.00\\nATOM     28  O   GLY A   7      -3.256  -0.665  -3.105  1.00  0.00\\nATOM     29  N   GLY A   8      -1.325   0.020  -3.414  1.00  0.00\\nATOM     30  CA  GLY A   8      -1.991   1.210  -3.929  1.00  0.00\\nATOM     31  C   GLY A   8      -2.347   1.048  -5.402  1.00  0.00\\nATOM     32  O   GLY A   8      -2.959   1.930  -6.003  1.00  0.00\\nATOM     33  N   GLY A   9      -2.084   0.258  -5.794  1.00  0.00\\nATOM     34  CA  GLY A   9      -2.355   0.054  -7.212  1.00  0.00\\nATOM     35  C   GLY A   9      -3.833  -0.220  -7.457  1.00  0.00\\nATOM     36  O   GLY A   9      -4.492  -0.883  -6.656  1.00  0.00\\n\");\n\tviewer_16908637703553598.setStyle({\"model\": -1},{\"cartoon\": {\"colorscheme\": {\"prop\": \"b\", \"map\": {\"0\": \"#FF7D45\", \"1\": \"#FFDB13\", \"2\": \"#65CBF3\", \"3\": \"#0053D6\"}}}, \"stick\": {}});\n\tviewer_16908637703553598.zoomTo();\nviewer_16908637703553598.render();\n});\n</script>",
      "text/html": [
       "<div id=\"3dmolviewer_16908637703553598\"  style=\"position: relative; width: 800px; height: 600px;\">\n",
       "        <p id=\"3dmolwarning_16908637703553598\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
       "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
       "        </div>\n",
       "<script>\n",
       "\n",
       "var loadScriptAsync = function(uri){\n",
       "  return new Promise((resolve, reject) => {\n",
       "    //this is to ignore the existence of requirejs amd\n",
       "    var savedexports, savedmodule;\n",
       "    if (typeof exports !== 'undefined') savedexports = exports;\n",
       "    else exports = {}\n",
       "    if (typeof module !== 'undefined') savedmodule = module;\n",
       "    else module = {}\n",
       "\n",
       "    var tag = document.createElement('script');\n",
       "    tag.src = uri;\n",
       "    tag.async = true;\n",
       "    tag.onload = () => {\n",
       "        exports = savedexports;\n",
       "        module = savedmodule;\n",
       "        resolve();\n",
       "    };\n",
       "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
       "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
       "});\n",
       "};\n",
       "\n",
       "if(typeof $3Dmolpromise === 'undefined') {\n",
       "$3Dmolpromise = null;\n",
       "  $3Dmolpromise = loadScriptAsync('https://cdnjs.cloudflare.com/ajax/libs/3Dmol/2.0.3/3Dmol-min.js');\n",
       "}\n",
       "\n",
       "var viewer_16908637703553598 = null;\n",
       "var warn = document.getElementById(\"3dmolwarning_16908637703553598\");\n",
       "if(warn) {\n",
       "    warn.parentNode.removeChild(warn);\n",
       "}\n",
       "$3Dmolpromise.then(function() {\n",
       "viewer_16908637703553598 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_16908637703553598\"),{backgroundColor:\"white\"});\n",
       "viewer_16908637703553598.zoomTo();\n",
       "\tviewer_16908637703553598.addModelsAsFrames(\"ATOM      1  N   GLY A   1      11.309   0.598   1.723  1.00  0.00\\nATOM      2  CA  GLY A   1      10.408   1.213   2.690  1.00  0.00\\nATOM      3  C   GLY A   1       8.966   0.783   2.451  1.00  0.00\\nATOM      4  O   GLY A   1       8.596   0.411   1.338  1.00  0.00\\nATOM      5  N   GLY A   2       8.077   1.637   2.685  1.00  0.00\\nATOM      6  CA  GLY A   2       6.786   0.962   2.631  1.00  0.00\\nATOM      7  C   GLY A   2       6.405   0.614   1.198  1.00  0.00\\nATOM      8  O   GLY A   2       6.515   1.446   0.297  1.00  0.00\\nATOM      9  N   GLY A   3       6.154  -0.391   1.211  1.00  0.00\\nATOM     10  CA  GLY A   3       5.779  -0.911  -0.098  1.00  0.00\\nATOM     11  C   GLY A   3       4.285  -0.751  -0.347  1.00  0.00\\nATOM     12  O   GLY A   3       3.462  -1.307   0.380  1.00  0.00\\nATOM     13  N   GLY A   4       4.043   0.138  -0.532  1.00  0.00\\nATOM     14  CA  GLY A   4       2.637   0.395  -0.821  1.00  0.00\\nATOM     15  C   GLY A   4       2.365   0.349  -2.319  1.00  0.00\\nATOM     16  O   GLY A   4       1.213   0.287  -2.749  1.00  0.00\\nATOM     17  N   GLY A   5       3.104   0.160  -3.048  1.00  0.00\\nATOM     18  CA  GLY A   5       2.936   0.231  -4.495  1.00  0.00\\nATOM     19  C   GLY A   5       1.816  -0.689  -4.965  1.00  0.00\\nATOM     20  O   GLY A   5       0.993  -0.305  -5.796  1.00  0.00\\nATOM     21  N   GLY A   6       1.920  -1.767  -4.681  1.00  0.00\\nATOM     22  CA  GLY A   6       0.838  -2.646  -5.108  1.00  0.00\\nATOM     23  C   GLY A   6      -0.428  -2.393  -4.299  1.00  0.00\\nATOM     24  O   GLY A   6      -1.539  -2.623  -4.779  1.00  0.00\\nATOM     25  N   GLY A   7      -0.050  -2.160  -3.316  1.00  0.00\\nATOM     26  CA  GLY A   7      -1.269  -1.895  -2.561  1.00  0.00\\nATOM     27  C   GLY A   7      -2.026  -0.705  -3.137  1.00  0.00\\nATOM     28  O   GLY A   7      -3.256  -0.665  -3.105  1.00  0.00\\nATOM     29  N   GLY A   8      -1.325   0.020  -3.414  1.00  0.00\\nATOM     30  CA  GLY A   8      -1.991   1.210  -3.929  1.00  0.00\\nATOM     31  C   GLY A   8      -2.347   1.048  -5.402  1.00  0.00\\nATOM     32  O   GLY A   8      -2.959   1.930  -6.003  1.00  0.00\\nATOM     33  N   GLY A   9      -2.084   0.258  -5.794  1.00  0.00\\nATOM     34  CA  GLY A   9      -2.355   0.054  -7.212  1.00  0.00\\nATOM     35  C   GLY A   9      -3.833  -0.220  -7.457  1.00  0.00\\nATOM     36  O   GLY A   9      -4.492  -0.883  -6.656  1.00  0.00\\n\");\n",
       "\tviewer_16908637703553598.setStyle({\"model\": -1},{\"cartoon\": {\"colorscheme\": {\"prop\": \"b\", \"map\": {\"0\": \"#FF7D45\", \"1\": \"#FFDB13\", \"2\": \"#65CBF3\", \"3\": \"#0053D6\"}}}, \"stick\": {}});\n",
       "\tviewer_16908637703553598.zoomTo();\n",
       "viewer_16908637703553598.render();\n",
       "});\n",
       "</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<py3Dmol.view at 0x7f5c5d3de710>"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PLDDT_BANDS = [\n",
    "  (0, 50, '#FF7D45'),\n",
    "  (50, 70, '#FFDB13'),\n",
    "  (70, 90, '#65CBF3'),\n",
    "  (90, 100, '#0053D6')\n",
    "]\n",
    "view = py3Dmol.view(width=800, height=600)\n",
    "view.addModelsAsFrames(designed_pdb)\n",
    "\n",
    "color_map = {i: bands[2] for i, bands in enumerate(PLDDT_BANDS)}\n",
    "style = {'cartoon': {'colorscheme': {'prop': 'b', 'map': color_map}}}\n",
    "\n",
    "style['stick'] = {}\n",
    "\n",
    "view.setStyle({'model': -1}, style)\n",
    "view.zoomTo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb208d37-fbe8-4e17-ad5f-9521c4ee5523",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
